<html xmlns:concordion="http://www.concordion.org/2007/concordion" xmlns:cc="http://fewi.custom.commands.FewiExtensions">
<head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<style>
	table {border-collapse: collapse; padding-bottom:5px; margin-bottom:5px;}
	table, th, td {border: 1px solid #CCCCCC;}
	td, th {padding:2px;}
	.man {background-color:#FFFFCC;}
	ul {padding: 0px; padding-bottom:3px; padding-top:0px; margin: 0px;}
	ol {padding: 0px; padding-bottom:3px; padding-top:0px; margin: 0px;}
	li {padding-left:2px; padding-top:0px; padding-bottom:0px; margin-left:25px;}
	p {margin:2px 0px;}
	.hilite {background-color:#FFFFE0;}
	td.xl65 {background-color:#FFFF00;}
	td>ul {padding: 0px; padding-bottom:1px; padding-top:0px; margin: 0px;}
	td>ul li {padding-left:2px; padding-top:0px; padding-bottom:0px; margin-left:10px;}
</style>
<!-- declare strict doctype to look for disallowed html errors
 **  **  **  **  ** REMOVE THE LINE before committing the file **  **  **  **  **  **  **
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
 **  **  **  **  ** REMOVE THE LINE before committing the file **  **  **  **  **  **  **
 -->


    <title>Build Sprint Tests for Firien Install</title>
</head>

<body>

<div id="TOC">
	<p>Contents: <a href="#maintitle">Title</a> | <a href="#acceptance">Acceptance Criteria</a> | <a href="#manual">Manual Tests</a> | <a href="#auto">Automated Tests</a></p>
</div>

<div id="maintitle">
    <h1>Build Sprint Tests for Firien Install</h1>
    <ul>
    	<li>User story #18 Allele collections are multi select</li>
    	<li>User story #24, E3: Display allele subtypes on allele detail and cre specificity pages</li>
    	<li>User story #32,</li>
    	<li>User story #35,</li>
    	<li>User story #38,</li>
    	<li>User story #41,</li>
    	<li>User story #81,</li>
        <li>User story #85 Alleles that have transmission status of cell line but have IMSR links get changed to a transmission state of Germ line</li>
        <li>User story #17, E3: Display collection on allele detail and cre specificity pages</li>
        <li>User story #92,</li>
        <li>User story #105, E5: Allele and marker detail pages in prod wi</li>
        <li>User story #110, E5: Remove allele sets from schema (5.18)</li>
        <li>User story #114,</li>
        <li>User story #115,</li>
        <li>User story #122,</li>
        <li>User story #128,</li>
        <li>User story #132,</li>
        <li>User story #156, Caption for Molecular Image Thumbnail</li>
        <li>User story #181,</li>
        <li>User story #184,</li>
        <li>User story #198, E8: Add Australian Phenome Facility(APF) link on Allele Detail page</li>
        
    </ul>
</div>

<div id="acceptance">
    <h2>Acceptance Criteria</h2>
	<ul>
		<li>not needed</li>
	</ul>
</div>



<div id="tests">
    <h2>Tests</h2>
    <p>Skip to <a href="#manual">Manual Tests</a> |  <a href="#auto">Automated Tests</a></p>
	<div id="manual">
<!-- To insert manual test results, use the following:
<p concordion:assertTrue="true">testing assertTrue - displays as green</p>
<p concordion:assertTrue="false">testing assertFalse - displays as red</p>
 -->
		<h3>Manual tests</h3>
		<p>From sto85, Alleles that have transmission status of cell line but have IMSR links get changed to a transmission state of Germ line</p>
		<ol>
    		<li>Alleles having genotypes with MP annotations loaded via the HTMP loader have their Germline Transmission display the text "Earliest citation of germline transmission"
 			<p concordion:assertTrue="false">klf 12/11/2012</p> 
    		<ol>
                <li>Navigate to the Allele Detail Page for  Acap3&lt;tm1a(EUCOMM)Hmgu&gt;, MGI:4436565 (previously had germ line transmission status = cell line).</li>
                <li>Verify the WI field "Germline Transmission" displays <b>Earliest citation of germline transmission: </b></li>
    		</ol>
    		</li>
    		<li>Alleles having their transmission status set to Germline via this load (above) have the appropriate (Sanger or Europhenome) load J:# associated with the transmission status, and it is linked to the correct References page.<br/>
    		<i>Example: Earliest citation of germline transmission: J:175295</i> 
    		<ol>
                <li><span class="regressHi">Regress</span> Navigate to the Allele Detail Page for Acap3&lt;tm1a(EUCOMM)Hmgu&gt;, MGI:4436565 (previously had germ line transmission status = cell line).</li>
                <li><span concordion:assertTrue="false">klf 12/11/2012</span> - Verify the WI field "Germline Transmission" displays <b>Earliest citation of germline transmission: J:165965</b>.</li>
                <li><span concordion:assertTrue="false">klf 12/11/2012</span> - Verify clicking the link opens the page at <b>http://www.informatics.jax.org/reference/J:165965</b>.</li>
                <li><span class="regressHi">Regress</span> Navigate to the Allele Detail Page for  Adam17&lt;tm1a(EUCOMM)Wtsi&gt;, MGI:4434074 (previously had germ line transmission status = cell line).</li>
                <li><span concordion:assertTrue="false">klf 12/11/2012</span> - Verify the WI field "Germline Transmission" displays <b>Earliest citation of germline transmission: J:175295</b>.</li>
                <li>Verify clicking the link opens the page at <b>http://www.informatics.jax.org/reference/J:175295</b>.</li>
    		</ol>
    		</li>
    		<li>A. Alleles previously having the transmission status  &quot;Germline&quot;  in the EI (displaying a reference J:# and phrase in the WI) do not change.
 			<p>WI displays existing data for the following alleles. They should not change.</p>
    		<ol>
                <li><span class="regressHi">Regress</span> Navigate to the Allele Detail Page for  Apoe&lt;tm1Unc&gt;, MGI:1857129.</li>
                <li><span concordion:assertTrue="false">klf 12/11/2012</span> - Verify the WI field "Germline Transmission" displays <b>Earliest citation of germline transmission: J:1050</b>.</li>
                <li><span concordion:assertTrue="false">klf 12/11/2012</span> - Verify clicking the link opens the page at <b>http://www.informatics.jax.org/reference/J:1050</b>.</li>
                <li>Navigate to the Allele Detail Page for B9d1&lt;tm1a(EUCOMM)Wtsi&gt;, MGI:4432284</li>
                <li><span concordion:assertTrue="false">passed - kstone 2012-12-11</span> - <b>Earliest citation of germline transmission: J:178421.</b></li>
<!--                 <li>Navigate to the Allele Detail Page for Csrp2bp&lt;tm1a(KOMP)Wtsi&gt;, MGI:4363521.</li>
                <li><span concordion:assertTrue="false">new</span> - <b>Earliest citation of germline transmission: J:164007.</b></li>
                <li>Navigate to the Allele Detail Page for Dkk3&lt;tm1Cni&gt;, MGI:3621929.</li>
                <li><span concordion:assertTrue="false">new</span> - <b>Earliest citation of germline transmission: J:106922.</b></li>
                <li>Navigate to the Allele Detail Page for Mysm1&lt;tm1a(KOMP)Wtsi&gt;, MGI:4362924.</li>
                <li><span concordion:assertTrue="false">new</span> - <b>Earliest citation of germline transmission: J:179287.</b></li>
 -->
            </ol>
            <p>B. Alleles with the transmission status "Not Applicable" do not change.</p>
            <ol>
                <li><span class="regressHi">Regress</span> Navigate to the Allele Detail Page for Ankrd11&lt;Yod&gt;</li>
                <li><span concordion:assertTrue="false">passed - kstone 2012-12-11</span> - In the EI, verify the transmission status is <b>Not Applicable</b> and <b>no Transmission reference exists</b>.</li>
                <li><span concordion:assertTrue="false">passed - kstone 2012-12-11</span> - Verify the WI field "Germline Transmission" <b>does not appear</b> and the EI says "Not Applicable" .</li>
<!--                 <li>Navigate to the Allele Detail Page for Cdh23&lt;v&gt;, MGI:1856228.</li>
                <li><span concordion:assertTrue="false">new</span> - In the EI, verify the transmission status is <b>Not Applicable</b> and <b>no Transmission reference exists</b>.</li>
                <li><span concordion:assertTrue="false">new</span> - Verify the WI field "Germline Transmission" <b>does not appear</b> and the EI says "Not Applicable" .</li>
                <li>Navigate to the Allele Detail Page for Ctnnb1&lt;Bfc&gt;, MGI:2656734.</li>
                <li><span concordion:assertTrue="false">new</span> - In the EI, verify the transmission status is <b>Not Applicable</b> and <b>no Transmission reference exists</b>.</li>
                <li><span concordion:assertTrue="false">new</span> - Verify the WI field "Germline Transmission" <b>does not appear</b> and the EI says "Not Applicable" .</li>
                <li>Navigate to the Allele Detail Page for Mecom&lt;Jbo&gt;, MGI:2158381.</li>
                <li><span concordion:assertTrue="false">new</span> - In the EI, verify the transmission status is <b>Not Applicable</b> and <b>no Transmission reference exists</b>.</li>
                <li><span concordion:assertTrue="false">new</span> - Verify the WI field "Germline Transmission" <b>does not appear</b> and the EI says "Not Applicable" .</li>
                <li>Navigate to the Allele Detail Page for Myo6&lt;sv&gt;, MGI:1856555.</li>
                <li><span concordion:assertTrue="false">new</span> - In the EI, verify the transmission status is <b>Not Applicable</b> and <b>no Transmission reference exists</b>.</li>
                <li><span concordion:assertTrue="false">new</span> - Verify the WI field "Germline Transmission" <b>does not appear</b> and the EI says "Not Applicable" .</li>
 -->
    		</ol>
    		<p>C. Alleles with the transmission status "Chimeric" do change.</p>
    		<ol>
    			<li>Lori set Zfp715&lt;tm1a(EUCOMM)Hmgu&gt; to have status "Chimeric". (KLF verified around noon on 12/3/2012.) </li> 
    			<li><span concordion:assertTrue="false">klf 13:39 12/3/2012.</span> - Verify load changes it to "Germline".</li>
    			<li><span concordion:assertTrue="false">passed - kstone 2012-12-11</span> - Navigate to Allele Detail Page and verify WI field "Germline Transmission" displays <b>Earliest citation of germline transmission: J:165965</b>.</li> 
    		</ol>
    		<p>D. Alleles with the transmission status "Not Specified" do not change.</p>
    		<ol>
    			<li>There are only a handful of these alleles and all are "Reserved" or "In Progress" Test not needed.</li>
    		</ol>
    		</li>
    		<li><span class="regressHi">Regress</span> Collection appears in Mutation Origin ribbon on Allele Detail Pages.(sto17)
 			<p concordion:assertTrue="false">passed. klf 2013-12-31</p>
    		<ol>
                <li>Navigate to the Allele Detail page for <b>0610037L13Rik&lt;tm1(KOMP)Vlcg&gt;</b> (MGI:3847523)</li>
                <li>Verify that the Collection value <b>KOMP-Regeneron</b> appears as the last item in the Mutation Origin ribbon.</li>
                <li>Navigate to the Allele Detail page for <b>Acan&lt;b2b183Clo&gt;</b> (MGI:5297419)</li>
                <li>Verify that the Collection value <b>B2B/CvDC</b>  appears as the last item in the Mutation Origin ribbon.</li>
                <li>Navigate to the Allele Detail page for <b>Adamts20&lt;hip&gt;</b> (MGI:3847117)</li>
                <li>Verify that the Collection value <b>Australian PhenomeBank</b>  appears as the last item in the Mutation Origin ribbon.</li>
                <li>Navigate to the Allele Detail page for <b>Dbx1&lt;tm1.1(cre)Mull&gt;</b> (MGI:5014205)</li>
                <li>Verify that the Collection value <b>Neuroscience Blueprint cre</b>  appears as the last item in the Mutation Origin ribbon.</li>
                <li>Navigate to the Allele Detail page for <b>Tg(ACTFLPe)9205Dym</b> (MGI:2448985)</li>
                <li>Verify that the Collection value <b>Neuroscience Blueprint cre</b> appears as the last item in the Transgene Origin ribbon.</li>
    		</ol>
    		</li>

    		<li><span class="regressHi">Regress</span> Collection does not appear when the allele is not assigned to a collection.(sto85)
 			<p concordion:assertTrue="false">passed. klf 2013-12-31</p>
    		<ol>
                <li>Navigate to the Allele Detail page for <b>Pax6&lt;2Neu&gt;</b> (MGI:2156812)</li>
                <li>Verify that <b>no Collection value</b> appears in the Mutation Origin ribbon.</li>
                <li>Navigate to the Allele Detail page for <b>Tg(GFAP-cre)#Gtm</b> (MGI:3057344)</li>
                <li>Verify that <b>no Collection value</b> appears in the Transgene Origin ribbon.</li>
    		</ol>
    		</li>

    		<li>GO Graphs appear properly. - CANNOT TEST THIS ONE UNTIL BUILD SPRINT. NOT INSTALLED ON VALINOR/CARDOLAN. (from sto110, E5: Remove allele sets from schema)
 			<p concordion:assertTrue="false">new March 2014</p>
    		<ol>

                <li>Navigate to the Pax6 Marker Detail Page.</li>
                <li>Follow the link for GO Classifications.</li>
                <li>Follow the link for GO Graphical View.</li>
                <li>Verify the GO Graph renders in the page with text and table data, as on Public.</li>
    		</ol>
    		</li>
    		<li><span class="regressHi">Regress</span> Allele subtypes appear in Mutation Description ribbon on Allele Detail Pages 
    			as a comma separated list following the allele type. (sto24)<br/> The types but not the subtypes are supported on Prod. (sto105)
    			<p concordion:assertTrue="true">passed - klf 2014-03-10</p> 
    		<ol>
                <li>Navigate to the Allele Detail page for Kdr&lt;tm1Jrt&gt; (MGI:1857168) (<a href="http://cardolan.informatics.jax.org/allele/MGI:1857168">link</a>)</li>
                <li>Verify that the subtype value <b>(Null/knockout, Reporter)</b> appears after Allele Type <b>Targeted</b> in the Mutation Description ribbon.</li>
                <li>Navigate to the Allele Detail page for Olig2&lt;tm1(cre/Esr1*)Htak&gt; (MGI:2183410) (<a href="http://cardolan.informatics.jax.org/allele/MGI:2183410">link</a>)</li>
                <li>Verify that the subtype value <b>(Inducible, Recombinase)</b> appears after Allele Type <b>Targeted</b> in the Mutation Description ribbon.</li>
                <li>Navigate to the Allele Detail page for Tg(BCL2)1TSK (MGI:1891501) (<a href="http://cardolan.informatics.jax.org/allele/MGI:1891501">link</a>)</li>
                <li>Verify that the subtype value <b>(Inserted expressed sequence)</b> appears after Transgene Type <b>Transgenic</b> in the Transgene Description ribbon.</li>
    		</ol>
    		</li>

    		<li><span class="regressHi">Regress</span> Allele Subtypes do not appear when the allele is not assigned to a subtype. (sto24)<br/>
    		Also true on Prod. (sto105)
 			<p concordion:assertTrue="false">passed - klf 2014-03-10</p> 
    		<ol>
                <li>Navigate to the Allele Detail page for Lith20&lt;SM/J&gt; (MGI:3577251) (<a href="http://cardolan.informatics.jax.org/allele/MGI:3577251">link</a>)</li>
                <li>Verify that <b>no subtype value</b> appears in the Mutation Description ribbon following the Allele Type: QTL.</li>
                <li>Navigate to the Allele Detail page for Dst&lt;Tg4&gt; (MGI:1856421) (<a href="http://cardolan.informatics.jax.org/allele/MGI:1856421">link</a>)</li>
                <li>Verify that <b>no subtype value</b> appears in the Transgene description ribbon following the Transgene Type: Transgenic.</li>
    		</ol>
    		</li>
    		<li>The correct image/caption is displayed to the left of the molecular image in the molecular description ribbon when more than one thumbnail exists for an allele.  (sto156)
 			<p concordion:assertTrue="false">passed jeffc 04/15/2014</p>
    		<ol>
                <li>Navigate to the Allele Detail page for Dock2&lt;mu&gt; (MGI:5319388) (<a href="http://cardolan.informatics.jax.org/allele/MGI:5319388">link</a>)</li>
                <li>Verify that to the left of the molecular image is displayed the thumbnail caption</li>
                <li>The caption should read <b>Schematic showing the duplication and location of the premature stop codon in the Dock2&lt;mu&gt; allele found in Irf5&lt;tm1Ttg&gt;/Irf5&lt;tm1Ttg&gt; mice</b></li>
                <li>The image and caption should be for the thumbnail J;182658; thumbnail; 3b which can be verified in the image module of the EI</li>
    		</ol>
    		</li>
    		<li>Verify these Alleles have the Australian Phenome Facility link in the Mutation Description ribbon(under  Mutation Details) and the link works. (sto198)
 			<p concordion:assertTrue="false">Passed jeffc 04/08/2014</p>
    		<ol>
                <li>Navigate to these allele detail pages.</li>
                <li>Verify the Australian Phenome Facility link exists and works:
                <table>
                	<tr><th>Allele Symbol</th><th>MGI ID</th></tr>
					<tr><td>Adamts20&lt;hip&gt;</td><td>MGI:3847117</td></tr>
					<tr><td>Nphp3&lt;pol&gt;</td><td>MGI:3847130</td></tr>
                </table>
                </li>
    		</ol>
    		</li>    		
    		<li>Verify that allele subtypes are displayed in the category column after the generated type (if there are subtypes)
 			<p concordion:assertTrue="false">passed - jeffc 03/10/2014</p> (sto24)
    		<ol>
    			<li>Do a nomenclature query for <a href="http://cardolan.informatics.jax.org/allele/summary?nomen=grhpr">Grhpr</a></li>
    			<li>Verify that allele <b>Grhpr&lt;Gt(OST383093)Lex&gt;</b> has only <b>Gene trapped</b> in the Category column. <i>It has no subtypes or transmission status display.</i></li>
    			<li>Verify that allele <b>Grhpr&lt;tm1(KOMP)Vlcg&gt;</b> has only <b>Targeted</b> 
    				followed by <b>(Null/knockout, Reporter)</b>  
    				followed by <b>(Cell Line)</b> in the Category column. </li>
    		</ol>
    		</li>    		
			<li>The allele report can be downloaded in tab-delimited format.(sto41)
 			<p concordion:assertTrue="false">passed. jeffc 03/10/2014</p>
    		<ol>
                <li>Submit a query for <b>chrX</b>.</li>
                <li>Click the button labeled "Text File".</li>
                <li>Verify the downloaded file can be opened and is in tab-delimited format.</li>
                <li>Verify that the same number of alleles on the Allele Summary appears in the downloaded file (9,228 alleles on 1/27/2014).</li>
                <li>Verify that the downloaded file contains the following columns:
                <table>
                	<tr><th>Column</th></tr>
                	<tr><td>MGI Allele ID</td></tr>
                	<tr><td>Allele symbol</td></tr>
                	<tr><td>Allele name</td></tr>
                	<tr><td>Chromosome</td></tr>
                	<tr><td>Synonyms</td></tr>
                	<tr><td>Allele type</td></tr>
                	<tr><td>Allele attributes</td></tr>
                	<tr><td>Transmission</td></tr>
                	<tr><td>Abnormal phenotypes reported in these systems</td></tr>
                	<tr><td>Human Disease Models</td></tr>
                </table>
                </li>
                <li>Verify that the following alleles exist in the downloaded file, with associated info:
                <p>NOTE: the last allele, Zrsr2&lt;tm1a(EUCOMM)Wtsi&gt;, is &quot;Deleted&quot; so it will not exist in Public data.</p>
                <table>
				<tr>
					<th>MGI Allele ID</th>
					<th>Allele Symbol</th>
					<th>Allele Name</th>
					<th>Chromosome</th>
					<th>Synonyms</th>
					<th>Allele type</th>
					<th>Allele attributes</th>
					<th>Transmission</th>
					<th>Abnormal phenotypes reported in these systems</th>
					<th>Human Disease Models</th>
				</tr>
				<tr>
					<td>MGI:3628655</td>
					<td>Abcb7&lt;tm1.1Mdf&gt;</td>
					<td>targeted mutation 1.1, Mark D Fleming</td>
					<td>X</td>
					<td>Abcb7&lt;-&gt;</td>
					<td>Targeted</td>
					<td>Null/knockout</td>
					<td>Germline</td>
					<td>cellular | mortality/aging</td>
					<td></td>
				</tr>
				<tr>
					<td>MGI:3628654</td>
					<td>Abcb7&lt;tm1Mdf&gt;</td>
					<td>targeted mutation 1, Mark D Fleming</td>
					<td>X</td>
					<td></td>
					<td>Targeted</td>
					<td>Conditional ready</td>
					<td>Germline</td>
					<td>cardiovascular | embryogenesis | growth/size | homeostasis | liver/biliary | mortality/aging</td>
					<td></td>
				</tr>
				<tr>
					<td>MGI:4433571</td>
					<td>Abcd1&lt;tm1a(EUCOMM)Wtsi&gt;</td>
					<td>targeted mutation 1a, Wellcome Trust Sanger Institute</td>
					<td>X</td>
					<td></td>
					<td>Targeted</td>
					<td>Conditional ready, Null/knockout, Reporter</td>
					<td>Germline</td>
					<td>hematopoietic | other | vision/eye</td>
					<td></td>
				</tr>
				<tr>
					<td>MGI:2179253</td>
					<td>Aggr2&lt;A/J&gt;</td>
					<td>A/J</td>
					<td>X</td>
					<td></td>
					<td>QTL</td>
					<td></td>
					<td></td>
					<td>behavior</td>
					<td></td>
				</tr>
				<tr>
					<td>MGI:4129850</td>
					<td>Akap17b&lt;Gt(RRI131)Byg&gt;</td>
					<td>gene trap RRI131, BayGenomics</td>
					<td>X</td>
					<td></td>
					<td>Gene trapped</td>
					<td></td>
					<td>Chimeric</td>
					<td>craniofacial | embryogenesis | growth/size | nervous system</td>
					<td></td>
				</tr>
				<tr>
					<td>MGI:3034098</td>
					<td>Ar&lt;tm1Verh&gt;</td>
					<td>targeted mutation 1, Guido Verhoeven</td>
					<td>X</td>
					<td>AR&lt;fl&gt; | ARflox | AR&lt;flox&gt; | AR&lt;loxP-exon2-loxP&gt;</td>
					<td>Targeted</td>
					<td>Conditional ready</td>
					<td>Germline</td>
					<td>adipose | cellular | endocrine/exocrine | growth/size | homeostasis | immune | reproductive | tumorigenesis</td>
					<td>Diabetes Mellitus, Noninsulin-Dependent; NIDDM 125853 | Prostate Cancer 176807</td>
				</tr>
				<tr>
					<td>MGI:4880890</td>
					<td>Zmat1&lt;tm1a(KOMP)Wtsi&gt;</td>
					<td>targeted mutation 1a, Wellcome Trust Sanger Institute</td>
					<td>X</td>
					<td></td>
					<td>Targeted</td>
					<td>Conditional ready, Null/knockout, Reporter</td>
					<td>Cell Line</td>
					<td></td>
					<td></td>
				</tr>
				<tr>
					<td>MGI:4197613</td>
					<td>Zmym3&lt;Gt(OST130993)Lex&gt;</td>
					<td>gene trap OST130993, Lexicon Genetics</td>
					<td>X</td>
					<td></td>
					<td>Gene trapped</td>
					<td></td>
					<td>Cell Line</td>
					<td></td>
					<td></td>
				</tr>
				<tr>
					<td>MGI:4842786</td>
					<td>Zrsr2&lt;tm1a(EUCOMM)Wtsi&gt;</td>
					<td>targeted mutation 1a, Wellcome Trust Sanger Institute</td>
					<td>X</td>
					<td></td>
					<td>Targeted</td>
					<td>Conditional ready, Null/knockout, Reporter</td>
					<td>Cell Line</td>
					<td></td>
					<td></td>
				</tr>
                </table>
                </li>
    		</ol>			
			</li>

			<li>The allele report can be downloaded in Excel format.(sto41)
 			<p concordion:assertTrue="false">passed. -jeffc 03/10/2014</p>
    		<ol>
                <li>Submit a query for <b>chrX</b>.</li>
                <li>Click the button labeled "Excel File".</li>
                <li>Verify the downloaded file can be opened with Excel.</li>
                <li>Verify that the same number of alleles on the Allele Summary appears in the downloaded file (9,228 alleles on 1/27/2014).</li>
                <li>Verify that the downloaded file contains the following columns (SAME AS ABOVE):
                <table>
                	<tr><th>Column</th></tr>
                	<tr><td>MGI Allele ID</td></tr>
                	<tr><td>Allele symbol</td></tr>
                	<tr><td>Allele name</td></tr>
                	<tr><td>Chromosome</td></tr>
                	<tr><td>Synonyms</td></tr>
                	<tr><td>Allele type</td></tr>
                	<tr><td>Allele attributes</td></tr>
                	<tr><td>Transmission</td></tr>
                	<tr><td>Abnormal phenotypes reported in these systems</td></tr>
                	<tr><td>Human Disease Models</td></tr>
                </table>
                </li>
                <li>Verify that the following alleles exist in the downloaded file, with associated info (SAME AS ABOVE):
                <p>NOTE: the last allele, Zrsr2&lt;tm1a(EUCOMM)Wtsi&gt;, is &quot;Deleted&quot; so it will not exist in Public data.</p>
                <table>
				<tr>
					<th>MGI Allele ID</th>
					<th>Allele Symbol</th>
					<th>Allele Name</th>
					<th>Chromosome</th>
					<th>Synonyms</th>
					<th>Allele type</th>
					<th>Allele attributes</th>
					<th>Transmission</th>
					<th>Abnormal phenotypes reported in these systems</th>
					<th>Human Disease Models</th>
				</tr>
				<tr>
					<td>MGI:3628655</td>
					<td>Abcb7&lt;tm1.1Mdf&gt;</td>
					<td>targeted mutation 1.1, Mark D Fleming</td>
					<td>X</td>
					<td>Abcb7&lt;-&gt;</td>
					<td>Targeted</td>
					<td>Null/knockout</td>
					<td>Germline</td>
					<td>cellular | mortality/aging</td>
					<td></td>
				</tr>
				<tr>
					<td>MGI:3628654</td>
					<td>Abcb7&lt;tm1Mdf&gt;</td>
					<td>targeted mutation 1, Mark D Fleming</td>
					<td>X</td>
					<td></td>
					<td>Targeted</td>
					<td>Conditional ready</td>
					<td>Germline</td>
					<td>cardiovascular | embryogenesis | growth/size | homeostasis | liver/biliary | mortality/aging</td>
					<td></td>
				</tr>
				<tr>
					<td>MGI:4433571</td>
					<td>Abcd1&lt;tm1a(EUCOMM)Wtsi&gt;</td>
					<td>targeted mutation 1a, Wellcome Trust Sanger Institute</td>
					<td>X</td>
					<td></td>
					<td>Targeted</td>
					<td>Conditional ready, Null/knockout, Reporter</td>
					<td>Germline</td>
					<td>hematopoietic | other | vision/eye</td>
					<td></td>
				</tr>
				<tr>
					<td>MGI:2179253</td>
					<td>Aggr2&lt;A/J&gt;</td>
					<td>A/J</td>
					<td>X</td>
					<td></td>
					<td>QTL</td>
					<td></td>
					<td></td>
					<td>behavior</td>
					<td></td>
				</tr>
				<tr>
					<td>MGI:4129850</td>
					<td>Akap17b&lt;Gt(RRI131)Byg&gt;</td>
					<td>gene trap RRI131, BayGenomics</td>
					<td>X</td>
					<td></td>
					<td>Gene trapped</td>
					<td></td>
					<td>Chimeric</td>
					<td>craniofacial | embryogenesis | growth/size | nervous system</td>
					<td></td>
				</tr>
				<tr>
					<td>MGI:3034098</td>
					<td>Ar&lt;tm1Verh&gt;</td>
					<td>targeted mutation 1, Guido Verhoeven</td>
					<td>X</td>
					<td>AR&lt;fl&gt; | ARflox | AR&lt;flox&gt; | AR&lt;loxP-exon2-loxP&gt;</td>
					<td>Targeted</td>
					<td>Conditional ready</td>
					<td>Germline</td>
					<td>adipose | cellular | endocrine/exocrine | growth/size | homeostasis | immune | reproductive | tumorigenesis</td>
					<td>Diabetes Mellitus, Noninsulin-Dependent; NIDDM 125853 | Prostate Cancer 176807</td>
				</tr>
				<tr>
					<td>MGI:4880890</td>
					<td>Zmat1&lt;tm1a(KOMP)Wtsi&gt;</td>
					<td>targeted mutation 1a, Wellcome Trust Sanger Institute</td>
					<td>X</td>
					<td></td>
					<td>Targeted</td>
					<td>Conditional ready, Null/knockout, Reporter</td>
					<td>Cell Line</td>
					<td></td>
					<td></td>
				</tr>
				<tr>
					<td>MGI:4197613</td>
					<td>Zmym3&lt;Gt(OST130993)Lex&gt;</td>
					<td>gene trap OST130993, Lexicon Genetics</td>
					<td>X</td>
					<td></td>
					<td>Gene trapped</td>
					<td></td>
					<td>Cell Line</td>
					<td></td>
					<td></td>
				</tr>
				<tr>
					<td>MGI:4842786</td>
					<td>Zrsr2&lt;tm1a(EUCOMM)Wtsi&gt;</td>
					<td>targeted mutation 1a, Wellcome Trust Sanger Institute</td>
					<td>X</td>
					<td></td>
					<td>Targeted</td>
					<td>Conditional ready, Null/knockout, Reporter</td>
					<td>Cell Line</td>
					<td></td>
					<td></td>
				</tr>
                </table>
                </li>
    		</ol>			
			</li>
			<li>Verify a query by coordinates returns the same alleles as on Public(sto32)(sto122)
 			<p concordion:assertTrue="false">passed -jeffc 03-10-2014</p>
    		<ol>
                <li>Submit a query for <b>chr12, between 56691767-56712822 bp</b></li>
                <li>Verify <b>18 alleles</b> appear for the markers Eccp, Pax9, Slc25a21, Splwt1, Tbqt3, and Tmc1m3.</li>
    		</ol>
			</li>

    		<li>Searching by chromosome returns the correct alleles now we have moved it to the FeWI (sto38)
 			<p concordion:assertTrue="true">passed -jeffc 03-10-2014</p>
    		<table>
	    		<tr><th>Chromosome</th><th>Python Result<br/><a href="http://cardolan.informatics.jax.org/searches/allele_form.shtml">query form</a></th><th>FEWI Result<br/><a href="http://cardolan.informatics.jax.org/allele/">query form</a></th></tr>
	    		<tr><td>18</td><td>11089 </td><td>11100</td></tr>
	    		<tr><td>19</td><td>11653 </td><td>11671</td></tr>
	    		<tr><td>Y</td><td>150 </td><td>150</td></tr>
	    		<tr><td>19,Y</td><td> </td><td>11800</td></tr>
	    		<tr></tr>
    		</table>
    		</li>
    		<li>Seaching with cm Position returns the correct alleles. cM values may be separated by a space or a dash. (revised for sto81)
 			<p concordion:assertTrue="true">passed. -jeffc 03-10-2014</p>
    		<table>
	    		<tr><th>Chromosome</th><th>cM</th><th>Python Result<br/><a href="http://cardolan.informatics.jax.org/searches/allele_form.shtml">query form</a></th><th>FEWI Result<br/><a href="http://cardolan.informatics.jax.org/allele/">query form</a></th></tr>
	    		<tr><td>19</td><td>50 - 75</td><td>507</td><td>507</td></tr>
	    		<tr><td>X</td><td>30-50</td><td>2770</td><td>2770</td></tr>
    		</table>
    		</li>
    		
    		<li>Searching with Genome Coordinates returns the correct alleles. Genome coordinate values may be separated by a space or a dash. (revised for sto81)
 			<p concordion:assertTrue="false">passed. -jeffc 03-10-2014</p>
    		<table>
	    		<tr><th>Chromosome</th><th>Coordinates</th><th></th><th>Python Result</th><th>FEWI Result</th></tr>
	    		<tr><td>3</td><td>(&gt;=) 125618205<br/>125618205-999999999</td><td>bp</td><td>4330</td><td>4330</td></tr>
	    		<tr><td>3</td><td>(&lt;=) 125618205<br/>1-125618205</td><td>bp</td><td>12768</td><td>12768</td></tr>
	    		<tr><td>3</td><td> 125618205 - 125622026</td><td>bp</td><td>11</td><td>11</td></tr>
	    		<tr><td>3</td><td> 125.618 - 125.622</td><td>Mbp</td><td>11</td><td>11</td></tr>
	    		<tr><td>3</td><td> 125.618 - 99999</td><td>Mbp</td><td>4330</td><td>4330</td></tr>
	    		<tr><td>3</td><td> 125.618..99999</td><td>Mbp</td><td>4330</td><td>4330</td></tr>
    		</table>
			</li>

			<li>Searching by MP or OMIM ID excludes alleles associated with "normal" and "Not" annotations. (sto35)
			<p concordion:assertTrue="false">passed. -jeffc 03-10-2014</p>
    		<table>
	    		<tr>
	    			<th>ID</th>
	    			<th>Term</th>
	    			<th>EXCLUDED Alleles</th>
	    			<th>Notes to testers</th>
	    		</tr>
	    		<tr>
	    			<td>MP:0005394</td>
	    			<td>taste/olfaction phenotype</td>
	    			<td>
	    			<ul>
	    				<li>Syn1&lt;tm1Lli&gt;</li>
	    				<li>Syn2&lt;tm1Kao&gt;</li>
	    				<li>Syn3&lt;tm1Jfe&gt;</li>
	    			</ul>
	    			</td>
	    			<td>The markers Syn1, Syn2, and Syn3 should be absent.  No other alleles of these markers are annotated  to this phenotype term or its children.</td>
	    		</tr>
	    		<tr>
	    			<td>100800</td>
	    			<td>Achondroplasia; ACH</td>
	    			<td>Fgfr3&lt;tm1Dor&gt;</td>
	    			<td>Other alleles of Fgfr3 should be displayed.</td>
	    		</tr>
	    		<tr>
	    			<td>102200</td>
	    			<td>Pituitary Adenoma, Growth Hormone-Secreting</td>
	    			<td>Fbn1&lt;Tsk&gt;</td>
	    			<td>The marker Fbn1 should be absent; no other Fbn1 alleles are annotated to this disease.</td>
	    		</tr>
    		</table>
			</li>

    		<li>Querying by Phenotype returns alleles annotated to the term via any kind of genotype (OMG, Conditional, or Non-Conditional Complex).<br/>
    		The affected systems column displays a comma separated list of systems for all OMG or Conditional genotypes of each allele, but not systems associated with Non-Conditional Complex genotypes. (sto115)  
 			<p concordion:assertTrue="false">passed. -jeffc 03-10-2014</p>  
    		<ol>
    			<li>
    			<table>
	    			<tr>
	    				<th>Query for...</th>
	    				<th>Allele</th>
	    				<th>Return allele? (Y/N)</th>
	    				<th>"Affected System" terms</th>
	    				<th>Display "Affected System" terms? (Y/N)</th>
	    				<th>notes to tester</th>
    				</tr>
    				<tr>
    					<td>decreased splenocyte proliferation <b>MP:0009335</b></td>
    					<td>Hprt&lt;b-m3&gt;</td>
    					<td>yes</td>
    					<td>behavior/neurological, nervous system</td>
    					<td>yes</td>
    					<td>queried term is under hematopoietic system and immune system but annotation is to COMPLEX genotype; allele also annotated to &quot;phenotypic reversion&quot; as OMG het</td>
    				</tr>
    				<tr>
    					<td>decreased splenocyte proliferation <b>MP:0009335</b></td>
    					<td>Hprt&lt;b-m3&gt;</td>
    					<td>yes</td>
    					<td>cellular, hematopoietic system, immune system</td>
    					<td>no</td>
    					<td>queried term is under hematopoietic system and immune system but annotation is to COMPLEX genotype; allele also annotated to &quot;phenotypic reversion&quot; as OMG het</td>
    				</tr>
    				<tr>
    					<td>alveolitis <b>MP:0010922</b> </td>
    					<td>Apoe&lt;tm1Unc&gt;</td>
    					<td>yes</td>
    					<td>cardiovascular system, hematopoietic system, immune system, respiratory system</td>
    					<td>yes</td>
    					<td>annotation for the queried term is to a 3-marker CONDITIONAL genotype; MANY other affected systems also displayed</td>
    				</tr>
    				<tr>
    					<td>alveolitis <b>MP:0010922</b> </td>
    					<td>Apoe&lt;tm1Unc&gt;</td>
    					<td>yes</td>
    					<td>limb/digits/tail, other phenotype, reproductive system</td>
    					<td>no</td>
    					<td>annotation for the queried term is to a 3-marker CONDITIONAL genotype; but annotations for this allele in these 3 systems involve only complex genotypes</td>
    				</tr>
    				<tr>
    					<td>abnormal homeostasis <b>MP:0001764</b> </td>
    					<td>Lrp1&lt;tm2Her&gt;</td>
    					<td>yes</td>
    					<td>adipose tissue, behavior/neurological, cardiovascular system, growth/size, hematopoietic system, homeostasis/metabolism, immune system, liver/biliary system, mortality/aging, muscle, nervous system, other phenotype, reproductive system, skeleton</td>
    					<td>yes</td>
    					<td>all genotypes of this allele are CONDITIONAL</td>
    				</tr>
    			</table>
    			</li>
    		</ol>
    		</li>
   		<li>Verify that MP and OMIM annotations associated with <b>Complex</b> (multiple allele pair), not conditionally-targeted genotypes are suppressed in the display, but that searching for such annotations <b>still does return</b> the allele.(sto32) (updated for sto115)
    		<p concordion:assertTrue="true">passed. -jeffc 03-10-2014</p>
    		<ol>
                <li>For the allele <b>Kras<sup>tm1Mok</sup></b>, do not display MP header terms associated with complex (multi-gene) genotypes unless they are ALSO associated with a simple or a conditional genotype.<br/>
                Do not display:
                <ul>
                	<li>homeostasis</li>
                	<li>immune</li>
                </ul>
                Do display:
                <ul>
                	<li>cardiovascular</li>
                	<li>cellular</li>
                	<li>embryogenesis</li>
                	<li>growth/size</li>
                	<li>mortality/aging</li>
                	<li>muscle</li>
                	<li>nervous system</li>
                </ul> 
                </li>
                <li>For the allele <b>Kras<sup>tm1Tyj</sup></b>,<br/>
                Do not display:
                <ul>
                	<li>limbs/digits/tail</li>
                	<li>vision/eye</li>
                </ul>
                Do display:
                <ul>
                	<li>cardiovascular</li>
                	<li>cellular</li>
                	<li>embryogenesis</li>
                	<li>growth/size</li>
                	<li>hematopoietic</li>
                	<li>homeostasis</li>
                	<li>integument</li>
                	<li>liver/biliary</li>
                	<li>mortality/aging</li>
                </ul>
                </li>
    		</ol>
    		</li>

    		<li>Verify that MP and OMIM associations with <b>Compound</b> genotypes are suppressed in the display, but that searching for such annotations <b>still does return</b> the allele.(sto32)  (updated for sto115)
    		<p concordion:assertTrue="true">passed -jeffc 03-10-2014</p>
    		<ol>
                <li>For the allele <b>Del(2)49Ddu</b>, verify that <b>no data appears</b> in the columns for Affected Analytical Systems or Similar Human Diseases</li>
    		</ol>
    		</li>
    		
			<li>Searching by MP or OMIM ID includes alleles from appropriate genotypes. (sto35) (updated for sto115)
			<p concordion:assertTrue="true">passed. -jeffc 03-10-2014</p>
    		<table>
	    		<tr>
	    			<th>ID</th>
	    			<th>Term</th>
	    			<th>Returned Alleles</th>
	    			<th>Notes to testers</th>
	    		</tr>
	    		<tr>
	    			<td>135100</td>
	    			<td>Fibrodysplasia Ossificans Progressiva; FOP</td>
	    			<td>Acvr1&lt;tm1Emsh&gt;</td>
	    			<td>OMG het/wt.</td>
	    		</tr>
	    		<tr>
	    			<td>135100</td>
	    			<td>Fibrodysplasia Ossificans Progressiva; FOP</td>
	    			<td>Tg(Eno2-Bmp4)3Jake</td>
	    			<td>&quot;Simple&quot; transgene in a hemizygous genotype.</td>
	    		</tr>
	    		<tr>
	    			<td>135100</td>
	    			<td>Fibrodysplasia Ossificans Progressiva; FOP</td>
	    			<td>Tg(CAG-LacZ,-ACVR1*,-EGFP)35-1Mis</td>
	    			<td>This transgene includes a reporter as well as a mutant protein.</td>
	    		</tr>
	    		<tr>
	    			<td>137580</td>
	    			<td>Gilles De La Tourette Syndrome; GTS</td>
	    			<td>Slitrk1&lt;tm1Jaru&gt;</td>
	    			<td>OMG hom.</td>
	    		</tr>
	    		<tr>
	    			<td>137580</td>
	    			<td>Gilles De La Tourette Syndrome; GTS</td>
	    			<td>Tg(DRD1-ctxA)7Burt</td>
	    			<td>&quot;Simple&quot; transgene in hemizygous genotype.</td>
	    		</tr>
	    		<tr>
	    			<td>137580</td>
	    			<td>Gilles De La Tourette Syndrome; GTS</td>
	    			<td>Tg(mI56i-cre,EGFP)1Kc</td>
	    			<td>This transgene is a recombinase and a reporter. The Conditional genotype includes the Met allele.</td>
	    		</tr>
	    		<tr>
	    			<td>137580</td>
	    			<td>Gilles De La Tourette Syndrome; GTS</td>
	    			<td>Met&lt;tm1Sst&gt;</td>
	    			<td>This allele is in a Conditional genotype with the recombinase/reporter transgene.</td>
	    		</tr>
	    		<tr>
	    			<td>101200</td>
	    			<td>Apert Syndrome</td>
	    			<td>
	    			<ul>
	    				<li>Fgfr2&lt;tm2Cxd&gt;</li>
	    				<li>Fgfr2&lt;tm1Ewj&gt;</li>
	    				<li>Fgfr2&lt;tm2Ewj&gt;</li> 
	    			</ul>
	    			</td>
	    			<td>All 3 of these alleles are in Conditional genotypes annotated to the term.  Fgfr2&lt;tm2Ewj&gt; is also in an OMG annotated to the term.</td>
	    		</tr>
	    		<tr>
	    			<td>101200</td>
	    			<td>Apert Syndrome</td>
	    			<td>Tg(EIIa-cre)C5379Lmgd</td>
	    			<td>This recombinase allele is in a Conditional genotype with 3 different Fgfr2 alleles.</td>
	    		</tr>
	    		<tr>
	    			<td>101200</td>
	    			<td>Apert Syndrome</td>
	    			<td>
	    			<ul>
	    				<li>Spry2&lt;tm1Ayos&gt;</li>
	    				<li>Spry4&lt;tm1Ayos&gt;</li>
	    			</ul>
	    			</td>
	    			<td>These alleles are in a COMPLEX genotype together that is annotated to this term. They are returned by the query but THIS DISEASE IS NOT DISPLAYED. (sto115)</td>
	    		</tr>
    		</table>
			</li>

			<li>Searching by MP or OMIM alternate ID includes alleles from appropriate genotypes. (sto35) 
			(updated for sto115)<p concordion:assertTrue="false">passed -jeffc 03-10-2014</p>
    		<table>
	    		<tr>
	    			<th>Alternate ID</th>
	    			<th>Term</th>
	    			<th>Returned Alleles</th>
	    			<th>Notes to testers</th>
	    		</tr>
	    		<tr>
	    			<td>Fyler:0540</td>
	    			<td>persistent truncus arteriosus type iv</td>
	    			<td>b2b1854Clo<br/>
	    				b2b1941Clo</td>
	    			<td>These are the only alleles annotated to this term.  Our ID is MP:0011664.</td>
	    		</tr>
	    		<tr>
	    			<td>Fyler:540</td>
	    			<td>persistent truncus arteriosus type iv</td>
	    			<td>b2b1854Clo<br/>
	    				b2b1941Clo</td>
	    			<td>Same as above, but different alternate ID. These are the only alleles annotated to this term. The term has no child terms.  Our ID is MP:0011664.</td>
	    		</tr>
	    		<tr>
	    			<td>601649</td>
	    			<td>Saethre-Chotzen Syndrome; SCS</td>
	    			<td>Twist1&lt;Pde&gt;<br/>
	    				Twist1&lt;Pas&gt;<br/>
	    				Twist1&lt;tm1Bhr&gt;<br/>
	    				Msx2&lt;tm1Rilm&gt;</td>
	    			<td>Alternate ID for Saethre-Chotzen Syndrome; SCS (101400). Four alleles annotated to this disease.  The Msx2 allele is part of a COMPLEX genotype; allele is returned but THIS DISEASE TERM IS NOT DISPLAYED (sto115).</td>
	    		</tr>
			</table>
			</li>

			<li>Searching by MP or OMIM terms  or synonyms returns alleles associated with genotypes annotated to the matching term and its child terms. (sto35)<br/>
				(Note: includes tests for plain text, text including non-alphanumeric characters, and text strings that include non-Boolean &quot;and&quot;, &quot;or&quot;, and &quot;not&quot; as plain words.)
			<p concordion:assertTrue="false">passed -jeffc 03-10-2014</p>
    		<table>
	    		<tr>
	    			<th>MP or OMIM Term or Synonym Text</th>
	    			<th>Included Alleles</th>
	    			<th>Notes to testers</th>
	    		</tr>
	    		<tr>
	    			<td>Ossificans</td>
	    			<td>
	    			<ul>
						<li>Acvr1&lt;tm1Emsh&gt;</li>
						<li>Tg(Eno2-Bmp4)3Jake</li>
						<li>Tg(CAG-LacZ,-ACVR1*,-EGFP)35-1Mis</li>
	    			</ul>
	    			</td>
	    			<td>The only matching term is Fibrodysplasia Ossificans Progressiva; FOP (135100)</td>
	    		</tr>
	    		<tr>
	    			<td>Tremor/Ataxia Syndrome</td>
	    			<td>
	    			<ul>
		    			<li>Fmr1&lt;tm2Cgr&gt;</li>
		    			<li>Tg(Pcp2-FMR1*)1Dln</li>
	    			</ul>
	    			</td>
	    			<td>Part of the name of Fragile X Tremor/Ataxia Syndrome; FXTAS (300623).<br/>
	    				These are the only 2 alleles annotated but without quotation marks, <b>1,845 alleles</b> are returned because of <i>Syndrome</i>.</td>
	    		</tr>
	    		<tr>
	    			<td>Ataxia/Tremor Syndrome</td>
	    			<td>
	    			<ul>
	    				<li>Fmr1&lt;tm2Cgr&gt;</li>
	    				<li>Tg(Pcp2-FMR1*)1Dln</li>
	    			</ul>
	    			</td>
	    			<td>Permutation of the name of Fragile X Tremor/Ataxia Syndrome; FXTAS (300623). Should return same alleles.<br/>
	    				These are the only 2 alleles annotated but without quotation marks, <b>same 1,845 alleles</b> as above are returned because of <i>Syndrome</i>.</td>
	    		</tr>
	    		<tr>
	    			<td>"Ataxia/Tremor Syndrome"</td>
	    			<td><i>none</i></td>
	    			<td>Permutation of the name of Fragile X Tremor/Ataxia Syndrome; FXTAS (300623) but includes quoted string. Should NOT return same alleles.</td>
	    		</tr>
	    		<tr>
	    			<td>"Bones and/or Mirror"</td>
	    			<td>
	    			<ul>
						<li>cl</li>
						<li>Dbf</li>
						<li>Fkbp8&lt;tm1Kei&gt;</li>
						<li>Fras1&lt;bl&gt;</li>
						<li>Grip1&lt;eb&gt;</li>
						<li>Pitx1&lt;tm1Rsd&gt;</li>
						<li>pma</li>
						<li>Ret&lt;tm1Kln&gt;</li>
						<li>skc3</li>
						<li>Tcirg1&lt;oc&gt;</li>
						<li>vsd</li>
	    			</ul>
	    			</td>
	    			<td>Words found in Clubfoot, Congenital, with or without Deficiency of Long Bones and/or Mirror-Image Polydactyly; CCF (119800).<br/>
	    				The Ret allele is in a genotype which includes the recombinase transgene.<br/>
	    				13 alleles annotated to this disease.</td>
	    		</tr>
	    		<tr>
	    			<td>Bones AND Mirror</td>
	    			<td>
	    			<ul>
						<li>cl</li>
						<li>Dbf</li>
						<li>Fkbp8&lt;tm1Kei&gt;</li>
						<li>Fras1&lt;bl&gt;</li>
						<li>Grip1&lt;eb&gt;</li>
						<li>Pitx1&lt;tm1Rsd&gt;</li>
						<li>pma</li>
						<li>Ret&lt;tm1Kln&gt;</li>
						<li>skc3</li>
						<li>Tcirg1&lt;oc&gt;</li>
						<li>vsd</li>
						<li>...and more...</li>
	    			</ul>
	    			</td>
	    			<td>Words in the name of Clubfoot, Congenital, with or without Deficiency of Long Bones and/or Mirror-Image Polydactyly; CCF (119800).<br/>
	    				This genotype includes the recombinase transgene.<br/>
	    				Should include same alleles as above, plus more from Notes combined with MP or OMIM annotations.</td>
	    		</tr>
	    		<tr>
	    			<td>Hypohidrotic/Hair/Tooth</td>
	    			<td>
	    			<ul>
	    				<li>Edar&lt;dl&gt;</li>
	    				<li>Edar&lt;Tg(OVE1B)Ove&gt;</li>
	    				<li>Eda&lt;Ta&gt;</li>
	    				<li>Eda&lt;Ta-6J&gt;</li>
	    			</ul>
	    			</td>
	    			<td>Part of the name of Ectodermal Dysplasia 10b, Hypohidrotic/Hair/Tooth Type, Autosomal Recessive; ECTD10B (224900),<br/>
	    				2 alleles annotated to this disease.<br/>
	    				Part of synonym for Ectodermal Dysplasia 1, Hypohidrotic, X-Linked; XHED (305100)<br/>
	    				2 alleles annotated to this disease.</td>
	    		</tr>
	    		<tr>
	    			<td>"Acrocephalosyndactyly, Type III; ACS3"</td>
	    			<td>
	    			<ul>
						<li>Twist1&lt;Pde&gt;</li>
						<li>Twist1&lt;Pas&gt;</li>
						<li>Twist1&lt;tm1Bhr&gt;</li>
						<li>EXCLUDE: Msx2&lt;tm1Rilm&gt;</li>
	    			</ul>
	    			</td>
	    			<td>Synonym for Saethre-Chotzen Syndrome; SCS (101400). 4 alleles annotated to this disease but Msx2&lt;tm1Rilm&gt; is only part of a complex genotype, so not returned.</td>
	    		</tr>
	    		<tr>
	    			<td>&quot;ethanol aversion&quot;</td>
	    			<td>
	    			<ul>
						<li>Cacna1b&lt;tm1Hssh&gt;</li>
						<li>Ccl2&lt;tm1Rol&gt;</li>
						<li>Ccl3&lt;tm1Unc&gt;</li>
						<li>Ccr2&lt;tm1Mae&gt;</li>
						<li>Drd2&lt;tm1Low&gt;</li>
						<li>Npy2r&lt;tm1Pern&gt;</li>
						<li>Tg(Camk2a-Fyn-531)1Nko</li>
	    			</ul>
					</td>
	    			<td>Synonym for &quot;alcohol aversion&quot; (MP:0002570).  Total should be <b>7 alleles</b>.</td>
	    		</tr>
	    		<tr>
	    			<td>"BALB/cByJ"</td>
	    			<td>These from MP Notes:
	    			<ul>
						<li>Acads&lt;del-J&gt;</li>
						<li>Axd</li>
						<li>Celf4&lt;Ff&gt;</li>
						<li>Col1a2&lt;tm1.1Mcbr&gt;</li>
						<li>Hld</li>
						<li>Lvs</li>
						<li>Mcoln3&lt;Va-J&gt;</li>
						<li>Phex&lt;Hyp-Duk&gt;</li>
						<li>Rpe65&lt;C57BL/6J&gt;</li>
						<li>Smad6&lt;tm1Glvn&gt;</li>
						<li>Tsed</li>
						<li>Tyr&lt;c-2J&gt;</li>
	    			</ul>
	    			<p>And these from TEXT-QTL Notes:</p>
	    			<ul>
	    				<li>Abmm&lt;DBA/2J&gt;</li>
	    				<li>Ccrs4&lt;BTBR&gt;</li>
	    			</ul>
					</td>
	    			<td>This text is not found in any MP or OMIM term, only in MP annotation Notes (General and Background Sensitivity, returning <b>12 alleles</b>) and TEXT-QTL notes, returning a total of <b>743 alleles</b>.</td>
	    		</tr>
	    		<tr>
	    			<td>&quot;however germ cell apoptosis"</td>
	    			<td>Fts</td>
	    			<td>This phrase is found only in Normal notes and only for this allele.</td>
	    		</tr>
	    		<tr>
	    			<td>"decreased IgG2a level"</td>
	    			<td>
	    			<ul>
	    				<li>Aicda&lt;tm1Hon&gt;</li>
	    				<li>Bach2&lt;tm1Igar&gt;</li>
	    				<li>Card11&lt;m1Btlr&gt;</li>
	    			</ul>
	    			</td>
	    			<td>Should return same set of <b>107 alleles</b> as MP:0008496.  Should NOT tokenize before and after the case-changes or the "2" in "IgG2a".</td>
	    		</tr>
	    		<tr>
	    			<td>IgG2b</td>
	    			<td>Around 200 alleles</td>
	    			<td>Should return DIFFERENT set of alleles from query for <i>Ig G 2 b</i> (44,045 alleles) and <i>IgG 2 b</i> (43,944 alleles).  Our query should NOT tokenize before and after the case-changes or the "2".</td>
	    		</tr>
    		</table>
			</li>
			<li>Searching by phenotype can match text in an allele's "General" note. (sto114)
			<p concordion:assertTrue="false">passed - jeffc 03-10-2014</p>
				<ul>
					<li>Do a phenotype/disease query for <a href="http://cardolan.informatics.jax.org/allele/summary?phenotype=S74469"><b>S74469</b></a></li>
					<li>Verify that <b>A&lt;iy&gt;</b> is returned. (<i>Has general note "Genbank ID for this mutation: S74469"</i>)</li>
				</ul>
			</li>

			<li>Searching by combinations of MP or OMIM terms, synonyms or IDs using Booleans returns alleles associated with genotypes annotated to the matching term and its child terms. (sto35)
			(updated for sto115 and sto181)<p concordion:assertTrue="false">passed -klf 03-13-2014</p>
    		<table>
	    		<tr>
	    			<th>MP or OMIM Term<br/>
	    				(Note quotation marks, parentheses, etc.)</th>
	    			<th>Included Alleles</th>
	    			<th>Notes to testers</th>
	    		</tr>
	    		<tr>
	    			<td>Netherton Chanarin</td>
	    			<td>
	    			<ul>
	    				<li>Dsg4&lt;lah&gt;</li>
	    				<li>Spink5&lt;tm1Drh&gt;</li>
	    				<li>Spink5&lt;tm1Hov&gt;</li>
	    				<li>F2rl1&lt;tm1Cgh&gt;</li>
	    			</ul>
	    			</td>
	    			<td>These words belong to 2 separate OMIM terms. Query should OR them together returning <b>8 alleles</b>. F2rl1 allele is included but DISEASE IS NOT DISPLAYED due to COMPLEX genotype with Spink5&lt;tm1Hov&gt;. (sto115)</td>
	    		</tr>
	    		<tr>
	    			<td>Netherton OR Chanarin</td>
	    			<td>
	    			<ul>
	    				<li>Dsg4&lt;lah&gt;</li>
	    				<li>F2rl1&lt;tm1Cgh&gt;</li>
	    				<li>Spink5&lt;tm1Drh&gt;</li>
	    			</ul>
	    			</td>
	    			<td>Should get same result as above: <b>8 alleles</b>.</td>
	    		</tr>
	    		<tr>
	    			<td>"split foot"</td>
	    			<td></td>
	    			<td>There should be no allele results.There are 2 matching OMIM terms but they have no annotations.</td>
	    		</tr>
	    		<tr>
	    			<td>"enhanced behavioral response to cocaine" AND "enhanced behavioral response to alcohol"</td>
	    			<td>Only these:
	    			<ul>
	    				<li>Drd4&lt;tm1Dkg&gt;</li>
	    				<li>hipr</li>
	    				<li>Ppp1r1b&lt;tm1Aaf&gt;</li>
	    			</ul>
	    			</td>
	    			<td>ANDing two MP terms. 30 alleles are annotated to &quot;...cocaine&quot; (MP:0009754).  38 alleles are annotated to &quot;...alcohol&quot; (MP:0009751) or its child term &quot;hyperactivity...&quot; (MP:0002629). Only these <b>3 alleles</b> are annotated to both terms.</td>
	    		</tr>
	    		<tr>
	    			<td>MP:0009754 AND MP:0009751</td>
	    			<td>Only these:
	    			<ul>
	    				<li>Drd4&lt;tm1Dkg&gt;</li>
	    				<li>hipr</li>
	    				<li>Ppp1r1b&lt;tm1Aaf&gt;</li>
	    			</ul>
	    			</td>
	    			<td>ANDing the IDs of the two terms above. Should return same set of <b>3 alleles</b>. (These alleles are annotated to the child term of &quot;...alcohol&quot;.)</td>
	    		</tr>
	    		<tr>
	    			<td>MP:0009754 OR MP:0009751</td>
	    			<td>
	    			<ul>
	    				<li>Actre1&lt;BALB/cJ&gt;</li>
	    				<li>Aldh2&lt;Gt(OST7285)Lex&gt;</li>
	    				<li>Camk4&lt;tm1Gsc&gt;</li>
	    				<li>Drd4&lt;tm1Dkg&gt;</li>
	    				<li>hipr</li>
	    				<li>Tg(Drd1a-cre)1Gsc</li>
	    				<li>Tg(Drd2-EGFP)S118Gsat</li>
	    				<li>Unc79&lt;Lwt&gt;</li>
	    			</ul>
	    			</td>
	    			<td>ORing the IDs of the two terms above.  Should return all alleles annotated to just one  of the terms in addition to the 3 annotated to both terms.  Total should be <b>67 alleles</b>.</td>
	    		</tr>
	    		<tr>
	    			<td>MP:0009754 AND NOT MP:0009751</td>
	    			<td>
	    			<ul>
	    				<li>Camk4&lt;tm1Gsc&gt;</li>
	    				<li>Chrm4&lt;tm2.1Jwe&gt;</li>
	    				<li>Clock&lt;m1Jt&gt;</li>
	    				<li><i>Exclude: </i> Drd4&lt;tm1Dkg&gt;</li>
	    				<li><i>Exclude:</i> hipr</li>
	    				<li><i>Exclude:</i> Ppp1r1b&lt;tm1Aaf&gt;</li>
	    			</ul>
	    			</td>
	    			<td>Combining the IDs of the two terms above with NOT.  Should exclude the set of 3 alleles for the AND case, returning alleles annotated only to the first ID ("...cocaine").  Total should be <b>29 alleles</b>.</td>
	    		</tr>
	    		<tr>
	    			<td>"response to cocaine"</td>
	    			<td>
	    			<ul>
	    				<li>Camk4&lt;tm1Gsc&gt;</li>
	    				<li>Chrm4&lt;tm2.1Jwe&gt;</li>
	    				<li>Clock&lt;m1Jt&gt;</li>
						<li>EXCLUDE: Creb1&lt;tm3Gsc&gt;</li>
	    				<li>Datd&lt;C57BL/6J&gt;</li>
	    				<li>Datd&lt;DBA/2J&gt;</li>
	    				<li>Dlg4&lt;tm2.1Grnt&gt;</li>
	    				<li>En1&lt;tm2(cre)Wrst&gt;</li>
						<li>Grm5&lt;tm1Fcq&gt;</li>
						<li>Gt(ROSA)26Sor&lt;tm1(EYFP)Cos&gt;</li>
						<li>Per2&lt;tm1Brd&gt;</li>
						<li>Slc6a3&lt;tm1(cre)Xz&gt;</li>
	    				<li>Slc6a3&lt;tm1.1(cre)Bkmn&gt;</li>
						<li>Tg(Drd1a-cre)EY262Gsat</li>
						<li>Tg(Drd1a-EGFP)X60Gsat</li>
						<li>EXCLUDE: Tg(Camk2a-cre)1Gsc</li>
	    			</ul>
					</td>
	    			<td><p>Should return alleles annotated to the term shown above, alleles annotated to "impaired behavioral response to cocaine" (MP:0009758), and alleles associated with annotations having notes containing this text from any of 3 Notes Types (General, Normal, and Background Sensitivity). Total should be <b>62 alleles</b>.</p>
	    				<p>Creb1 and Tg(Camk2a-cre)1Gsc alleles are excluded because the matching text appears in a note that is associated with a Normal annotation, which we exclude.</p> 
	    				<p>Two Cyfip alleles present in a trans-het genotype added to production 1/9/2014 so we are returning them where Public is not. Data will appear on public after 1/20/2014?</p>
	    				<p>Two Datd QTL alleles are included because matching text appears in the &quot;TEXT-QTL&quot; note.</p>
	    			</td>
	    		</tr>
	    		<tr>
	    			<td>"response to morphine"</td>
	    			<td>Oprd1&lt;tm1Jep&gt;<br/>...and more</td>
	    			<td>Annotated to the term in COMPLEX genotype only. Allele should be returned by the query.  Header is displayed because additional terms appear for other genotypes. (sto115)</td>
	    		</tr>
	    		<tr>
	    			<td>"enhanced behavioral response to cocaine" AND NOT (Parkinson OR Autism)</td>
	    			<td>
	    			<ul>
	    				<li>EXCLUDE: Dlg4&lt;tm2.1Grnt&gt;</li>
	    				<li>EXCLUDE: En1&lt;tm2(cre)Wrst&gt;</li>
	    				<li>EXCLUDE: Slc6a3&lt;tm1.1(cre)Bkmn&gt;</li>
	    			</ul>
	    			</td>
	    			<td>Should return only <b>29 alleles</b>. </td>
	    		</tr>
	    		<tr>
	    			<td>"response to cocaine" AND NOT behavior</td>
	    			<td><i>none</i></td>
	    			<td>There aren't any alleles that have the  exact phrase "response to cocaine" but do not also have the word "behavior" in a Note or Annotation somewhere. </td>
	    		</tr>
	    		<tr>
	    			<td>(cocaine AND response AND to) AND NOT behavior</td>
	    			<td>
	    			<ul>
	    				<li>Nr1i3&lt;tm1Ddm&gt;</li>
	    				<li>Datd&lt;C57BL/6J&gt;</li>
	    				<li>Datd&lt;DBA/2J&gt;</li>
	    				<li>EXCLUDE: Chrna3&lt;tm1.1Hwrt&gt;</li>
	    				<li>EXCLUDE: Igh&lt;tm1.1Erpi&gt;</li>
	    			</ul>
	    			</td>
	    			<td>Alleles may have the words "cocaine", "response", and "to" in separate Notes or annotated Terms but must not also have the word "behavior". <br/>
	    				Searching TEXT-QTL Mapping Experiment Notes occurs only for allele with Allele Type = QTL. Other alleles with matching text in TEXT-QTL Notes must not be returned. (sto181)<br/>
	    				(Examples: Chrna3&lt;tm1.1Hwrt&gt; has "response to" in one TEXT-QTL experiment note (J:39049) associated with marker Chrna3 and "cocaine" in another TEXT-QTL experiment note (J:30687) but this allele is NOT a QTL allele so it should not be returned. <br/>
	    				Igh&lt;tm1.1Erpi&gt; has "response to" in a note for an MP annotation and "cocaine" in a TEXT-QTL experiment note associated with marker Igh (J:23071) but this allele is NOT a QTL allele so it should not be returned.)
	    			</td>
	    		</tr>
    		</table>
			</li>
			
			<li>Searching by MP or OMIM terms  or synonyms or MP Notes text disregards "words" on MGI's modified Stop Words list. (sto35)<br/>
				(Note: This should be the same Stop Words list we are using for the Quick Search.  See <a href="http://prodwww.informatics.jax.org/wiki/index.php/TR:7754_Requirements">Search Tool Requirements</a> wiki page.<br/>
				This list is:  <i>a, an, and, are, as, be, but, if, is, it, such, that, the, their, then, there, these, they, this, was, will</i>.<br/>
				Excluded from the list, therefore, treated as words are: <i>at, by, for, in, into, no, not, of, on, or, to, with</i>.)
			<p concordion:assertTrue="false">passed -jeffc 03-11-2014</p>
			<table>
			<tr>
				<th>Input</th>
				<th>Included Alleles Returned</th>
				<th>Notes to Testers</th>
			</tr>
			<tr>
				<td>a, an, and, are, as, be, but, if, is, it, or</td>
				<td>lots</td>
				<td>These are Stop Words so they should be ignored. The query should return all annotated alleles.</td>
			</tr>
			<tr>
				<td>such, that, the, their, then, there, these, they, this, was, will</td>
				<td>lots</td>
				<td>These are Stop Words so they should be ignored. The query should return all annotated alleles.</td>
			</tr>
			<tr>
				<td>at, by, for, in, into, no</td>
				<td>different list</td>
				<td>These are not Stop Words so they should be respected.  The query should return only alleles having one or more of these words in an annotated term.</td>
			</tr>
			<tr>
				<td>not, of, on,  to, with</td>
				<td>different list</td>
				<td>These are not Stop Words so they should be respected.  The query should return only alleles having one or more of these words in an annotated term.</td>
			</tr>
			<tr>
				<td>in AND by AND of AND for AND grows AND &quot;sparse coat&quot;</td>
				<td>
				<ul>
					<li>Ass1&lt;bar&gt;</li>
					<li>Ass1&lt;fold&gt;</li>
					<li>Gsdma3&lt;Dfl&gt;</li>
				</ul>
				</td>
				<td><p>Only these 3 alleles contain all the terms.  &quot;In&quot;, &quot;By&quot;,  &quot;Of&quot;, and &quot;For&quot; are not Stop Words so they should be respected.</p>
					<p>The note text which returns the Ass1 alleles is: <i>enough hair grows in by 3 weeks of age for a sparse coat</i>. The portion of note text which returns the Gsdma3 allele is: <i>...a second sparse coat re-grows but is lost by 4 months of age...</i>.</p>
					<p>Searching with just <i>grows AND "sparse coat"</i> returns 3 additional alleles.</p>
				</td>
			</tr>
			<tr>
				<td>differentiation AND is AND perturbed AND  such AND that AND they AND differentiate AND into AND hybrid</td>
				<td>
				<ul>
					<li>Onecut1&lt;tm1Flem&gt;</li>
					<li>Pten&lt;tm1Hwu&gt;</li>
					<li>Tfap2a&lt;tm2Will&gt;</li>
				</ul>
				</td>
				<td>The text <i>differentiation is perturbed such that they differentiate into hybrid</i> is found in an MP Note.  "Is", "Such", "That", and "They" are Stop Words and should produce the same results if they are not included in the query.  Returns only these <b>3 alleles.</b></td>
			</tr>
			</table>
			</li>			

			<li>The "Anatomical Systems Affected by Phenotypes" pop-up tool populates the Pheno/Disease input field with the MP IDs of the selected systems.
			<p concordion:assertTrue="false">passed. -jeffc 03-11-2014</p>
			<table>
			<tr>
				<th>Selected Text</th>
				<th>Expected MP ID</th>
			</tr>
			<tr>
				<td>adipose tissue</td>
				<td>MP:0005375</td>
			</tr>
			<tr>
				<td>behavioral/neurological</td>
				<td>MP:0005386</td>
			</tr>
			<tr>
				<td>cardiovascular system</td>
				<td>MP:0005385</td>
			</tr>
			<tr>
				<td>no phenotypic analysis</td>
				<td>MP:0003012</td>
			</tr>
			<tr>
				<td>normal phenotype</td>
				<td>MP:0002873</td>
			</tr>
			<tr>
				<td>vision/eye </td>
				<td>MP:0005391</td>
			</tr>
			<tr>
				<td>tumorigenesis </td>
				<td>MP:0002006</td>
			</tr>
			<tr>
				<td>taste/olfaction </td>
				<td>MP:0005394</td>
			</tr>
			</table>
			</li>

			<li>The "Anatomical Systems Affected by Phenotypes" pop-up tool adds the selected MP IDs to whatever text is already in the Pheno/Disease input field
			<p concordion:assertTrue="false">passed. -jeffc 03-11-2014</p>
			<ol>
				<li>Enter some text in the Phenotypes/Diseases form input text area.</li>
				<li>Open the "Anatomical Systems Affected by Phenotypes" pop-up tool.</li>
				<li>Select a checkbox and click "Add to query".</li>
				<li>Verify that the MP IDs are added to the existing text in the form text area.</li>
			</ol>
			</li>

    		<li>When allele is annotated to term "Normal phenotype"   or its children in OMG or Conditional genotype, and no other terms in ANY genotypes, display "no abnormal phenotype observed". (sto115)
			<p concordion:assertTrue="false">passed. -jeffc 03-11-2014</p>  
			<ol>
				<li>From the Allele Query Form, submit a Nomenclature query for the items listed below. Verify that for the listed allele, the "Abnormal Phenotypes Reported in these Systems" column displays the listed text.
				<table>
					<tr>
						<th>Nomenclature field</th>
						<th>Allele on summary</th>
						<th>"Systems" column text:<br/>
						<b>no abnormal phenotype observed</b></th>
						<th>notes to tester</th>
					</tr>
					<tr>
						<td>Alkbh8</td>
						<td>Alkbh8&lt;tm1Klng&gt; (MGI:4822190)</td>
						<td>yes</td>
						<td></td>
					</tr>
					<tr>
						<td>Aph1b</td>
						<td>Aph1b&lt;tm1Bdes&gt; (MGI:3522628)</td>
						<td>yes</td>
						<td>in OMG and in complex geno with MGI:3530442 Aph1c&lt;tm1.1Bdes&gt; - only annot for either allele</td>
					</tr>
					<tr>
						<td>Aph1c</td>
						<td>Aph1c&lt;tm1.1Bdes&gt; (MGI:3530442)</td>
						<td>yes</td>
						<td></td>
					</tr>
					<tr>
						<td>Hprt</td>
						<td>Hprt&lt;tm1.1Pobe&gt; (MGI:3582526)</td>
						<td>yes</td>
						<td></td>
					</tr>
					<tr>
						<td>Hprt</td>
						<td>Hprt&lt;tm1Dwm&gt; MGI:4462242</td>
						<td>yes</td>
						<td>annotated to &quot;phenotypic reversion&quot;, child of &quot;normal phenotype&quot;, in two OMGs, and no other terms</td>
					</tr>
					<tr>
						<td>Vps52</td>
						<td>Vps52&lt;t-w5+Art&gt; MGI:5469354</td>
						<td>no</td>
						<td>annotated to &quot;phenotypic reversion&quot;, child of &quot;normal phenotype&quot; in COMPLEX genotypes, and no other terms</td>
					</tr>
					<tr>
						<td>Hprt</td>
						<td>Hprt&lt;b-m3&gt; (MGI:1857299)</td>
						<td>no</td>
						<td>annotated to &quot;phenotypic reversion&quot;, child of &quot;normal phenotype&quot; in one genotype, but has abnormal annotations in other genotypes</td>
					</tr>
					<tr>
						<td>Hprt</td>
						<td>Hprt&lt;tm1(Agt)Sig&gt; (MGI:2449451)</td>
						<td>no</td>
						<td>only annotation is a "normal" annotation so column should be blank (those annotations get ignored)</td>
					</tr>
					<tr>
						<td>Abr</td>
						<td>Abr&lt;tm1Jhg&gt; (MGI:2388713)</td>
						<td>no</td>
						<td>has "no abnormal phenotype detected" annot in OMG, but has MP annotations in COMPLEX genotype with Bcr&lt;tm1Hkp&gt;, so should not display this way</td>
					</tr>
					<tr>
						<td>Etn2</td>
						<td>Etn2&lt;Sd&gt; (MGI:1857746)</td>
						<td>no</td>
						<td>has "no abnormal phenotype detected" annot but many others too, so should not display this way</td>
					</tr>
				</table>
				</li>
			</ol>
    		</li>
    		
    		<li>When allele is annotated to term "No phenotypic analysis"  in OMG or Conditional genotype, and no other terms in ANY genotypes, display "No Phenotypic Analysis". (sto115)
 			<p concordion:assertTrue="false">passed. -jeffc 03-11-2014</p>  
			<ol>
				<li>From the Allele Query Form, submit a Nomenclature query for the items listed below. Verify that for the listed allele, the "Abnormal Phenotypes Reported in these Systems" column displays the listed text.
				<table>
					<tr>
						<th>Nomenclature field</th>
						<th>Allele on summary</th>
						<th>"Systems" column text:<br/>
						<b>no phenotypic analysis</b></th>
						<th>notes to tester</th>
					</tr>
					<tr>
						<td>Ano6</td>
						<td>Ano6&lt;tm1.1Naga&gt; (MGI:5538269)</td>
						<td>yes</td>
						<td></td>
					</tr>
					<tr>
						<td>Ezh2</td>
						<td>Ezh2&lt;tm1Sho&gt; (MGI:3790775)</td>
						<td>yes</td>
						<td>has 2 geno but this is only annot</td>
					</tr>
					<tr>
						<td>Hprt</td>
						<td>Hprt&lt;tm11Aird&gt; (MGI:5514145)</td>
						<td>yes</td>
						<td>only annotation and is in an OMG</td>
					</tr>
					<tr>
						<td>Acvr1b</td>
						<td>Acvr1b&lt;tm1.2Pbe&gt; (MGI:5441234)</td>
						<td>yes</td>
						<td>in a CONDITIONAL genotype; this is the only annotation</td>
					</tr>
					<tr>
						<td>Col6a1</td>
						<td>Col6a1&lt;tm1Yah&gt; (MGI:4868459)</td>
						<td>no</td>
						<td>in a COMPLEX Non-Conditional genotype; this is the only annotation - we do not display COMPLEX annotations; cell should be blank</td>
					</tr>
					<tr>
						<td>Prmt2</td>
						<td>Prmt2&lt;tm1Yah&gt; (MGI:4868409)</td>
						<td>no</td>
						<td>in two COMPLEX Non-Conditional genotypes; this is the only annotation - we do not display COMPLEX annotations; cell should be blank</td>
					</tr>
					<tr>
						<td>Cstb</td>
						<td>Cstb&lt;tm1Yah&gt; (MGI:4868410)</td>
						<td>no</td>
						<td>annotated to this term in a COMPLEX Non-Conditional genotypes; but all other annotations are "N" so should be reported as &quot;no abnormal phenotype observed&quot;</td>
					</tr>
					<tr>
						<td>Etn2</td>
						<td>Etn2&lt;Sd&gt; (MGI:1857746)</td>
						<td>no</td>
						<td>has "no phenotypic analysis" annot but many others too, so should not display this way</td>
					</tr>
					<tr>
						<td>Emx2</td>
						<td>Emx2&lt;tm2(cre)Sia&gt; (MGI:3579416)</td>
						<td>no</td>
						<td>has "no phenotypic analysis" in a CONDITIONAL genotype but many others too in another CONDITIONAL genotype, so should not display this way</td>
					</tr>
				</table>
				</li>
			</ol>
    		</li>
    		<li>Verify that disease column excludes "Not" annotations when there are no other annotations for a disease term for a given allele (sto32) (updated for sto115)
 			<p concordion:assertTrue="false">passed. -jeffc 03-11-2014</p>
    		<ol>
                <li>Verify that the row for allele <b>Foxp3<sup>sf</sup></b> displays <b>I<b>chthyosis and Male Hypogonadism 308200, </b>Immunodysregulation, Polyendocrinopathy, and Enteropathy, X-Linked; 304790, Wiskott-Aldrich Syndrome; WAS 301000</b> under Similar Human Diseases.</li>
                <li>Verify that the row for allele <b>Foxp3<sup>sf</sup></b> DOES NOT display <b>Ichthyosis, X-Linked; XLI OMIM ID: 308100</b>  under Similar Human Diseases.</li>
                <li>Verify that the row for allele <b>Nos1<sup>tm1Plh</sup></b> displays <b>Achalasia, Familial Esophageal 200400, Pyloric Stenosis, Infantile Hypertrophic, 1; IHPS1 179010</b> (<i>Note: there is an annotation and a "Not" annotation for Pyloric Stenosis, so this disease SHOULD appear</i>).</li>
            </ol>
    		</li>
    		<li>Verify you can query by a single collection and multiple collections(sto18)
    		<p concordion:assertTrue="false">jeffc 01/21/2014</p>
    		<ol>
    			<li>Open the Allele query form</li>
    			<li>Select the collection <b>Australian PhenomeBank</b> and click the search</li>
    			<li>Verify you return <b>51</b> results (EI has 54 but 3 of them are wild-type)</li>
    			<li>Select the collection <b>Beutler Mutagenesis</b>, leaving <b>Australian PhenomeBank</b> still selected and click the search</li>
    			<li>Verify the results returns <b>242</b> results</li>
    		</ol>
    		</li>
   		<li><span class="regressHi">Regress</span> Collection appears in Allele Information ribbon on Recombinase Specificity Pages, between Molecular Description and Find Mice.
 			<p concordion:assertTrue="false">passed. -klf 2014-01-02</p> (sto17)
    		<ol>
                <li>Navigate to a Recombinase Specificity page (there may be more than one) for <b>Dbx1&lt;tm1.1(cre)Mull&gt;</b> (MGI:5014205)</li>
                <li>Verify that the Collection value <b>Neuroscience Blueprint cre</b>  appears in the correct place.</li>
                <li>Navigate to a Recombinase Specificity page (there may be more than one) for <b>Tg(ACTFLPe)9205Dym</b> (MGI:2448985)</li>
                <li>Verify that the Collection value <b>Neuroscience Blueprint cre</b> appears in the correct place.</li>
    		</ol>
    		</li>

    		<li><span class="regressHi">Regress</span> Collection does not appear when the allele is not assigned to a collection.
 			<p concordion:assertTrue="false">passed. -klf 2014-01-02</p> (sto17)
    		<ol>
                <li>Navigate to a Recombinase Specificity page (there may be more than one) for <b>Hprt&lt;tm1(cre)Mnn&gt;</b> (MGI:2181632)</li>
                <li>Verify that <b>no Collection value</b> appears.</li>
                <li>Navigate to a Recombinase Specificity page (there may be more than one) for <b>Tg(CMV-cre/ERT)1Ipc</b> (MGI:2387570)</li>
                <li>Verify that <b>no Collection value</b> appears.</li>
    		</ol>
    		</li>
    		<li><span class="regressHi">Regress</span> Allele Subtypes appear in Allele Information ribbon after Type as comma delimited list.
 			<p concordion:assertTrue="false">passed - kstone 2014/01/23</p> (sto24)
    		<ol>
                <li>Navigate to a Recombinase Specificity page (there may be more than one) for 
                	<a href="http://cardolan.informatics.jax.org//recombinase/specificity?id=MGI:4420223&amp;systemKey=4856343">Gad2&lt;tm2(cre)Zjh&gt;</a></li>
                <li>Verify that the subtypes value <b>(Inserted expressed sequence, Recombinase)</b>  appears in the correct place.</li>
    		</ol>
    		</li>

    		<li><span class="regressHi">Regress</span> Image link URLs appear on the Recombinase Activity Detail page
 			<p concordion:assertTrue="false">klf 6/6/13</p>
 			<p>Note: this test requires fake data.</p>
    		<ol>
                <li>Navigate to the Recombinase Activity Detail page for <b>Calb2&lt;tm1(cre)Zjh&gt; - Nervous System</b>. </li>
                <li>Verify that images in the "Images" ribbon display linked text between the J:# and the Figure Label text.</li>
                <li>Verify that tiny thumbnails in the "Recombinase Activity" ribbon, "Reference, Source" column, display linked text following the Figure Label text.</li>
    		</ol>
    		</li>

    		<li><span class="regressHi">Regress</span> Image link URLs open the linked resource in a new window
 			<p concordion:assertTrue="false">klf 6/6/13</p>
    		<ol>
                <li>Navigate to the Recombinase Activity Detail page for <b>Calb2&lt;tm1(cre)Zjh&gt; - Nervous System</b>. </li>
                <li>Verify "GRS Image" is hyperlinked to ndp.jax.org in the Images ribbon for Fig. Calb2 Brain (dorsal) Adult, and opens in a new window.</li>
                <li>Verify "GRS Image" is hyperlinked to ndp.jax.org in the "Reference, Source" column wherever Fig. Calb2 Brain (dorsal) Adult is cited, and opens in a new window.</li>
                <li>***Verify "AIBS Data" is hyperlinked to connectivity.brain-map.org in the Images ribbon for Fig. Calb2 Spinal cord (wh matter) P7, and opens in a new window.</li>
                <li>Verify "AIBS Data" is hyperlinked to connectivity.brain-map.org in the "Reference, Source" column wherever Fig. Calb2 Spinal cord (wh matter) P7 is cited, and opens in a new window.</li>
                <li>Verify each link opens in a new tab/window.</li>
                <li>*** Only the one marked *** needs regression testing.</li>
    		</ol>
    		</li>

    		<li><span class="regressHi">Regress</span> The \Link(...) markup does not affect the display of images on GXD Expression Detail WI pages.
 			<p concordion:assertTrue="false">klf 4/30/2013</p>
    		<ol>
                <li>Open the record for image MGI:5310731 (in situ reporter (knock-in) assay used by both Cre and GXD for Six1&lt;tm1(cre,ALPP)Xli&gt;)</li>
                <li>Enter \Link(https://www.google.com/|Test Link|)  </li>
                <li>Save the note and the image record.</li>
                <li>After Exporter and/or Mover steps...</li>
                <li>Navigate to the Recombinase Activity Detail page for Six1&lt;tm1(cre,ALPP)Xli&gt; - Alimentary System.</li>
                <li>Verify "Test Link" is hyperlinked to Google.com in the Images ribbon for Fig. 5.</li>
                <li>Verify "Test Link" is hyperlinked to Google.com in the "Reference, Source" column wherever Figure 5 is cited.</li>
                <li>Navigate to the Recombinase Activity Detail page for Six1&lt;tm1(cre,ALPP)Xli&gt; - Sensory Organs.</li>
                <li>Verify "Test Link" is hyperlinked to Google.com in the Images ribbon for Fig. 5.</li>
                <li>Verify "Test Link" is hyperlinked to Google.com in the "Reference, Source" column wherever Figure 5 is cited.</li>
                <li>Navigate to the GXD Image Detail page for the image (http://cardolan.informatics.jax.org/image/MGI:5310731).</li>
                <li>Verify "Test Link" does not appear on the page.</li>
                <li>Verify the page looks the same as it does on Public.</li>
    		</ol>
    		</li>
    		<li>Autocomplete exists and displays correct information
    		<p concordion:assertTrue="false">jeffc 04/14/2014 verified again for sto184</p>
    		<ol>
                <li>Open <a href="http://cardolan.informatics.jax.org/diseasePortal">disease portal query form</a> </li>
                <li>Type <b>systemic lupus</b> into phenotypes and diseases box</li>
                <li>Verify that both <b>OMIM</b> and <b>Mammalian Phenotype</b> terms appear. 
                	<i>The first two terms should be "Systemic Lupus Erythmatosus; SLE", an OMIM term, 
                	and "decreased resistance to systemic lupus erythematosus", an MP term.</i>
                </li>
                <li>Verify that the first term <b>Systemic Lupus Erythematosus; SLE</b> is displayed with the text 
                	<i>OMIM</i> immediately following it.
                </li>
                <li>Verify that the second term <b>decreased resistance to systemic lupus erythematosus</b>
                	is displayed with the text <i>[increased susceptibility to systemic lupus erythematosus] Mammalian Phenotype,
                	the brackets indicating that you matched a synonym.</i>
                </li>
                <li>Verify that the "disease clumping" term <b>Systemic Lupus Erythematosus</b> (with no abbreviation) does not appear in the autocomplete. (sto117)</li>
            </ol>
            </li>
            <li>Autocomplete populates box with IDs
            <p concordion:assertTrue="false">passed - jeffc 04/14/2014 verified for sto184</p>
    		<ol>
                <li>Type <b>systemic lupus</b> into phenotypes and diseases box</li>
                <li>Click (or highlight and hit enter/tab) the first term <b>Systemic Lupus Erythematosus; SLE</b>.</li>
                <li>Verify that the text <b>152700,</b> is put into the phenotype/disease query box</li>
                <li>Clear form. Type <b>wet lung</b> into phenotypes and diseases box</li>
                <li>Click (or highlight and hit enter/tab) the first term <b>wet lung [pulmonary edema]</b>.</li>
                <li>Verify that the text <b>MP:0003828,</b> is put into the phenotype/disease query box</li>
            </ol>
            </li>
            <li>Autocomplete supports multiples
            <p concordion:assertTrue="false">passed - jeffc 04/14/2014 verified for sto184</p>
    		<ol>
                <li>Type <b>systemic lupus</b> into phenotypes and diseases box</li>
                <li>Click (or highlight and hit enter/tab) the first term <b>Systemic Lupus Erythematosus; SLE</b>.</li>
                <li>Verify that the text <b>152700,</b> is put into the phenotype/disease query box</li>
                <li>Do not clear the form. Continue typing <b>wet lung</b> into phenotypes and diseases box</li>
                <li>Click (or highlight and hit enter/tab) the first term <b>wet lung [pulmonary edema]</b>.</li>
                <li>Verify that the text <b>MP:0003828,</b> is added to the phenotype/disease query box</li>
            </ol>
            </li>
            <li>"You searched for" section displays the searched IDs as well as the terms they map to
           <p concordion:assertTrue="false">passed - jeffc 04/14/2014 verified for sto184</p>
    		<ol>
                <li>Type <b>systemic lupus</b> into phenotypes and diseases box</li>
                <li>Click (or highlight and hit enter/tab) the first term <b>Systemic Lupus Erythematosus; SLE</b>.</li>
                <li>Verify that the text <b>152700,</b> is put into the phenotype/disease query box</li>
                <li>Do not clear the form. Continue typing <b>wet lung</b> into phenotypes and diseases box</li>
                <li>Click (or highlight and hit enter/tab) the first term <b>wet lung [pulmonary edema]</b>.</li>
                <li>Verify that the text <b>MP:0003828,</b> is added to the phenotype/disease query box</li>
                <li>Click "Search"</li>
                <li>Verify that the "you searched for" section displays the following text after 
                	"Phenotypes or Diseases matching": 
                	<b>[152700(Systemic Lupus Erythematosus; SLE), MP:0003828(pulmonary edema),]</b>
                </li>
            </ol> 
            </li>   		
    		<li>Confirm Hovering over a Gene symbol on the Genes summary, displays text of Gene's full name and it's feature type(sto132)
    		<p concordion:assertTrue="false">passed jeffc 04/24/2012</p>
    		<ol>
                <li>Query for marker <b>pax*</b></li>
                <li>Hover your mouse over Gene Symbol <b>Pax1</b></li>
                <li>Verify that a popup displays saying <b>Name: paired box 1 Type: protein coding gene</b></li>
                <li>Hover your mouse over Gene Symbol <b>PAX9</b></li>
                <li>Verify that a popup displays saying <b>Name: paired box 9 Type: protein coding gene</b></li>
                <li>Query for location <b>Chr12:3000000-10000000</b></li>
                <li>Hover your mouse over Gene Symbol <b>Apob</b></li>
                <li>Verify that a popup displays saying <b>Name: apolipoprotein B Type: protein coding gene</b></li>
                <li>Hover your mouse over Gene Symbol <b>OSR1</b></li>
                <li>Verify that a popup displays saying <b>Name: odd-skipped related transcription factor 1</b></li>
                <li>Query for disease <b>diabetes</b></li>
                <li>Hover your mouse over Gene Symbol <b>Idd3</b></li>
                <li>Verify that a popup displays saying <b>Name: insulin dependent diabetes susceptibility 3 Type: QTL</b></li>
                <li>Hover your mouse over Gene Symbol <b>IL6</b></li>
                <li>Verify that a popup displays saying <b>Name: interleukin 6 (interferon, beta 2)</b></li>
    		</ol>
    		</li>
			<li>Confirm on the HDMC Disease tab that numbers in the Mouse Models column are links(sto128)
    		<p concordion:assertTrue="false">jeffc 03/17/2014</p>
    		<ol>
                <li>Query for marker <b>Actb</b> on the Human Disease and Mouse Model query form</li>
                <li>Click on the Diseases tab and locate the Mouse Models column</li>
                <li>Verify that all rows that contain a number are links</li>
    		</ol>
    		</li>
		    <li>Confirm Hovering over a Gene symbol on the Grid summary, displays text of Gene's full name and it's feature type(sto132)
    		<p concordion:assertTrue="false">passed jeffc 04/23/2014</p>
    		<ol>
                <li>Query for marker <b>pax*</b></li>
                <li>Hover your mouse over Gene Symbol <b>Pax1</b></li>
                <li>Verify that a popup displays saying <b>Name: paired box 1 Feature Type: protein coding gene </b></li>
                <li>Hover your mouse over Gene Symbol <b>PAX9</b></li>
                <li>Verify that a popup displays saying <b>Name: paired box 9 Feature Type: protein coding gene </b></li>
                <li>Query for location <b>Chr12:3000000-10000000</b></li>
                <li>Hover your mouse over Gene Symbol <b>Apob</b></li>
                <li>Verify that a popup displays saying <b>Name: apolipoprotein B Feature Type: protein coding gene</b></li>
                <li>Hover your mouse over Gene Symbol <b>OSR1</b></li>
                <li>Verify that a popup displays saying <b>Name: odd-skipped related transcription factor 1</b></li>
                <li>Query for disease <b>diabetes</b></li>
                <li>Hover your mouse over Gene Symbol <b>Ighm</b></li>
                <li>Verify that a popup displays saying <b>Name: immunoglobulin heavy constant mu Feature Type: gene segment</b></li>
                <li>Hover your mouse over Gene Symbol <b>CTSB</b></li>
                <li>Verify that a popup displays saying <b>Name: Cathepsin B</b></li>
    		</ol>
    		</li>
			<li>A link to IMSR for all strains or lines available is displayed for each mouse marker in the popup that has such a link on its Allele Detail Page.
 			(sto136)<p concordion:assertTrue="false">passed. -klf 2014-04-22</p> 
 			<ol>
 				<li>On the HMDC Portal, query for genes <b>h2-d1,Hspa1a,vpreb1,H3f3a,Olfr160,hamp2,clec10a,mgl2</b>.</li>
 				<li>From the Grid Tab, click the cell in the row for Human HLA-B and the column for hematopoietic system.</li>
 				<li>Verify that the pop-up contains IMSR links for <b>H2-D1</b> and <b>H2-K1</b>.  Verify that the count of "strains or lines" is the same as on an Allele Detail Page for those markers (5 and 3 as of 4/18/2014).</li>
 				<li>From the Grid Tab, click the cell in the row for for Human HLA-B and the column for nervous system.</li>
 				<li>Verify that the pop-up contains an IMSR links only for <b>H2-D1</b>.  Verify that the count of "strains or lines" is the same as on an Allele Detail Page for that marker (5 as of 4/18/2014).</li>
 				<li>From the Grid Tab, click the cell in the row for H3f3a, H3f3b and the column for growth/size.</li>
 				<li>Verify that the pop-up contains IMSR links for <b>H3f3a</b> and <b>H3f3b</b>.  Verify that the count of "strains or lines" is the same as on an Allele Detail Page for those markers (11 and 8 as of 4/18/2014).</li>
 				<li>Reset the query.  Query for gene <b>Bmp4</b>.</li>
 				<li>From the Grid Tab, click the cell in the row for Bmp4 and the column for behavior/neurological.</li>
 				<li>Verify that the pop-up contains an IMSR link for <b>Bmp4</b>.  Verify that the count of "strains or lines" is the same as on an Allele Detail Page for Bmp4 (4 as of 4/22/2014).</li>
 				<li>Click the cell in the row for Tg(Eno2-Bmp4)3Jake and the column for behavior/neurological.</li>
 				<li>Verify that the pop-up does not contain an IMSR link. (There are no strains at IMSR as of 4/18/2014.)</li>
 			</ol>
			</li>
    		<li>Verify these markers have the APF link or both Mutagenetix and APB links in the Alleles and Phenotypes ribbon and the link(s) works. (sto92)
 			<p concordion:assertTrue="false">passed jeffc 04/08/2014</p>
    		<ol>
                <li>Navigate to these marker detail pages.</li>
                <li>Verify the APF link exists and works (link will look like this: <b>Incidental mutations (data from APF )</b>):
                <table>
                	<tr><th>Marker Symbol</th><th>MGI ID</th></tr>
					<tr><td>Atp4a</td><td>MGI:88113</td></tr>
					<tr><td>Fitm2</td><td>MGI:2444508</td></tr>
					<tr><td>hyi</td><td>MGI:1915430</td></tr>
					<tr><td>Gm20388</td><td>MGI:5141853</td></tr>
					<tr><td>Lvrq1</td><td>MGI:3042875</td></tr>
					<tr><td>Gbx2</td><td>MGI:95668</td></tr>
					<tr><td>Rgs19</td><td>MGI:1915153</td></tr>
                </table>
                </li>
    		</ol>
    		</li>
            <li>For alleles, the "Type" column from a Quick Search displays the modified text from above, based on the allele's Generation Type. The allele's Subtypes (Attributes), if any, are ignored.
            <p concordion:assertTrue="true">klf 2014-03-10</p>
            <ol>
                <li>Open the MGI main page and enter the text indicated in the table below into the Quick Search box, Press search</li>
                <li>When the results display click the <b>Show all ..</b> link</li>
                <li>Verify the following results:
                <table>
					<tr>
						<th>Search for...</th>
						<th>Allele symbol</th>
						<th>Search Type</th>
						<th>Generation Type</th>
						<th>Note</th>
					</tr>
					<tr>
						<td>Pax6</td>
						<td>Pax6&lt;Sey-3H&gt;</td>
						<td>Chemically induced allele</td>
						<td>Chemically and radiation induced</td>
						<td></td>
					</tr>
					<tr>
						<td>Pax6</td>
						<td>Pax6&lt;Sey-Neu&gt;</td>
						<td>Chemically induced allele</td>
						<td>Chemically induced (ENU)</td>
						<td></td>
					</tr>
					<tr>
						<td>Pax6</td>
						<td>Pax6&lt;9Neu&gt;</td>
						<td>Chemically induced allele</td>
						<td>Chemically induced (other)</td>
						<td></td>
					</tr>
					<tr>
						<td>Pax6</td>
						<td>Pax6&lt;Gt(E049H01)Wrst&gt;</td>
						<td>Gene trapped allele</td>
						<td>Gene trapped</td>
						<td></td>
					</tr>
					<tr>
						<td>Pax6</td>
						<td>Pax6&lt;Sey-H&gt;</td>
						<td>Radiation induced allele</td>
						<td>Radiation induced</td>
						<td></td>
					</tr>
					<tr>
						<td>Pax6</td>
						<td>Pax6&lt;Sey&gt;</td>
						<td>Spontaneous allele</td>
						<td>Spontaneous</td>
						<td></td>
					</tr>
					<tr>
						<td>Pax6</td>
						<td>Pax6&lt;tm1Pgr&gt;</td>
						<td>Targeted allele</td>
						<td>Targeted</td>
						<td></td>
					</tr>
					<tr>
						<td>Pax6</td>
						<td>Tg(Pax6-cre,GFP)2Pgr</td>
						<td>Transgene</td>
						<td>Transgenic</td>
						<td></td>
					</tr>
					<tr>
						<td>Pax6</td>
						<td>Pax6&lt;Neu11&gt;-deleted</td>
						<td>column will be blank</td>
						<td>Not Specified</td>
						<td>Only displayed on production data search</td>
					</tr>
					<tr>
						<td>chab1</td>
						<td>chab1&lt;AKR/J&gt;</td>
						<td>QTL</td>
						<td>QTL</td>
						<td></td>
					</tr>
					<tr>
						<td>cript</td>
						<td>cript&lt;Tn(sb-rtTA)2427.SB1Ove&gt;</td>
						<td>Transposon induced allele</td>
						<td>Transposon induced</td>
						<td></td>
					</tr>
					<tr>
						<td></td>
						<td>none exist</td>
						<td>Not Applicable</td>
						<td>Not Applicable</td>
						<td>They exist in data but not visible in WI</td>
					</tr>
					<tr>
						<td></td>
						<td>none exist</td>
						<td>Endonuclease-mediated allele</td>
						<td>Endonuclease-mediated</td>
						<td>None exist yet</td>
					</tr>
					<tr>
						<td></td>
						<td>none exist</td>
						<td>Other</td>
						<td>Other</td>
						<td>None exist yet</td>
					</tr>
				</table>
				</li>
	            </ol>
            </li>

			<li>Expression data is displayed associated with EMAPA/EMAPS terms in the Quick Search.
            <p concordion:assertTrue="false">failing 2014-03-10</p>
            <p>Some example terms which have changed between the AD and EMAPA/EMAPS:</p>
            <table>
						<tr>
							<th>AD Term</th>
							<th>EMAPA Term</th>
							<th>EMAPA ID</th>
						</tr>
						<tr>
							<td>TS12: vitelline vein<br /> ...through...<br /> TS18:
								vitelline vein
							</td>
							<td>vitelline vein embryonic part TS12-18</td>
							<td>EMAPA:16244</td>
						</tr>
						<tr>
							<td>TS15: mesonephros mesenchyme<br /> ...through...<br />
								TS16: mesonephros mesenchyme
							</td>
							<td>mesonephros associated mesenchyme TS15-16</td>
							<td>EMAPA:16745</td>
						</tr>
						<tr>
							<td>TS15: telencephalon<br /> ...through...<br /> TS16:
								telencephalon
							</td>
							<td>early telencephalon TS15-16</td>
							<td>EMAPA:16652</td>
						</tr>
						<tr>
							<td>asdf</td>
							<td>adsf</td>
							<td>asdf</td>
						</tr>
					</table>
					<ol>
						<li>Navigate to the Quick Search and search for <b>vitelline vein embryonic part</b>.</li>
						<li>Verify the Vocabulary Terms "bucket" displays <b>vitelline vein embryonic part TS12-18</b>.</li>
						<li>Verify the Genome Features "bucket" displays the following markers, associated with the stage-specific EMAPS terms for this EMAPA term:
						<ul>
							<li>Abi3bp</li>
							<li>Aplnr</li>
							<li>Chst15</li>
							<li>Gata4</li>
							<li>Hhex</li>
							<li>Kdr</li>
							<li>Pitx2</li>
							<li>Rps28</li>
							<li>Stab2</li>
							<li>Tagln2</li>
							<li>Tal1</li>
							<li>Tgfb1</li>
							<li>Tgfb2</li>
						</ul>
						</li>
					</ol>
			</li>

			<li>Quick Search data is aligned with other data.
            <p concordion:assertTrue="false">failing 2014-03-11</p>
			<ol>
				<li>Quick Search: query for marker <b>Gm7092</b>.</li>
				<li>Verify quick search results show marker symbol <b>Erich4</b> (<a href="http://cardolan.informatics.jax.org/marker/MGI:3646269">MGI:3646269</a>).</li>
				
			</ol>
			</li>


    		
	</ol>
</div>

	<div id="auto">
		<h3>Automated tests</h3>
		<ol>
    		<li>Test 1
    		<ol>
                <li>Step 1</li>
                <li>Step 2</li>
    		</ol>
    		</li>
		</ol>
 <!--  EXAMPLE TEST CASE USING DYNAMIC DATA SPREADSHEET AT https://docs.google.com/spreadsheet/ccc?key=0ArepxibBtJW2dEl0eEg1bElOc1Mxc1FnNGtmSmlUSlE#gid=2 
     			<li>
    			Query for <b concordion:set="#structure">4-cell stage</b> should return 
    			<b cc:assertEquals="getResultCount(#structure)"><b cc:dynamicData="countFor4CellStageResults"></b></b> assay results, 
    			<b cc:assertEquals="getAssayCount(#structure)"><b cc:dynamicData="countFor4CellStageAssays"></b></b> assays, 
    			and <b cc:assertEquals="getGeneCount(#structure)"><b cc:dynamicData="countFor4CellStageGenes"></b></b> genes.<br/></li>
  -->
 	</div>

</div>
</body>
</html>