<html xmlns:concordion="http://www.concordion.org/2007/concordion" xmlns:cc="http://fewi.custom.commands.FewiExtensions">
<head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
    <!-- Not working on the server quite yet. Putting styles back into this local doc. 12/20/2012 klf.
     Put test suite specific styles here. To add global styles, see fewi/test/concordion.css -->
<style>
li {padding-left:2px; padding-top:0px; padding-bottom:0px; margin-left:25px;}
ol {padding: 0px; padding-bottom:3px; padding-top:0px; margin: 0px;}
ul {padding: 0px; padding-bottom:3px; padding-top:0px; margin: 0px;}
p {margin:2px 0px;}
table {border-collapse: collapse; padding-bottom:5px; margin-bottom:5px;}
table, th, td {border: 1px solid #CCCCCC;}
td, th {padding:2px;}
td>ul {padding: 0px; padding-bottom:1px; padding-top:0px; margin: 0px;}
td>ul li {padding-left:2px; padding-top:0px; padding-bottom:0px; margin-left:10px;}
td.xl65 {background-color:#FFFF00;}


h5 {font-size:1em; font-variant:small-caps; margin-top:4px; padding-top:4px;}
tr.small th, tr.small td {font-size:0.75em;}
tr.blue th {background-color:#D0E0F0;}
tr.gray td {background-color:#EEEEEE;}

.hide {display:none;}
.hilite {background-color:#FFFF00;} /*bright yellow*/
.lowlite {background-color:#FFFFDD;} /*palest yellow*/
.man {background-color:#FFFFCC;} /*less-pale yellow*/

.regressHi {background-color:#FFAB57;} /*medium-orange*/
.regressLo {background-color:#90DDF5;} /*light blue*/
</style>
<!-- declare strict doctype to look for disallowed html errors
 **  **  **  **  ** REMOVE THE LINE before committing the file **  **  **  **  **  **  **
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
 **  **  **  **  ** REMOVE THE LINE before committing the file **  **  **  **  **  **  **
 -->


    <title>Load Allele Collections</title>
</head>

<body>

<div id="TOC">
	<p>Contents: <a href="#maintitle">Title</a> | <a href="#acceptance">Acceptance Criteria</a> | <a href="#manual">Manual Tests</a> | <a href="#auto">Automated Tests</a></p>
</div>

<div id="maintitle">
    <h1>Load Allele Collections</h1>
    <ul>
        <li>User story 15, E3: Implement schema change and data migration</li>
    </ul>
</div>

<div id="acceptance">
    <h2>Acceptance Criteria</h2>
	<ul>
		<li>new vocab is in the prod db</li>
		<li>schema supports an allele having a collection</li>
		<li>NOTE: will have decided the 1 vs many issue <i>(decided: an allele may have exactly 0 or 1 collection --Cindy)</i></li>
		<li>rules for autoassignment of collections for selected sets of allele defined and implemented</li>
		<li>data migration script exists and has been run</li>
		<li>verify via SQL</li>
	</ul>
</div>



<div id="tests">
    <h2>Tests</h2>
    <p>Skip to <a href="#manual">Manual Tests</a> |  <a href="#auto">Automated Tests</a></p>
	<div id="manual">
<!-- To insert manual test results, use the following:
<p concordion:assertTrue="true">testing assertTrue - displays as green</p>
<p concordion:assertTrue="false">testing assertFalse - displays as red</p>
 -->
		<h3>Manual tests</h3>
		<ol>
    		<li>A new vocabulary for Allele Collections exists and contains the values specified below.
 			<p concordion:assertTrue="true">jeffc 12/30/2013</p>
    		<ol>
                <li>Query the Sybase VOC_Vocab table to verify the new Allele Collection vocabulary exists.</li>
                <li>Query the Sybase VOC_Term table to verify the following terms exist for that vocabulary <i>(list as of 12/30/2013)</i>:
                <ul>
					<li>Australian PhenomeBank</li>
					<li>B2B/CvDC</li>
					<li>Beutler Mutagenix</li>
					<li>Deltagen</li>
					<li>EUCOMM</li>
					<li>GENSAT</li>
					<li>Harwell ENU Mutagenesis</li>
					<li>KOMP-CSD</li>
					<li>KOMP-Regeneron</li>
					<li>Lexicon</li>
					<li>Mutagenesis for Dev. Defects</li>
					<li>Neuroscience Blueprint cre</li>
					<li>NorCOMM</li>
					<li>Pleiades Promoter Project</li>
					<li>RIKEN GSC ENU Project</li>
					<li>Sanger Inst. Gene Trap Res.</li>
					<li>Sanger miRNA knockouts</li>
					<li>Not Specified</li>
                </ul>
                </li>
                <li><a href="http://rohan.informatics.jax.org/usrlocalmgi/scrum-dog/dbutils/mgidbmigration/TR11515/alleletype/alleletype-after-SQL.csh.log">Collection list can be found here</a></li>
    		</ol>
    		</li>

    		<li>No allele has more than one Collection value
 			<p concordion:assertTrue="true">jeffc 12/30/2013</p>
    		<ol>
                <li>Tested via migration script.</li>
    		</ol>
    		</li>

    		<li>Every allele in the database has a Collection value.
 			<p concordion:assertTrue="true">jeffc 12/30/2013</p>
    		<ol>
                <li>Tested via migration script.</li>
    		</ol>
    		</li>

    		<li>Every Collection value in the database is associated with at least 1 allele.
 			<p concordion:assertTrue="true">jeffc 12/30/2013</p>
    		<ol>
                <li>Tested via migration script.</li>
    		</ol>
    		</li>

    		<li>The following alleles are assigned to the indicated Collection. 
 			<p concordion:assertTrue="true">jeffc 12/30/2013</p>
    		<ol>
                <li>Query Sybase ALL_Allele for the appropriate _Collection_key for each allele.</li>
                <li>
                <table>
                	<tr><th>Allele symbol</th><th>Allele MGI ID</th><th>Collection</th><th>Notes for Tester</th></tr>
                	<tr><td>Btk<sup>vila</sup></td><td>MGI:4820731</td><td>Australian PhenomeBank</td><td>List (or J:104190?)</td></tr>
                	<tr><td>Cfc1<sup>b2b970Clo</sup></td><td>MGI:5311378</td><td>B2B/CvDC</td><td>J:175213</td></tr>
                	<tr><td>Cd40lg<sup>m1Btlr</sup></td><td>MGI:4819117</td><td>Beutler Mutagenix</td><td>List</td></tr>
                	<tr><td>Gabrp<sup>tm1Dgen</sup></td><td>MGI:3604585</td><td>Deltagen</td><td>By name</td></tr>
                	<tr><td>Bcap31<sup>tm1a(EUCOMM)Wtsi</sup></td><td>MGI:4432267</td><td>EUCOMM</td><td>By logical DB</td></tr>
                	<tr><td>Tg(Gpr12-EGFP)LD58Gsat</td><td>MGI:4847093</td><td>GENSAT</td><td>By name</td></tr>
                	<tr><td>Gena160</td><td>MGI:5490100</td><td>Harwell ENU Mutagenesis</td><td>List</td></tr>
                	<tr><td>Kank3<sup>tm1(KOMP)Mbp</sup></td><td>MGI:4453999</td><td>KOMP-CSD</td><td>By logical DB</td></tr>
                	<tr><td>Arf3<sup>tm1(KOMP)Vlcg</sup></td><td>MGI:5085749</td><td>KOMP-Regeneron</td><td>By logical DB</td></tr>
                	<tr><td>Crtam<sup>tm1Lex</sup></td><td>MGI:5007093</td><td>Lexicon</td><td>By name</td></tr>
                	<tr><td>Mc4r<sup>Glu3</sup></td><td>MGI:2671841</td><td>Mutagenesis for Dev. Defects</td><td>J:89098, <b>J:85113</b>, J:109155, J:161341</td></tr>
					<tr><td>nur33</td><td>MGI:3038764</td><td>Mutagenesis for Dev. Defects</td><td><b>J:89098</b>, J:85113, J:109155, J:161341</td></tr>
					<tr><td>--</td><td>--</td><td>Mutagenesis for Dev. Defects</td><td>J:89098, J:85113, <b>J:109155</b>, J:161341<br/>All alleles also have one other J: number.</td></tr>
					<tr><td>--</td><td>--</td><td>Mutagenesis for Dev. Defects</td><td>J:89098, J:85113, J:109155, <b>J:161341</b><br/>All alleles also have one other listed J: number.</td></tr>
                	<tr><td>Calb2<sup>tm2.1(cre/ERT2)Zjh</sup></td><td>MGI:4880758</td><td>Neuroscience Blueprint cre</td><td>List</td></tr>
                	<tr><td>Iqcd<sup>tm1(NCOM)Mfgc</sup></td><td>MGI:4880499</td><td>NorCOMM</td><td>By logical DB</td></tr>
                	<tr><td>Hprt<sup>tm151(Ple99-EGFP)Ems</sup></td><td>MGI:5303234</td><td>Pleiades Promoter Project</td><td>List</td></tr>
                	<tr><td>Rom1<sup>Rgsc1156</sup></td><td>MGI:3814074</td><td>RIKEN GSC ENU Project</td><td>By name and allele type</td></tr>
                	<tr><td>Aebp2<sup>Gt(BC0681)Wtsi</sup></td><td>MGI:3871767</td><td>Sanger Inst. Gene Trap Res.</td><td>By name and allele type</td></tr>
                	<tr><td>Mir133a-1<sup>tm1Wtsi</sup></td><td>MGI:5320126</td><td>Sanger miRNA knockouts</td><td>List</td></tr>
                	<tr><td>Adsl<sup>Gt(AW0008)Wtsi</sup></td><td>MGI:3857089</td><td>Not Specified</td><td>this one should fail because of wrong collection</td></tr>
                	<tr><td>Grm1<sup>tvrm84</sup></td><td>MGI:4867882</td><td>Not Specified</td><td></td></tr>
                </table>
                </li>
                <li><a href="http://rohan.informatics.jax.org/usrlocalmgi/scrum-dog/dbutils/mgidbmigration/TR11515/allelecollection/allelecollectionTests.csh.log.sorted">Passed test results</a></li>
    		</ol>
    		</li>

    		<li>Verify collection counts are close to expected counts
 			<p concordion:assertTrue="true">jeffc 12/30/2013</p>
    		<ol>
    		<li>
                <table>
                <tr><th>Collection Name</th><th>Estimated Count</th><th>Real Count</th></tr>
        		<tr><td>KOMP-Regeneron</td><td>about 4100</td><td>4141</td></tr>
        		<tr><td>KOMP-CSD</td><td>abour 10000</td><td>10626</td></tr> 
        		<tr><td>NorCOMM</td><td>about 580</td><td>580</td></tr> 
        		<tr><td>EUCOMM</td><td>about 16000</td><td>17372</td></tr> 
        		<tr><td>B2B/CvDC</td><td>172</td><td>202</td></tr> 
        		<tr><td>Pleiades Promoter Project</td><td>about 300</td><td>325</td></tr> 
        		<tr><td>Beutler Mutagenix</td><td>few 100</td><td>191</td></tr> 
        		<tr><td>Mutagenesis for Dev. Defects</td><td>few 100</td><td>396</td></tr> 
        		<tr><td>Australian PhenomeBank</td><td>about 100</td><td>54</td></tr> 
        		<tr><td>Deltagen</td><td>234</td><td>234</td></tr> 
        		<tr><td>Lexicon</td><td>a lot</td><td>199940</td></tr> 
        		<tr><td>Neuroscience Blueprint cre</td><td>34 or more</td><td>34</td></tr> 
        		<tr><td>GENSAT</td><td>about 1350</td><td>1366</td></tr> 
        		<tr><td>Harwell ENU Mutagenesis</td><td>about 250</td><td>234</td></tr> 
        		<tr><td>Sanger miRNA knockouts</td><td>about 250</td><td>255</td></tr> 
        		<tr><td>Sanger Inst. Gene Trap Res.</td><td>a lot</td><td>12088</td></tr> 
        		<tr><td>RIKEN GSC ENU Project</td><td>about 350</td><td>349</td></tr> 
        		<tr><td>Not Specified</td><td>unknown</td><td>560483</td></tr> 
        </table></li>
			<li><a href="http://rohan.informatics.jax.org/usrlocalmgi/scrum-dog/dbutils/mgidbmigration/TR11515/alleletype/alleletype-after-SQL.csh.log">Collection counts can be found here</a></li>
		    	</ol>
    			</li>
		</ol>
	</div>
	<div id="auto">
		<h3>Automated tests</h3>
		<ol>
    		<li>Test 1
    		<ol>
                <li>Step 1</li>
                <li>Step 2</li>
    		</ol>
    		</li>
		</ol>
 	</div>
</div>

</body>
</html>