<html xmlns:concordion="http://www.concordion.org/2007/concordion" xmlns:cc="http://fewi.custom.commands.FewiExtensions">
<head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
    <!-- Not working on the server quite yet. Putting styles back into this local doc. 12/20/2012 klf.
     Put test suite specific styles here. To add global styles, see fewi/test/concordion.css -->
<style>
li {padding-left:2px; padding-top:0px; padding-bottom:0px; margin-left:25px;}
ol {padding: 0px; padding-bottom:3px; padding-top:0px; margin: 0px;}
ul {padding: 0px; padding-bottom:3px; padding-top:0px; margin: 0px;}
p {margin:2px 0px;}
table {border-collapse: collapse; padding-bottom:5px; margin-bottom:5px;}
table, th, td {border: 1px solid #CCCCCC;}
td, th {padding:2px;}
td>ul {padding: 0px; padding-bottom:1px; padding-top:0px; margin: 0px;}
td>ul li {padding-left:2px; padding-top:0px; padding-bottom:0px; margin-left:10px;}
td.xl65 {background-color:#FFFF00;}
dt {font-weight:bold;}


h5 {font-size:1em; font-variant:small-caps; margin-top:4px; padding-top:4px;}
tr.small th, tr.small td {font-size:0.75em;}
tr.blue th {background-color:#D0E0F0;}
tr.gray td {background-color:#EEEEEE;}

.hide {display:none;}
.hilite {background-color:#FFFF00;} /*bright yellow*/
.lowlite {background-color:#FFFFDD;} /*palest yellow*/
.man {background-color:#FFFFCC;} /*less-pale yellow*/

.regressHi {background-color:#FFAB57;} /*medium-orange*/
.regressLo {background-color:#90DDF5;} /*light blue*/
</style>
<!-- declare strict doctype to look for disallowed html errors
 **  **  **  **  ** REMOVE THE LINE before committing the file **  **  **  **  **  **  **
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
 **  **  **  **  ** REMOVE THE LINE before committing the file **  **  **  **  **  **  **
 -->


    <title>HDP File Uploads</title>
</head>

<body>
<div id="TOC">
	<p>Contents: <a href="#maintitle">Title</a> | <a href="#acceptance">Acceptance Criteria</a> | <a href="#manual">Manual Tests</a> | <a href="#auto">Automated Tests</a></p>
</div>

<div id="maintitle">
    <h1>HDP File Uploads</h1>
    <ul>
		<li>User Story 158, Query form: Upload files of coordinates - VCF Format (5.16)</li>
		<li>User story 129, HDP: Upload file of gene symbols/ids (5.18)</li>
		<li>User story 130, HDP: VCF upload: allow user to disable filters (5.18)</li>
    </ul>
</div>

<div id="acceptance">
    <h2>Acceptance Criteria</h2>
	<ul>
		<li>location query offers a file upload option</li>
		<li>the reset button should also reset the file upload (right now it doesn't?)</li>
		<li>only one input (text entry or file upload) is supported. E.g.:
		<ul>
			<li>if text entered, disable file upload</li>
			<li>if file specified, disable text entry</li>
		</ul>
		</li>
		<li>only VCF files support (for now)</li>
		<li>hard-coded (not user settable) filters apply:
		<ul>
			<li>the FILTER column must empty for ALL rows, or must contain "PASS" in this row</li>
			<li>the ID column must be empty </li>
			<li>(Remember that "." counts as empty)</li>
		</ul>
		</li>
		<li>(sto129) nomenclature search offers a file upload option</li>
		<li>(sto129) accepts whatever the type-in accepts</li>
		<li>(sto129) single column file</li>
		<li>(sto129) only limited by system resources</li>
		<li>(sto130) default is still to filter as usual</li>
		<li>(sto130) if "disable" is selected, all input rows are used for the query</li>
		<li>(sto130) absolute file size limit still applies (200k rows?)</li>
	</ul>

</div>


<div id="tests">
    <h2>Tests</h2>
	<div id="manual">
<!-- To insert manual test results, use the following:
<p concordion:assertTrue="true">testing assertTrue - displays as green</p>
<p concordion:assertTrue="false">testing assertFalse - displays as red</p>
 -->
		<h3>Manual and Automated tests</h3>
		<p>Jump to: <a href="#vcf">VCF File</a> tests | <a href="#gene">Gene File</a> tests</p>
		
		<h4 id="vcf">VCF File Uploads</h4>
		<ol>
    		<li>Verify a "Choose File" button exists on the Human Disease Query page(sto158)
    		<p concordion:assertTrue="true">passed. -klf Dec 2013</p>
    		<ol>
                <li>Open the Human Disease query page</li>
                <li>Verify the existence of a "Choose File" button.</li>
    		</ol>
    		</li>

    		<li>Verify you can upload a VCF file with human data (sto158)
    		<p concordion:assertTrue="true">passed. -klf Dec 2013</p>
    		<ol>
                <li>Open the Human Disease query page and click the "Choose File" button</li>
                <li>Navigate to the TR directory /wts_projects/11400/11423/data_files/VCF_Files_From_Comp_Sci and click on the file <b>human1.vcf</b></li>
                <li>Verify the file is loaded properly by getting the popup <b>file successfully processed</b></li>
                <li>Click the "OK" button on the popup</li>
                <li>Verify the filename <b>human1.vcf</b> is displayed on the form page.</li>
    		</ol>
    		</li>

    		<li>Verify you can upload a VCF file with mouse data (sto158)
    		<p concordion:assertTrue="true">passed. -klf Dec 2013</p>
    		<ol>
                <li>Open the Human Disease query page and click the "Choose File" button</li>
                <li>Navigate to the TR directory /wts_projects/11400/11423/data_files/VCF_Files_From_Comp_Sci and click on the file <b>mouse1.vcf</b></li>
                <li>Verify the file is loaded properly by getting the popup <b>file successfully processed</b></li>
                <li>Click the "OK" button on the popup</li>
                <li>Verify the filename <b>mouse1.vcf</b> is displayed on the form page.</li>
    		</ol>
    		</li>

     		<li>Verify you get an error when trying to upload a VCF file with no data (sto158)
    		<p concordion:assertTrue="true">passed. -klf Dec 2013</p>
    		<ol>
                <li>Open the Human Disease query page and click the "Choose File" button</li>
                <li>Navigate to the TR directory /wts_projects/11400/11423/data_files/VCF_Files_for_Testing and click on the file <b>error.vcf</b></li>
                <li>Verify you get the error message <b>ERROR: No coordinates found in VCF file. Ensure file is in the correct format.</b></li>
    		</ol>
    		</li>

     		<li>Verify VCF file upload preprocesses the file keeping only "PASS" (case-insensitive), "." (dot) or null in Column 7 ("FILTER" column) (sto158)
    		<p concordion:assertTrue="true">passed. -klf Dec 2013</p>
    		<ol>
                <li>Open the Human Disease query page and click the "Choose File" button</li>
                <li>Navigate to the TR directory /wts_projects/11400/11423/data_files/VCF_Files_for_Testing and click on the file <b>human1-testPass6pass4Dot3.vcf</b></li>
                <li>This file should upload without error.  Select "human" coordinates and click "GO".</li>
                <li>The search should result in 6 rows in the Grid and 19 genes on the Genes tab.</li>
                <li>Navigate to the TR directory /wts_projects/11400/11423/data_files/VCF_Files_for_Testing and click on the file <b>human1-testgoodRows15.vcf</b></li>
                <li>This file should upload without error.  Select "human" coordinates and click "GO".</li>
                <li>Verify that the search results in the same 6 rows in the Grid and 19 genes on the Genes tab.</li>
    		</ol>
    		</li>
            
     		<li>Verify VCF file upload preprocesses the file keeping only rows with "." (dot) or null in Column 3 ("ID" column) (sto158)
    		<p concordion:assertTrue="true">passed. -klf Dec 2013</p>
    		<ol>
                <li>Open the Human Disease query page and click the "Choose File" button</li>
                <li>Navigate to the TR directory /wts_projects/11400/11423/data_files/VCF_Files_for_Testing and click on the file <b>human1-TEST-rsIDTestCol3.vcf</b></li>
                <li>This file should upload without error.  Select "human" coordinates and click "GO".</li>
                <li>Verify that the following genes ARE the ONLY genes returned in the Grid (<i>not in correct order</i>):
                <table>
                	<tr><th>Gene symbol</th></tr>
                	<tr><td>PRMT6</td></tr>
                	<tr><td>NOTCH2</td></tr>
                	<tr><td>ADAR</td></tr>
                	<tr><td>SH2D2A</td></tr>
                	<tr><td>NTRK1</td></tr>
                </table>
                </li>
                <li>Verify that the following genes ARE NOT returned in the Grid (<i>not in correct order</i>):
                <table>
                	<tr><th>Gene symbol</th></tr>
                	<tr><td>DDR2</td></tr>
                	<tr><td>RGL1</td></tr>
                	<tr><td>HMCN1</td></tr>
                </table>
                </li>
    		</ol>
    		</li>
    		
    		<li>Verify VCF files over 25MB are rejected.
    		<p concordion:assertTrue="true">passed. -klf Dec 2013</p>
    		<ol>
                <li>Open the Human Disease query page and click the "Choose File" button</li>
                <li>Navigate to the TR directory /wts_projects/11400/11423/data_files/VCF_Files_for_Testing and click on the file <b>VeryVeryLargeFile_human.vcf</b></li>
                <li>Verify an error message is displayed like "ERROR: Max file upload limit of 25MB exceeded."</li>
                <li>Verify clicking the "OK" button dismisses the error message.</li>
                <li>Verify the filename "VeryVeryLargeFile_human.vcf" is NOT displayed on the interface.</li>
                <li>Verify clicking the "GO" button returns no data (i.e., the file was not processed and uploaded).</li> 
    		</ol>
    		</li>

<!-- 
17921 Rows Ignored Due to Non-Empty ID Column
0 Rows Ignored Due to FILTER Other Than 'PASS'
 -->
			<li>Checkbox exists to disable the filters (so do not excludes rows having an RSID value, or rows having a Filter value other than 'PASS'). Checking the box causes a file upload to process (include) rows which would have been excluded. (sto130) 
    		<p concordion:assertTrue="false">new</p>
    		<ol>
                <li>Create a VCF test file with rows:
                <ul>
                	<li>Having RSID and Filter = 'PASS' - 2 rows.</li>
                	<li>Having no RSID and Filter = 'PASS' - 2 rows.</li>
                	<li>Having RSID and Filter other than 'PASS' - 2 rows.</li>
                	<li>Having no RSID and Filter other than 'PASS' - 2 rows.</li>
                </ul> 
                </li>
                <li>Click the "Choose File" button and select the test file.</li>
                <li>Verify that the "File successfully processed." dialog reports: 2 coordinates included;  6 rows ignored.</li>
                <li>Clear the query to begin again.</li>
                <li>Open the Human Disease query page and select the checkbox to disable the filters.</li>
                <li>Click the "Choose File" button and select the test file.</li>
                <li>Verify that the "File successfully processed." dialog reports: 8 coordinates included;  0 rows ignored.</li>
            </ol>
			</li>

			<li>Uploading a VCF file and then checking the box causes the file to be reprocessed. (sto130) 
    		<p concordion:assertTrue="false">new</p>
    		<ol>
                <li>Click the "Choose File" button and select the test file from above.</li>
                <li>Verify that the "File successfully processed." dialog reports: 2 coordinates included;  6 rows ignored.</li>
                <li>Select the checkbox to disable the filters.</li>
                <li>Verify that the "File successfully processed." dialog appears again and reports: 8 coordinates included;  0 rows ignored.</li>
            </ol>
			</li>

			<li>Test
    		<p concordion:assertTrue="false">new</p>
    		<ol>
                <li>step</li>
            </ol>
			</li>

			<li>Test
    		<p concordion:assertTrue="false">new</p>
    		<ol>
                <li>step</li>
            </ol>
			</li>

		</ol>

		<h4 id="gene">Gene File Uploads</h4>
		<ol>
    		<li>Uploading a tab-delimited file containing a single column listing gene symbols returns the same results as the same gene symbols would if typed into the interface input field. (sto129)
    		<p concordion:assertTrue="true">passed. -klf 4/17/2014</p>
    		<ol>
                <li>On the Human-Mouse: Disease Connection page, enter <b>105830</b> (Angelman Syndrome; AS) in the "Search by disease or phenotype terms" input area and submit the query.</li>
                <li>On the Genes tab, click the button to "Export: Text File". Edit to include only a single column of data containing gene symbols (some mouse, some human). Save the file.<br/>
                	(<em>Note: as of 4/17/2014 there were 5 genes with 4 unique gene symbols associated with this disease.</em>)</li>
                <li>On the Human-Mouse: Disease Connection page, click the button to upload a file of genes.</li>
                <li>Select the file from step 2 and upload it.</li>
                <li>Verify that the "processing" window reports the same number of rows as are in the file. Click the "OK" button to continue.</li>
                <li>Click the "GO" button to submit the file.</li>
                <li>Note the number of results returned on the <b>Grid tab</b>, on the <b>Genes tab</b>, and on the <b>Diseases tab</b>.</li>
                <li>Clear the HMDC query form.</li>
                <li>From the upload file in step 2, copy the marker symbols and paste into the "Search by gene" input field and click the "Go" button to submit the query.</li>
                <li>Verify that the same number of results were returned on the <b>Grid tab</b>, on the <b>Genes tab</b>, and on the <b>Diseases tab</b>.</li>
            </ol>
    		</li>

    		<li>Uploading a tab-delimited file containing a single column listing gene IDs returns the same results as the same gene IDs would if typed into the interface input field. (sto129)
    		<p concordion:assertTrue="true">passed. -klf 4/17/2014</p>
    		<ol>
                <li>On the Marker Query Form, enter <b>168600</b> (Parkinson Disease, Late-Onset; PD) in the "Phenotype/Human Disease" input area and submit the query.</li>
                <li>From the results summary table, click the button to "Export: Text File". Edit to include only a single column of data containing gene IDs. Save the file.<br/>
                	(<em>Note: as of 4/17/2014 there were 5 markers associated with this disease.</em>)</li>
                <li>On the Human-Mouse: Disease Connection page, click the button to upload a file of genes.</li>
                <li>Select the file from step 2 and upload it.</li>
                <li>Verify that the "processing" window reports the same number of rows as are in the file. Click the "OK" button to continue.</li>
                <li>Click the "GO" button to submit the file.</li>
                <li>Note the number of results returned on the <b>Grid tab</b>, on the <b>Genes tab</b>, and on the <b>Diseases tab</b>.</li>
                <li>Clear the HMDC query form.</li>
                <li>From the upload file in step 2, copy the marker IDs and paste into the "Search by gene" input field and click the "Go" button to submit the query.</li>
                <li>Verify that the same number of results were returned on the <b>Grid tab</b>, on the <b>Genes tab</b>, and on the <b>Diseases tab</b>.</li>
            </ol>
    		</li>

    		<li>The "Reset" button clears an uploaded file of gene inputs from the cache. (sto129)
    		<p concordion:assertTrue="true">passed. -klf 4/18/2014</p>
    		<ol>
                <li>From the test above, select the file from step 2 and upload it.</li>
                <li>Verify that the "processing" window reports the same number of rows as are in the file. Click the "OK" button to continue.</li>
                <li>Click the "GO" button to submit the file.</li>
                <li>Open the query form area and click the "Reset" button.</li>
                <li>Verify that the filename is removed from the interface.</li>
                <li>Verify the interface displays text "No file chosen".</li>
                <li>Click the "Go" button.</li>
                <li>Verify no genes are returned.</li>
            </ol>
    		</li>

    	</ol>	
	</div>
</div>

</body>
</html>