<html xmlns:concordion="http://www.concordion.org/2007/concordion" xmlns:cc="http://fewi.custom.commands.FewiExtensions">
<head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
    <!-- Not working on the server quite yet. Putting styles back into this local doc. 12/20/2012 klf.
     Put test suite specific styles here. To add global styles, see fewi/test/concordion.css -->
<style>
li {padding-left:2px; padding-top:0px; padding-bottom:0px; margin-left:25px;}
ol {padding: 0px; padding-bottom:3px; padding-top:0px; margin: 0px;}
ul {padding: 0px; padding-bottom:3px; padding-top:0px; margin: 0px;}
p {margin:2px 0px;}
table {border-collapse: collapse; padding-bottom:5px; margin-bottom:5px;}
table, th, td {border: 1px solid #CCCCCC;}
td, th {padding:2px;}
td>ul {padding: 0px; padding-bottom:1px; padding-top:0px; margin: 0px;}
td>ul li {padding-left:2px; padding-top:0px; padding-bottom:0px; margin-left:10px;}
td.xl65 {background-color:#FFFF00;}


h5 {font-size:1em; font-variant:small-caps; margin-top:4px; padding-top:4px;}
tr.small th, tr.small td {font-size:0.75em;}
tr.blue th {background-color:#D0E0F0;}
tr.gray td {background-color:#EEEEEE;}

.hide {display:none;}
.hilite {background-color:#FFFF00;} /*bright yellow*/
.lowlite {background-color:#FFFFDD;} /*palest yellow*/
.man {background-color:#FFFFCC;} /*less-pale yellow*/

.regressHi {background-color:#FFAB57;} /*medium-orange*/
.regressLo {background-color:#90DDF5;} /*light blue*/
</style>
<!-- declare strict doctype to look for disallowed html errors
 **  **  **  **  ** REMOVE THE LINE before committing the file **  **  **  **  **  **  **
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
 **  **  **  **  ** REMOVE THE LINE before committing the file **  **  **  **  **  **  **
 -->


    <title>Popup: Disease Data</title>
</head>

<body>

<div id="TOC">
	<p>Contents: <a href="#maintitle">Title</a> | <a href="#acceptance">Acceptance Criteria</a> | <a href="#manual">Manual Tests</a> | <a href="#auto">Automated Tests</a></p>
</div>

<div id="maintitle">
    <h1>Popup: Disease Data</h1>
    <ul>
        <li>User story 34, E4: Drill Down page -  OMIM Xs - debugging display</li>
        <li>User story 185, BUG: genoclusters in popups should only be those that matched the query</li>
        <li>sto117: Grid: Disease clumps : front end</li>
        <li>User story 196, Tagging of Genotypes in popups</li>
        <li>User story 195, Sort order of Genotypes in popups</li>
        <li>User story 171, Popups: headers and styling</li>
        <li>User story 136, HDP: Add IMSR links to popup grids (MGI 5.18)</li>
    </ul>
</div>

<div id="acceptance">
    <h2>Acceptance Criteria</h2>
	<ul>
		<li>shows all the OMG genotypes</li>
		<li>shows the human gene associations</li>
		<li>shows the disease</li>
		<li>sto185: when phenotypes/diseases are queried, we only displayed the genoclusters that match, and only columns that have data for the matched genoclusters</li>
		<li>sto196: Chimeric genotypes are identified on the Genocluster popup page</li>
		<li>sto196: Conditional genotypes are identified on the Genocluster popup page</li>
		<li>sto117: query semantics unchanged; groups are for display purposes only</li>
		<li>sto117: grid disease columns show clump terms rather than indiviual diseases</li>
		<li>sto117: [grid] cells are rollups of the checkmarks for the individual matching diseases</li>
		<li>sto117: the counts used for coloring the cell are the sums of counts for the underlying individual disease cells</li>
		<li>sto117: popups show the individual matching diseases for the group, the matching genotypes, and the matching human genes</li>
		<li>sto117: specific formatting not specified here</li>
		<li>sto196: Chimeric genotypes are identified on the Genocluster popup page</li>
		<li>sto196: Conditional genotypes are identified on the Genocluster popup page</li>
		<li>sto195: Genoclusters in a popup are sorted by allele state first, then alpha be (1st) allele symbol.</li>
		<li>sto195: use "smart" alpha sorting</li>
		<li>sto195: The sort order of states is: hom, het, hemi, conditional [chimeric: let them fall where they may]</li>	
		<li>sto171: [none specified]</li>
		<li>sto136: popup grid includes links to IMSR (MGI 5.18)</li>
		<li>sto136: one link per gene on the page (MGI 5.18)</li>
		<li>sto136: link searches IMSR for all mice (not ES cells) carrying any mutation in that gene (this is the same link as on the allele detail page) (MGI 5.18)</li>
	</ul>

	<h2>Specific Requirements</h2>
	<ul>
		<li>See <a href="HDPGridPopupsPhenotype.html">specific requirements</a>  dealing with which genotypes and associated phenotype or disease data to display, order, labeling,  grid color binning, and other details.</li>
	</ul>
</div>



<div id="tests">

    <h2>Tests</h2>
    <p>Skip to <a href="#manual">Manual Tests</a> |  <a href="#auto">Automated Tests</a></p>
	<div id="manual">
<!-- To insert manual test results, use the following:
<p concordion:manual="true">testing assertTrue - displays as green</p>
<p concordion:manual="false">testing assertFalse - displays as red</p>
 -->
		<h3>Manual tests</h3>
		<ol>
    		<li>Clicking a checkmark causes the popup to appear as a new, separate browser window.
 			<p concordion:manual="true">klf 9/10/2013</p> <i>(cannot automate)</i>
    		<ol>
                <li>Open the HDP query form.</li>
                <li>Enter <b>Gata1</b> in the Genes section and hit search.</li>
                <li>On the grid tab, click the checkmark at the intersection of <b>Gata1</b> and <b>Myelofibrosis</b>.</li>
                <li>Verify that a new window opens.</li>
    		</ol>
    		</li>
		
    		<li>Chimeric genotypes - In the popup, a checkmark appears for a &quot;compressed&quot; genotype and OMIM term for each OMG of the gene with mouse disease model annotations.  Chimeric genotypes are NOT compressed with other kinds of genotypes. (sto196)<br/>
    		(Note: There are no chimeric genotypes having the same allele pair as a non-chimeric genotype, so cannot verify that these remain separated.) 
 			<p concordion:manual="true">klf 11/04/2013</p> <i>(semi-automated)</i>
    		<ol>
                <li>Open the HDP query form</li>
                <li>Enter <b concordion:set="#pheno">135100</b> (Fibrodysplasia Ossificans Progressiva; FOP) in the Phenotypes and Diseases section and hit search</li>
                <li>Click the checkmark at the intersection of the <b concordion:set="#row">Acvr1</b> row and <b concordion:set="#column">Fibrodysplasia Ossificans Progressiva; FOP</b> column. </li>
                <li>Confirm that <b>1</b> genotype row is displayed for 
                	<b cc:assertValueAppearsOnlyOnce="diseaseGridPopupGenoClustersByPheno(#pheno,#row,#column)">Acvr1&lt;tm1Emsh&gt;/Acvr1&lt;+&gt;</b> (1 genotype exists).</li>
                <li>Confirm that the text "<b>(chimeric)</b>" appears following the genotype.  (<span concordion:manual="true">klf</span>)</li>
                <li>Confirm that human <b cc:assertResultsContain="diseaseGridPopupHumanMarkersByPheno(#pheno,#row,#column)">ACVR1</b> IS LISTED.</li>
                <li>Clear the query form.</li>
                <li>Enter <b concordion:set="#pheno">305600</b> (Focal Dermal Hypoplasia; FDH) in the Phenotypes and Diseases section and hit search</li>
                <li>Click the checkmark at the intersection of the <b concordion:set="#row">Porcn</b> row and <b concordion:set="#column">Focal Dermal Hypoplasia; FDH</b> column. </li>
                <li>Confirm that <b>1</b> genotype row is displayed for 
                	<b cc:assertValueAppearsOnlyOnce="diseaseGridPopupGenoClustersByPheno(#pheno,#row,#column)">Porcn&lt;tm1Vdv&gt;/Porcn&lt;+&gt;</b> (1 genotype exists).</li>
                <li>Confirm that the text "<b>(chimeric)</b>" appears following the genotype.  (<span concordion:manual="true">klf</span>)</li>
                <li>Confirm that <b>1</b> genotype row is displayed for 
                	<b cc:assertValueAppearsOnlyOnce="diseaseGridPopupGenoClustersByPheno(#pheno,#row,#column)">Porcn&lt;tm1Vdv&gt;/Y</b> (1 genotype exists).</li>
                <li>Confirm that the text "<b>(chimeric)</b>" appears following the genotype.  (<span concordion:manual="true">klf</span>)</li>
                <li>Confirm that human <b cc:assertResultsContain="diseaseGridPopupHumanMarkersByPheno(#pheno,#row,#column)">PORCN</b> IS LISTED.</li>
    		</ol>
    		</li>
    		
    		<li>Conditional genotypes - In the popup, a checkmark appears for a &quot;compressed&quot; genotype and OMIM term for each OMG of the gene with mouse disease model annotations.  Conditional genotypes are NOT compressed with other kinds of genotypes. (sto196)<br/>
    		(Note: There are no conditional genotypes having the same allele pair as a non-conditional genotype, so cannot verify that these remain separated.)
 			<p concordion:manual="true">klf 11/04/2013</p> <i>(semi-automated)</i>
    		<ol>
                <li>Open the HDP query form</li>
                <li>Enter <b concordion:set="#pheno">131200</b> (Endometriosis, Susceptibility to, 1) in the Phenotypes and Diseases section and hit search</li>
                <li>Click the checkmark at the intersection of the <b concordion:set="#row">Kras</b> row and <b concordion:set="#column">Endometriosis, Susceptibility to, 1</b> column. </li>
                <li>Confirm that <b>1</b> genotype row is displayed for 
                	<b cc:assertValueAppearsOnlyOnce="diseaseGridPopupGenoClustersByPheno(#pheno,#row,#column)">Kras&lt;tm4Tyj&gt;/Kras&lt;+&gt;</b> (1 genotype exists).</li>
                <li>Confirm that the text "<b>(conditional)</b>" appears following the genotype.  (<span concordion:manual="true">klf</span>)</li>
                <li>Confirm that human <b cc:assertNoResultsContain="diseaseGridPopupHumanMarkersByPheno(#pheno,#row,#column)">KRAS</b> IS NOT LISTED.</li>
    		</ol>
    		</li>
    		
    		<li>The only diseases displayed either match the queried term OR a share a geno-cluster with a phenotype/disease matching the queried term. The only genotypes displayed are associated directly with the queried term. Additional genotypes of a marker are excluded. (sto185)
 			<p concordion:manual="true">klf 10/29/2013</p> <i>(semi-automated)</i>
    		<ol>
                <li>Open the HDP query form</li>
                <li>Enter <b concordion:set="#pheno">270150</b> (Sjogren Syndrome) in the Phenotypes and Diseases section and <b>Aire</b> in the Genes section and hit search.</li>
                <li>Verify the only diseases listed are <b>Rheumatoid Arthritis</b> and <b>Autoimmune Polyendocrine Syndrome, Type I; APS1</b> (which shares a genotype with Sjogren Syndrome).</li>
                <li>Verify <b>Guillain-Barre Syndrome, Familial; GBS</b> is NOT listed (annotated to a different genotype of Aire).</li>
                <li>Click the checkmark or colored cell for <b concordion:set="#row">Aire</b> and <b concordion:set="#column">Rheumatoid Arthritis</b>; 
                	verify the genotype in the Disease Popup is <b cc:assertResultsContain="diseaseGridPopupGenoClustersByPheno(#pheno,#row,#column)">Aire&lt;tm1Mmat&gt;/Aire&lt;tm1Mmat&gt;</b>.</li>
                <li>Click the checkmark or colored cell for 
                	<b concordion:set="#row">Aire</b> and <b concordion:set="#column">Autoimmune Polyendocrine Syndrome, Type I, with or without Reversible Metaphyseal Dysplasia; APS1</b>; 
                	verify the same genotype appears (<b cc:assertResultsContain="diseaseGridPopupGenoClustersByPheno(#pheno,#row,#column)">Aire&lt;tm1Mmat&gt;/Aire&lt;tm1Mmat&gt;</b>) in the Disease Popup.</li>
    		</ol>
    		</li>

    		<li>Verify that Chimeric Genotypes are followed by the word <b>(Chimeric)</b> (sto196)
 			<p concordion:manual="true">klf 11/04/2013</p> <i>(semi-automated)</i>
    		<ol>
                <li>Open the HDP query form</li>
                <li>Enter <b concordion:set="#pheno">135100</b> (Fibrodysplasia Ossificans Progressiva; FOP) in the Phenotypes and Diseases section and hit search</li>
                <li>Click the checkmark at the intersection of the <b concordion:set="#row">Acvr1</b> row and <b concordion:set="#column">Fibrodysplasia Ossificans Progressiva; FOP</b> column. </li>
                <li>Confirm that <b>1</b> genotype row is displayed for 
                	<b cc:assertValueAppearsOnlyOnce="diseaseGridPopupGenoClustersByPheno(#pheno,#row,#column)">Acvr1&lt;tm1Emsh&gt;/Acvr1&lt;+&gt;</b> and is followed by <b>(Chimeric)</b>.</li>
            </ol>
            </li>
                
            <li>Verify that Conditional Genotypes are followed by the word <b>(Conditional)</b> (sto196)
 			<p concordion:manual="true">klf 11/04/2013</p> <i>(semi-automated)</i>
    		<ol>
                <li>Open the HDP query form</li>
                <li>Enter <b concordion:set="#pheno">131200</b> (Endometriosis, Susceptibility to, 1) in the Phenotypes and Diseases section and hit search</li>
                <li>Click the checkmark at the intersection of the <b concordion:set="#row">Kras</b> row and <b concordion:set="#column">Endometriosis, Susceptibility to, 1</b> column. </li>
                <li>Confirm that <b>1</b> genotype row is displayed for 
                	<b cc:assertValueAppearsOnlyOnce="diseaseGridPopupGenoClustersByPheno(#pheno,#row,#column)">Kras&lt;tm4Tyj&gt;/Kras&lt;+&gt;</b> and this genotype is followed by <b>(Conditional)</b>.</li>
            </ol>
            </li>
            
            <li>Verify that disease "clumps" in the main grid are expanded into all disease terms and associated with all mouse genotypes and human genes in the popup when querying by gene. (sto117)
 			<p concordion:manual="true">klf 11/04/2013</p> <i>(semi-automated)</i>
    		<ol>
                <li>Open the HDP query form</li>
                <li>Enter <b concordion:set="#genes">Col1a1</b> in the Genes section and hit search.</li>
                <li>Confirm that the Grid displays the disease column <b>Osteogenesis Imperfecta</b>.</li>
                <li>Click the checkmark or colored cell at the intersection of the <b concordion:set="#row">Col1a1</b> row and <b concordion:set="#column">Osteogenesis Imperfecta</b> column (5 genotype annotations to 3 disease terms). </li>
                <li>Confirm that Popup displays <b>4</b> columns 
                	for <b>Osteogenesis Imperfecta, Type I</b>, <b>Osteogenesis Imperfecta, Type II</b>, <b>Osteogenesis Imperfecta, Type III</b>, and <b>Osteogenesis Imperfecta, Type IV</b>.</li>
                <li>Confirm that <b>1</b> genotype row is displayed for <b concordion:set="#genoRow">Col1a1&lt;Aga2&gt;/Col1a1&lt;+&gt;</b> 
                	and for disease terms Osteogenesis Imperfecta, Type II (<b concordion:set="#diseaseCol">166210</b>) there is a 
                		<b concordion:assertEquals="diseaseGridPopupCheckByGene(#genes,#row,#column,#genoRow,#diseaseCol)">check</b>, 
                	and Osteogenesis Imperfecta, Type III (<b concordion:set="#diseaseCol">259420</b>) there is a 
                		<b concordion:assertEquals="diseaseGridPopupCheckByGene(#genes,#row,#column,#genoRow,#diseaseCol)">check</b>.</li>
                <li>Confirm that <b>1</b> genotype row is displayed for <b concordion:set="#genoRow">Col1a1&lt;tm1Jcm&gt;/Col1a1&lt;+&gt;</b> 
                	and for disease term Osteogenesis Imperfecta, Type II (<b concordion:set="#diseaseCol">166210</b>) there is a 
                		<b concordion:assertEquals="diseaseGridPopupCheckByGene(#genes,#row,#column,#genoRow,#diseaseCol)">check</b>.</li>
                <li>Confirm that <b>1</b> gentoype row is displayed for <b concordion:set="#genoRow">Col1a1&lt;tm1.1Jcm&gt;/Col1a1&lt;+&gt;</b>
                	 and for disease term Osteogenesis Imperfecta, Type I (<b concordion:set="#diseaseCol">166200</b>) there is a 
                	 	<b concordion:assertEquals="diseaseGridPopupCheckByGene(#genes,#row,#column,#genoRow,#diseaseCol)">check</b>.</li>
                <li>Confirm that <b>1</b> gentoype row is displayed for <b concordion:set="#genoRow">Col1a1&lt;Mov13&gt;/Col1a1&lt;+&gt;</b>
                	 and for disease term Osteogenesis Imperfecta, Type I (<b concordion:set="#diseaseCol">166200</b>) there is a 
                	 	<b concordion:assertEquals="diseaseGridPopupCheckByGene(#genes,#row,#column,#genoRow,#diseaseCol)">check</b>.</li>               
                <li>Confirm that the human gene <b concordion:set="#humanRow">COL1A1</b> is displayed 
                	and for disease terms Osteogenesis Imperfecta, Type I (<b concordion:set="#diseaseCol">166200</b>) there is a 
                	 	<b concordion:assertEquals="diseaseGridPopupHumanCheckByGene(#genes,#row,#column,#humanRow,#diseaseCol)">check</b>,
                	and Osteogenesis Imperfecta, Type II (<b concordion:set="#diseaseCol">166210</b>) there is a 
                		<b concordion:assertEquals="diseaseGridPopupHumanCheckByGene(#genes,#row,#column,#humanRow,#diseaseCol)">check</b>, 
                	and Osteogenesis Imperfecta, Type III (<b concordion:set="#diseaseCol">259420</b>) there is a 
                		<b concordion:assertEquals="diseaseGridPopupHumanCheckByGene(#genes,#row,#column,#humanRow,#diseaseCol)">check</b>, 
                	and Osteogenesis Imperfecta, Type IV (<b concordion:set="#diseaseCol">166220</b>) there is a 
                		<b concordion:assertEquals="diseaseGridPopupHumanCheckByGene(#genes,#row,#column,#humanRow,#diseaseCol)">check</b>.</li>
            </ol>
            </li>

            <li>Verify that disease "clumps" in the main grid are expanded into disease terms matching the query and associated with mouse genotypes and human genes matching the query in the popup when querying by disease term. (sto117)
 			<p concordion:manual="true">klf 11/04/2013</p> <i>(semi-automated)</i>
    		<ol>
                <li>Open the HDP query form</li>
                <li>Enter <b concordion:set="#pheno">259420</b> (Osteogenesis Imperfecta, Type III) in the Phenotypes and Diseases section and hit search.</li>
                <li>Confirm that the Grid displays the disease column <b>Osteogenesis Imperfecta</b>.</li>
                <li>Click the checkmark or colored cell at the intersection of the <b concordion:set="#row">Col1a1</b> row and <b concordion:set="#column">Osteogenesis Imperfecta</b> column (1 genotype is annotated to this disease term, and another term in this disease "clump"). </li>
                <li>Confirm that Popup displays <b>2</b> columns for <b>Osteogenesis Imperfecta, Type II</b>,  and <b>Osteogenesis Imperfecta, Type III</b>.</li>
                <li>Confirm that <b>1</b> genotype row is displayed for <b concordion:set="#genoRow">Col1a1&lt;Aga2&gt;/Col1a1&lt;+&gt;</b> 
                	and for disease terms Osteogenesis Imperfecta, Type II (<b concordion:set="#diseaseCol">166210</b>) there is a 
                		<b concordion:assertEquals="diseaseGridPopupCheckByPhenotype(#pheno,#row,#column,#genoRow,#diseaseCol)">check</b>, 
                	and for Osteogenesis Imperfecta, Type III (<b concordion:set="#diseaseCol">259420</b>) there is a 
                		<b concordion:assertEquals="diseaseGridPopupCheckByPhenotype(#pheno,#row,#column,#genoRow,#diseaseCol)">check</b>.</li>
                <li>Confirm that <b>0</b> genotype rows are displayed for <b>Col1a1&lt;tm1Jcm&gt;/Col1a1&lt;+&gt;</b> with checkmark or colored cell for disease term <b>Osteogenesis Imperfecta, Type II</b>.</li>
                <li>Confirm that <b>0</b> gentoype rows are displayed for <b>Col1a1&lt;tm1.1Jcm&gt;/Col1a1&lt;+&gt;</b> with checkmark or colored cell for disease term <b>Osteogenesis Imperfecta, Type I</b>.</li>
                <li>Confirm that <b>0</b> gentoype rows are displayed for <b>Col1a1&lt;Mov13&gt;/Col1a1&lt;+&gt;</b> with checkmark or colored cell for disease term <b>Osteogenesis Imperfecta, Type I</b>.</li>               
                <li>Confirm that the human gene <b concordion:set="#humanRow">COL1A1</b> is displayed 
                	and for disease term: Osteogenesis Imperfecta, Type III (<b concordion:set="#diseaseCol">259420</b>) there is a 
                		<b concordion:assertEquals="diseaseGridPopupHumanCheckByPhenotype(#pheno,#row,#column,#humanRow,#diseaseCol)">check</b>.</li>
            </ol>
            </li>

			<li>Verify genotypes are displayed in the correct order (sto195).
 			<p concordion:manual="true">klf 11/05/2013</p>
    		<ol>
                <li>Open the HDP query form</li>
                <li>Enter <b>300624</b> (Fragile X Mental Retardation Syndrome) in the Phenotypes and Diseases section and hit search</li>
                <li>Click the checkmark at the intersection of the <b>Fmr1</b> row and <b>Fragile X Mental Retardation Syndrome</b> column. </li>
                <li>Confirm that genotypes are listed in the following order: <span concordion:manual="true">(OK klf)</span>
                <table>
                	<tr><th>Genotype</th></tr>
					<tr><td>Fmr1tm&lt;tm1Cgr&gt;/Fmr1tm&lt;tm1Cgr&gt;</td></tr>
					<tr><td>Fmr1tm&lt;tm1.1Cidz&gt;/Y </td></tr>
					<tr><td>Fmr1tm&lt;tm1Cgr&gt;/Y </td></tr>
					<tr><td>Fmr1tm&lt;tm1Rbd&gt;/Y </td></tr>
					<tr><td>Fmr1tm&lt;tm1Usdn&gt;/Y </td></tr>
					<tr><td>Fmr1tm&lt;tm1Cidz&gt;/Fmr1tm&lt;tm1Cidz&gt;  Tg(Pcp2-cre)2Mpin/0 (conditional)</td></tr>
                </table>
                </li>
                <li>Clear the query form.<br/><br/></li>
                <li>Enter <b>300100</b> (Adrenoleukodystrophy) in the Phenotypes and Diseases section and hit search</li>
                <li>Click the checkmark at the intersection of the <b>Abcd1</b> row and <b>Adrenoleukodystrophy</b> column. </li>
                <li>Confirm that genotypes are listed in the following order: <span concordion:manual="true">(OK klf)</span>
                <table>
                	<tr><th>Genotype</th></tr>
					<tr><td> Abcd&lt;1tm1Kan&gt;/Abcd1&lt;tm1Kan &gt;</td></tr>
					<tr><td> Abcd1&lt;tm1Kds&gt;/Abcd1&lt;tm1Kds &gt;</td></tr>
					<tr><td> Abcd1&lt;tm1Ymd&gt;/Abcd1&lt;tm1Ymd &gt;</td></tr>
					<tr><td> Abcd1&lt;tm1Kan&gt;/Y </td></tr>
					<tr><td> Abcd1&lt;tm1Kds&gt;/Y </td></tr>
					<tr><td> Abcd1&lt;tm1Ymd&gt;/Y </td></tr>
                </table>
                </li>
                <li>Clear the query form.<br/><br/></li>
                <li>Enter <b>300755</b> (Agammaglobulinemia, X-Linked; XLA) in the Phenotypes and Diseases section and hit search</li>
                <li>Click the checkmark at the intersection of the <b>Btk</b> row and <b>Agammaglobulinemia</b> column. </li>
                <li>Confirm that genotypes are listed in the following order: <span concordion:manual="true">(OK klf)</span>
                <table>
                	<tr><th>Genotype</th></tr>
					<tr><td>Btk&lt;tm1Wk&gt;/Btk&lt;tm1Wk&gt; </td></tr>
					<tr><td>Btk&lt;xid&gt;/Btk&lt;xid&gt; </td></tr>
					<tr><td>Btk&lt;tm1Gs&gt;/Y </td></tr>
					<tr><td>Btk&lt;xid&gt;/Y </td></tr>
                </table>
                </li>
            </ol>
			</li>

    		<li>Bin colors correctly represent the number of annotations to a term for a genotype cluster (sto171)
 			<p concordion:manual="true">klf 11/12/13</p> 
    		<ol>
                <li>Open the HDP query form.</li>
                <li>Enter <b>Lyst Ift88 Npc1 Smpd1 Traf3ip2 Aire Tg(Th-MYCN)41Waw Myo7a</b> in the Genes section and hit search.</li>
                <li>On the grid tab, click the checkmark at the intersection of the Gene and Disease Cluster name specified in the table below.</li>
                <li>Verify that the following terms have the correct annotation count bin colors:
                <table>
                	<tr><th>Gene</th><th>Cluster name</th><th>Disease term</th><th>Genotype</th><th>Bin color number</th><th>(Count of annotations)</th></tr>
					<tr><td>Aire</td>	<td>Guillain-Barre Syndrome, Familial; GBS</td>	<td>Guillain-Barre Syndrome, Familial; GBS</td>	<td>Aire&lt;tm1Mand&gt;/Aire&lt;+&gt;</td>	<td>Bin 1</td>	<td>1</td></tr>
					<tr><td>Aire</td>	<td>Autoimmune Diseases</td>	<td>Autoimmune Polyendocrine Syndrome, Type I; APS1</td>	<td>Aire&lt;tm1.1Doi&gt;/Aire&lt;tm1.1Doi&gt;</td>	<td>Bin 2</td>	<td>5</td></tr>
					<tr><td>Ift88</td>	<td>Polycystic Kidney Disease</td>	<td>Polycystic Kidney Disease, Autosomal Recessive; ARPKD</td>	<td>Ift88&lt;Tg737Rpw&gt;/Ift88&lt;Tg737Rpw&gt;</td>	<td>Bin 3</td>	<td>9</td></tr>
					<tr><td>Lyst</td>	<td>Storage Pool Platelet Disease</td>	<td>Storage Pool Platelet Disease</td>	<td>Lyst&lt;bg-J&gt;/Lyst&lt;bg-J&gt;</td>	<td>Bin 1</td>	<td>1</td></tr>
					<tr><td>Lyst</td>	<td>Chediak-Higashi Syndrome; CHS</td>	<td>Chediak-Higashi Syndrome; CHS</td>	<td>Lyst&lt;bg&gt;/Lyst&lt;bg&gt;</td>	<td>Bin 3</td>	<td>9</td></tr>
					<tr><td>Myo7a</td>	<td>Usher Syndrome</td>	<td>Usher Syndrome, Type I; USH1</td>	<td>Myo7a&lt;816SB&gt;/Myo7a&lt;816SB&gt;</td>	<td>Bin 2</td>	<td>2</td></tr>
					<tr><td>Myo7a</td>	<td>Usher Syndrome</td>	<td>Usher Syndrome, Type I; USH1</td>	<td>Myo7a&lt;sh1&gt;/Myo7a&lt;sh1&gt;</td>	<td>Bin 2</td>	<td>2</td></tr>
					<tr><td>Npc1</td>	<td>Niemann-Pick Disease</td>	<td>Niemann-Pick Disease, Type C1; NPC1</td>	<td>Npc1&lt;m1N&gt;/Npc1&lt;m1N&gt;</td>	<td>Bin 3</td>	<td>9</td></tr>
					<tr><td>Smpd1</td>	<td>Niemann-Pick Disease</td>	<td>Niemann-Pick Disease, Type A</td>	<td>Smpd1&lt;tm1Esc&gt;/Smpd1&lt;tm1Esc&gt;</td>	<td>Bin 3</td>	<td>6</td></tr>
					<tr><td>Smpd1</td>	<td>Niemann-Pick Disease</td>	<td>Niemann-Pick Disease, Type B</td>	<td>Smpd1&lt;tm1Esc&gt;/Smpd1&lt;tm1Esc&gt;</td>	<td>Bin 3</td>	<td>6</td></tr>
					<tr><td>SMPD1</td>	<td>Niemann-Pick Disease</td>	<td>Niemann-Pick Disease, Type A</td>	<td>(none - human gene)</td>	<td>Human Bin 1</td>	<td>1</td></tr>
					<tr><td>SMPD1</td>	<td>Niemann-Pick Disease</td>	<td>Niemann-Pick Disease, Type B</td>	<td>(none - human gene)</td>	<td>Human Bin 1</td>	<td>1</td></tr>
					<tr><td>Tg(Th-MYCN)41Waw</td>	<td>Neuroblastoma</td>	<td>Neuroblastoma, Susceptibility to</td>	<td>Tg(Th-MYCN)41Waw/0</td>	<td>Bin 2</td>	<td>3</td></tr>
					<tr><td>Traf3ip2</td>	<td>Dermatitis, Atopic</td>	<td>Dermatitis, Atopic</td>	<td>Traf3ip2&lt;adjm&gt;/Traf3ip2&lt;adjm&gt;</td>	<td>Bin 2</td>	<td>5</td></tr>
                </table>
                </li>
            </ol>
    		</li>
    		
    		<li>Correct onMouseover behavior for the genotype cluster rows...
 			<p concordion:manual="true">klf 11/12/13</p> 
    		<ol>
                <li>Entire row turns pale yellow color.</li>
                <li>Tooltip says "click row to see phenotype details".</li>
                <li>Row separating human genes from mouse genotypes says "Human Gene" in boldface and has a medium-gray background.</li>
                <li>Tooltip for human gene row says "click gene for homology details"</li>
            </ol>
    		</li>

    		<li>Correct Popup Window Header look and text...
 			<p concordion:manual="true">klf 11/12/13</p> 
    		<ol>
                <li>Text says "Data for <b><i>Human gene symbol</i></b>, <b><i>Mouse gene symbol</i></b> and <b><i>disease cluster name</i></b>. (No bold or italics - that is for this test case only.)</li>
                <li>Text is larger font on a light blue background with a black border.</li>
            </ol>
    		</li>

			<li>A link to IMSR for all strains or lines available is displayed for each "causal" mouse marker in the popup that has such a link on its Allele Detail Page.
 			
 			(sto136)<p concordion:manual="true">passed. -klf 2014-04-22</p> 
 			<ol>
 				<li>On the HMDC Portal, query for  gene: <b>G6pc</b>.</li>
 				<li>From the Grid Tab, click the cell in the row for G6pc and the column for Glycogen Storage Disease</li>
 				<li>Verify that the pop-up contains IMSR links for <b>G6pc</b>.  Verify that the count of "strains or lines" is the same as on an Allele Detail Page for those markers (2 as of 4/18/2014).<br/>
 					(<em>Note: no link is necessary for markers of Recombinase alleles.</em>)</li>
 				<li>Reset the query. Query for disease: <b>151400</b> (Leukemia, Chronic Lymphocytic; CLL).</li>
 				<li>From the Grid Tab, click the cell in the row for Del(14Trim13-Rnaseh2b)6Rdf and the column for Leukemia, Chronic Lymphocytic.</li> 
 				<li>Verify that a count of "strains or lines" for <b>Del(14Trim13-Rnaseh2b)6Rdf</b> does not appear (there are no strains at IMSR as of 4/18/2014).</li>
 			</ol>
			</li> 

		</ol>
	</div>


	<div id="auto">
		<h3>Automated tests</h3>
		<ol>
    		<li>Happy Path:  In the popup, a checkmark appears for a &quot;compressed&quot; genotype and OMIM term for each OMG/simple  genotype of the particular gene with a mouse model annotation.<br/>
    			Annotations to different genotypes which aggregate to one "compressed" genotype are displayed only with the single, "compressed" genotype. <br/>
    			Verify human disease is listed only when association exists AND disease name or synonym matches queried term (if pheno/disease query input).
 			<p><i>fully automated</i></p>
    		<ol>
                <li>Open the HDP query form</li>
                <li>Enter <b concordion:set="#pheno">601665</b> (Obesity) in the Phenotypes and Diseases section and hit search.</li>
                <li>Click the checkmark at the intersection of the <b concordion:set="#row">Lepr</b> row and <b concordion:set="#column">Obesity</b> column.</li>
                <li>Confirm that <b>1</b> genotype is displayed for <b cc:assertValueAppearsOnlyOnce="diseaseGridPopupGenoClustersByPheno(#pheno,#row,#column)">Lepr&lt;db&gt;/Lepr&lt;db&gt;</b>.</li>
                <li>Confirm that the pop-up contains AT LEAST the following data (homozygous genotypes):
                	<table>
                	<tr><th>genotype</th><th>term</th><th>checkmark?</th><th>notes: count of genotypes aggregated</th></tr>
                	<tr><td cc:assertResultsContain="diseaseGridPopupGenoClustersByPheno(#pheno,#row,#column)">Lepr&lt;db&gt;/Lepr&lt;db&gt;</td><td>601665 Obesity</td><td>yes</td><td>4</td></tr>
                	</table>
                </li>
                <li>Confirm that human <b cc:assertNoResultsContain="diseaseGridPopupHumanMarkersByPheno(#pheno,#row,#column)">LEPR</b> is NOT LISTED.</li>
                <li>Clear the query form.</li>
                <li>Enter <b concordion:set="#pheno">105830</b> (Angelman Syndrome; AS) in the Phenotypes and Diseases section and hit search.</li>
                <li>Click the checkmark at the intersection of the <b concordion:set="#row">Ube3a</b> row and <b concordion:set="#column">Angelman Syndrome; AS</b> column.</li>
                <li>Confirm that <b>1</b> genotype is displayed for <b cc:assertValueAppearsOnlyOnce="diseaseGridPopupGenoClustersByPheno(#pheno,#row,#column)">Ube3a&lt;tm1Alb&gt;/Ube3a&lt;+&gt;</b>.</li>
                <li>Confirm that the pop-up contains AT LEAST the following data (heterozygous genotypes):
                	<table>
                	<tr><th>genotype</th><th>term</th><th>checkmark?</th><th>notes: count of genotypes aggregated</th></tr>
                	<tr><td  cc:assertResultsContain="diseaseGridPopupGenoClustersByPheno(#pheno,#row,#column)">Ube3a&lt;tm1Alb&gt;/Ube3a&lt;+&gt;</td><td>601665 Obesity</td><td>yes</td><td>2</td></tr>
                	</table>
                </li>
                <li>Confirm that human <b cc:assertResultsContain="diseaseGridPopupHumanMarkersByPheno(#pheno,#row,#column)">UBE3A</b> IS LISTED.</li>
                <li>(Future Note: Amelogenesis Imperfecta, Type IE; AI1E (OMIM ID: 301200) has a transgene hom NOT model - it isn't a cluster of multiple genotypes, but it is a NOT. Tg(Amelx)42Msnd/Tg(Amelx)42Msnd)</li>
                <li>(Future Note: Paragangliomas 4; PGL4 (OMIM ID: 115310) has 2 simple het NOT models that should form one geno-cluster row: Sdhd&lt;tm1Jlob&gt;/Sdhd&lt;+&gt;)</li>
    		</ol>
    		</li>

    		<li>Hemizygous genotypes: In the popup, a checkmark appears for a &quot;compressed&quot; genotype and OMIM term for each OMG of the particular gene with a mouse model annotation.
 			<p><i>fully-automated</i></p> 
    		<ol>
                <li>Open the HDP query form</li>
                <li>Enter <b concordion:set="#pheno">300624</b> (Fragile X Mental Retardation Syndrome) in the Phenotypes and Diseases section and hit search</li>
                <li>Click the checkmark at the intersection of the <b concordion:set="#row">Fmr1</b> row and <b concordion:set="#column">Fragile X Mental Retardation Syndrome</b> column. </li>
                <li>Confirm that <b>1</b> genotype row is displayed for <b cc:assertValueAppearsOnlyOnce="diseaseGridPopupGenoClustersByPheno(#pheno,#row,#column)">Fmr1&lt;tm1Cgr&gt;/Y</b> (2 genotypes exist).</li>
                <li>Confirm that human <b cc:assertResultsContain="diseaseGridPopupHumanMarkersByPheno(#pheno,#row,#column)">FMR1</b> IS LISTED.</li>
                <li>Clear the query form.</li>
                <li>Enter <b concordion:set="#pheno">152700</b> (Systemic Lupus Erythematosus; SLE) in the Phenotypes and Diseases section and hit search</li>
                <li>Click the checkmark at the intersection of the <b concordion:set="#row">Yaa</b> row and <b concordion:set="#column">Systemic Lupus Erythematosus</b> column. </li>
                <li>Confirm that <b>1</b> genotype row is displayed for <b cc:assertValueAppearsOnlyOnce="diseaseGridPopupGenoClustersByPheno(#pheno,#row,#column)">X/Yaa</b> (5 genotypes exist).</li>
                <li>Clear the query form.</li>
                <li>Enter <b concordion:set="#pheno">104300</b> (Alzheimer Disease; AD) in the Phenotypes and Diseases section and hit search</li>
                <li>Click the checkmark at the intersection of the <b concordion:set="#row">Tg(APPswe,PSEN1dE9)85Dbo</b> row and <b concordion:set="#column">Alzheimer Disease</b> column. </li>
                <li>Confirm that <b>1</b> genotype row is displayed for <b cc:assertValueAppearsOnlyOnce="diseaseGridPopupGenoClustersByPheno(#pheno,#row,#column)">Tg(APPswe,PSEN1dE9)85Dbo/0</b> (3 genotypes that are disease models exist).</li>
                <li>(Future Note: Ichthyosis, X-Linked; XLI (OMIM ID: 308100) has 2 hemizygous NOT models that should form one geno-cluster row: Foxp3&lt;sf&gt;/Y)</li>
    		</ol>
    		</li>

		</ol>
 	</div>
</div>

</body>
</html>