<html xmlns:concordion="http://www.concordion.org/2007/concordion" xmlns:cc="http://fewi.custom.commands.FewiExtensions">
<head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<style>
	table {border-collapse: collapse; padding-bottom:5px; margin-bottom:5px;}
	table, th, td {border: 1px solid #CCCCCC;}
	td, th {padding:2px;}
	.man {background-color:#FFFFCC;}
	ul {padding: 0px; padding-bottom:3px; padding-top:0px; margin: 0px;}
	ol {padding: 0px; padding-bottom:3px; padding-top:0px; margin: 0px;}
	li {padding-left:2px; padding-top:0px; padding-bottom:0px; margin-left:25px;}
	p {margin:2px 0px;}
</style>
<!-- declare strict doctype to look for disallowed html errors
 **  **  **  **  ** REMOVE THE LINE before committing the file **  **  **  **  **  **  **
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
 **  **  **  **  ** REMOVE THE LINE before committing the file **  **  **  **  **  **  **
-->


    <title>Marker Summary from Query</title>
</head>


<body>

<div id="TOC">
	<p>Contents: <a href="#maintitle">Title</a> | <a href="#acceptance">Acceptance Criteria</a> | <a href="#definitions">Definitions</a> | <a href="#logic">Logic Summary</a> | <a href="#manual">Manual Tests</a> | <a href="#auto">Automated Tests</a></p>
</div>

<div id="maintitle">
    <h1>Marker Summary from Query</h1>
    <ul>
        <li>User story 163, E9: MQ&amp;S: Search Functionality (MGI 5.18)</li>
    </ul>
</div>


<div id="acceptance">
    <h2>Acceptance Criteria</h2>
	<ul>
		<li>Small differences due to using a different text search engines are OK. Determining if a difference if "small" is mainly a judgement call by the SME, I think.</li>
		<li>nomen search has no operator - as in allele and GXD forms</li>
		<li>verify GO and InterPro work ok</li>
		<li>add pheno system selection box</li>
		<li>other fields should give exactly the same results, e.g., a query by coordinates, or GO term	</li>
	</ul>
</div>


<div id="definitions">
	<h2>Definitions</h2>
	<dl>
		<dt>Term</dt>
		<dd>definition line 1</dd>
		<dd>definition line 2</dd>
	</dl>
</div>


<div id="logic">
	<h2>Logic Summary</h2>

	<h3>Querying in Text Fields</h3>
	<p>Description of logic used in each of the text field inputs when a query is submitted:</p>

	<h4>Nomenclature</h4>
	<ul>
		<li>Uses primarily the same logic as <a href="../allele/QueryFormNomenclature.html">Allele Query and Summary.</a></li>
		<li>Does not include search against allele synonyms.</li>
		<li>Initial and trailing wildcard symbols (asterisk) are supported; "internal" asterisks are treated literally.</li>
		<li>Tokens are combined with Boolean AND. </li>
	</ul>

	<h4>GO and Protein Domain (InterPro)</h4>
	<ul>
		<li>Performs a tokenized query against Solr index.</li>
		<li>Initial and trailing wildcard symbols (asterisk) are supported.</li>
	</ul>

	<h4>Pheno/Disease</h4>
	<ul>
		<li>Performs a tokenized query against Solr index.</li>
		<li>Initial and trailing wildcard symbols (asterisk) are supported.</li>
		<li>Returns only genes having annotated alleles in One Marker Genotypes (OMG) or Conditional genotypes. EXCLUDES markers for annotated alleles present only in Complex genotypes.</li>
	</ul>
</div>




<div id="tests">
    <h2>Tests</h2>
    <p>Skip to <a href="#manual">Manual Tests</a> |  <a href="#auto">Automated Tests</a></p>
	<div id="manual">
<!-- To insert manual test results, use the following:
<p concordion:assertTrue="true">testing assertTrue - displays as green</p>
<p concordion:assertTrue="false">testing assertFalse - displays as red</p>
 -->
		<h3>Manual tests</h3>
		
		<h4>Nomenclature</h4>
		<ol>
    		<li>Current, old, and interim mouse marker symbols, names, and synonyms  and human ortholog symbols and names are supported.
 			<p concordion:assertTrue="false">new</p>
    		<ol>
                <li>Navigate to the marker query form. Search for the following and verify the indicated results are returned.</li>
                <li>
                <table>
								<tr>
									<th>Input<br /> <span style="color: #FFFFFF;">ForceCellToBeWider</span></th>
									<th>Include in return<br/>(may be others)</th>
									<th>Exclude from return</th>
									<th>Notes to tester</th>
								</tr>
								<tr>
									<td>Pax6</td>
									<td>
										<ul>
											<li>Pax6</li>
											<li>Pauper</li>
											<li>Del(2)Pax6<sup>11Neu</sup></li>
											<li>Tg(CAG-EGFP,-Pax6*5a,-lacZ)1Stoy</li>
										</ul>
									</td>
									<td>
										<ul>
											<li>Pax6os1</li>
											<li>Elp4</li>
										</ul>
									</td>
									<td>Pax6os1 and Elp4 are no longer returned because we no
										longer do default "contains" searching within tokens. These
										markers are returned via Current Symbol, Current Name and
										Synonym.</td>
								</tr>
								<tr>
									<td>Defb1</td>
									<td>Defb1</td>
									<td></td>
									<td>This is an interim marker, searching by Symbol.</td>
								</tr>
								<tr>
									<td>pentamer</td>
									<td>Ppnr</td>
									<td></td>
									<td>This is an interim marker, searching by a token in the
										Name.</td>
								</tr>
								<tr>
									<td>oral cancer</td>
									<td>
										<ul>
											<li>Cdk2ap1</li>
											<li>Oraov1</li>
										</ul>
									</td>
									<td></td>
									<td>Cdk2ap1 is returned via Old Name</td>
								</tr>
								<tr>
									<td>asip</td>
									<td>
										<ul>
											<li>a</li>
											<li>Pard3</li>
										</ul>
									</td>
									<td></td>
									<td>Ortholog has different symbol.</td>
								</tr>
								<tr>
									<td>&quot;CD3-TCR complex&quot;</td>
									<td>
										<ul>
											<li>Cd3d</li>
											<li>Cd3e</li>
											<li>Cd3g</li>
										</ul>
									</td>
									<td></td>
									<td>Query using quotation marks. Ortholog has different
										name.</td>
								</tr>
								<tr>
									<td>"RT1-Db"</td>
									<td>H2-Eb1</td>
									<td></td>
									<td>Rat ortholog.</td>
								</tr>
							</table>
						</li>
    		</ol>
    		</li>
    		
    		<li>Current allele symbols or names are supported.  Allele synonyms are NOT supported.
 			<p concordion:assertTrue="false">new</p>
    		<ol>
                <li>Navigate to the marker query form. Search for the following and verify the indicated results are returned.</li>
                <li>
                <table>
								<tr>
									<th>Input<br /> <span style="color: #FFFFFF;">ForceCellToBeWider</span></th>
									<th>Include in return<br/>(may be others)</th>
									<th>Exclude from return</th>
									<th>Notes to tester</th>
								</tr>
								<tr>
									<td>Gena265</td>
									<td></td>
									<td>Pax6</td>
									<td>Gena265 is a synonym for the allele Pax6<sup>Leca3</sup>.
									</td>
								</tr>
								<tr>
									<td>leprdb</td>
									<td></td>
									<td>Lepr</td>
									<td>leprdb is an allele synonym for Lepr<sup>db</sup>.
									</td>
								</tr>
								<tr>
									<td>subcapsularis</td>
									<td></td>
									<td>Mip</td>
									<td>Allele Synonym : cataracta hereditaria subcapsularis for Mip<sup>cat</sup>.</td>
								</tr>
								<tr>
									<td>corneal adhesion</td>
									<td>Pax6</td>
									<td></td>
									<td>Allele name for Pax6<sup>Leca3</sup>.
									</td>
								</tr>
								<tr>
									<td>Leca3</td>
									<td>Pax6</td>
									<td></td>
									<td>Token in the allele symbol for Pax6<sup>Leca3</sup>.
									</td>
								</tr>
							</table>
						</li>
    		</ol>
    		</li>

    		<li>Leading and trailing asterisks are treated as wildcards.  "Internal" asterisks are treated as literal characters.
 			<p concordion:assertTrue="false">new</p>
    		<ol>
                <li>Navigate to the marker query form. Search for the following and verify the indicated results are returned.</li>
                <li>
                <table>
								<tr>
									<th>Input<br /> <span style="color: #FFFFFF;">ForceCellToBeWider</span></th>
									<th>Include in return<br/>(may be others)</th>
									<th>Exclude from return</th>
									<th>Notes to tester</th>
								</tr>
								<tr>
									<td>Pax6*</td>
									<td>
										<ul>
											<li>Pax6</li>
											<li>Pax6os1</li>
											<li>Del(2)Pax6<sup>11Neu</sup></li>
											<li>Elp4</li>
											<li>Pauper</li>
											<li>Tg(Pax6-cre,GFP)1Pgr</li>
										</ul>
									</td>
									<td></td>
									<td></td>
								</tr>
								<tr>
									<td>*os*</td>
									<td>
									<ul>
										<li>Abhd11os</li>
										<li>cosa</li>
										<li>Dx6os1</li>
										<li>Mid1</li>
										<li>Os</li>
										<li>Os9</li>
										<li>Osbp</li>
										<li>Tg(Etn2*,Ptf1aos,Ptf1a)#Kymm</li>
										<li>Tg(Fos-EGFP)KR563Gsat</li>
									</ul>
									</td>
									<td></td>
									<td></td>
								</tr>
								<tr>
									<td>Cell</td>
									<td>Cell</td>
									<td>Cell</td>
									<td>Cell</td>
								</tr>
							</table>
						</li>
    		</ol>
    		</li>

    		<li>Non-alphanumeric characters other than asterisks are used to define token boundaries.
 			<p concordion:assertTrue="false">new</p>
    		<ol>
                <li>Navigate to the marker query form. Search for the following and verify the indicated results are returned.</li>
                <li>
                <table>
								<tr>
									<th>Input<br /> <span style="color: #FFFFFF;">ForceCellToBeWider</span></th>
									<th>Include in return<br />(may be others)
									</th>
									<th>Exclude from return</th>
									<th>Notes to tester</th>
								</tr>
								<tr>
									<td>Dach1-Pax6</td>
									<td>Tg(Dach1-Pax6)#Fjl</td>
									<td></td>
									<td></td>
								</tr>
								<tr>
									<td>-Pax6+Dach1</td>
									<td>Tg(Dach1-Pax6)#Fjl</td>
									<td></td>
									<td>Not a string in a real name or symbol; verifying that the tokens are ANDed and characters don't matter.</td>
								</tr>
								<tr>
									<td>Pax6*5a</td>
									<td>Tg(CAG-EGFP,-Pax6*5a,-lacZ)1Stoy</td>
									<td></td>
									<td></td>
								</tr>
								<tr>
									<td>-Pax6,-lacZ</td>
									<td>
										<ul>
											<li>Tg(CAG-EGFP,-Pax6*5a,-lacZ)1Stoy</li>
											<li>Tg(CAG-EGFP,-Pax6,-lacZ)1Stoy</li>
										</ul>
									</td>
									<td></td>
									<td></td>
								</tr>
								<tr>
									<td>BV100Gsat</td>
									<td>Tg(Pax6-EGFP)BV100Gsat</td>
									<td></td>
									<td></td>
								</tr>
								<tr>
									<td>Gsat</td>
									<td></td>
									<td>Tg(Pax6-EGFP)BV100Gsat</td>
									<td>Verifying that tokenization is not happening on character-numeric string changes.</td>
								</tr>
							</table>
						</li>
    		</ol>
    		</li>
   		</ol>
   		
   		<h4>Feature type</h4>
		<ol>
    		<li>Markers returned for the selected feature type are of that type
 			<p concordion:assertTrue="false">new</p><i>(fully-automated)</i>
    		<ol>
                <li>Select <b>polymorphic pseudogene</b> (count is 4 as of 3/25/2014) and submit the query with no other parameters.
                 <span class="hide" concordion:set="#mcvKey">7288449</span></li>
                <li>Verify <b cc:assertEquals="getMarkerCountByMcv(#mcvKey)"><b cc:dynamicData="markerCountMcv1"></b></b> markers are returned.</li>
                <li>Verify the markers returned are all of the selected type.</li>
                <li>Reopen the query and add selections for <b>RNase MRP RNA gene</b> (count=1) and <b>telomerase RNA gene</b> (count=1) and submit the query.
                	<span class="hide" concordion:set="#mcvKey">7288449,6238182,6238183</span></li>
                <li>Verify <b cc:assertEquals="getMarkerCountByMcv(#mcvKey)"><b cc:dynamicData="markerCountMcv2"></b></b> markers are returned.</li>
                <li>Verify the 2 additional markers are of the additional types just selected.</li>
    		</ol>
    		</li>
   		</ol>
   		
   		<h4>Map position</h4>
		<ol>
    		<li>Submitting a single chromosome selection returns markers for the chromosome.  Submitting multiple chromosome selections returns markers for all selected chromosomes. 
 			<p concordion:assertTrue="false">new</p><i>(fully automated)</i>
    		<ol>
                <li>Select chromosome <b concordion:set="#chr">19</b> and submit the query with no other parameters.</li>
                <li>Verify <b cc:assertEquals="getMarkerCountByChromosome(#chr)"><b cc:dynamicData="markerCountChr19"></b></b> markers are returned.</li>
                <li>Reopen the query select only chromosome <b concordion:set="#chr">Y</b> and submit the query.</li>
                <li>Verify <b cc:assertEquals="getMarkerCountByChromosome(#chr)"><b cc:dynamicData="markerCountChrY"></b></b> markers are returned.</li>
                <li>Reopen the query select chromosomes <b concordion:set="#chr">19, Y</b> and submit the query.</li>
                <li>Verify <b cc:assertEquals="getMarkerCountByChromosome(#chr)"><b cc:dynamicData="markerCountChr19andY"></b></b> markers are returned.</li>
    		</ol>
    		</li>

    		<li>Submitting a chromosome + cM position range returns markers in that range.
 			<p concordion:assertTrue="false">new</p>
 			<i>(semi-auto)</i>
    		<ol>
                <li>Select chromosome <b concordion:set="#chr">1</b> and cM range <b concordion:set="#cm">10-20</b> and submit the query with no other parameters.</li>
                <li>Verify <b>211</b> markers are returned.</li>
                <li>Verify the list includes the following markers:
                <table cc:verifySupersetOf="#symbol : getSymbolsByChrCm(#chr,#cm)">
        			<tr><th concordion:assertEquals="#symbol">Marker Symbol</th></tr>
                	<tr><td>Aff3</td></tr>
                	<tr><td>Cocia4</td></tr>
                	<tr><td>Creg2</td></tr>
                	<tr><td>D1Mit20</td></tr>
                	<tr><td>Eif5b</td></tr>
                	<tr><td>Gm4850</td></tr>
                	<tr><td>Prss40</td></tr>
                	<tr><td>Snord89</td></tr>
                	<tr><td>Zfp451</td></tr>
                </table>
                </li>
    		</ol>
    		</li>

    		<li>Submitting a chromosome + genome coordinates range returns markers in that range.
 			<p concordion:assertTrue="false">new</p>
 			<i>(semi-auto)</i>
    		<ol>
                <li>Select chromosome <b concordion:set="#chr">1</b> and genome coordinates range <b concordion:set="#coord">24332803-40772038</b> with units <b>bp</b> and submit the query with no other parameters.</li>
                <li>Verify <b>298</b> markers are returned.</li>
                <li>Verify the list includes the following markers:
                <table cc:verifySupersetOf="#symbol : getSymbolsByChrCoord(#chr,#coord)">
        			<tr><th concordion:assertEquals="#symbol">Marker Symbol</th></tr>
                	<tr><td>Aff3</td></tr>
                	<tr><td>Creg2</td></tr>
                	<tr><td>D1Mit20</td></tr>
                	<tr><td>Eif5b</td></tr>
                	<tr><td>Gm4850</td></tr>
                	<tr><td>Prss40</td></tr>
                	<tr><td>Snord89</td></tr>
                	<tr><td>Zfp451</td></tr>
                </table>
                </li>
                <li>Verify the following markers are not returned (no coordinates):
                <table cc:verifyNotRows="#symbol : getSymbolsByChrCoord(#chr,#coord)">
        			<tr><th concordion:assertEquals="#symbol">Marker Symbol</th></tr>
                	<tr><td>Cocia4</td></tr>
                	<tr><td>D1Mit412</td></tr>
                	<tr><td>Dbsty1</td></tr>
                </table>
                </li>
                <li>Reopen the query and select chromosome <b>1</b> and genome coordinates range <b>24.332-40.772</b> with units <b>Mbp</b> and submit the query with no other parameters.</li>
                <li>Verify <b>297</b> markers are returned.</li>
    		</ol>
    		</li>

    		<li>Submitting a marker range returns markers which fall between the start coordinate of the one marker and the end coordinate of the other marker.
 			<p concordion:assertTrue="false">new</p>
 			<i>(semi-auto)</i>
    		<ol>
                <li>Input <b concordion:set="#startMarker">D19Mit32</b> and <b concordion:set="#endMarker">Tbx10</b>.</li>
                <li>Verify the following markers are returned (these are sorted by genome coordinates):
                <table cc:verifySupersetOf="#symbol : getSymbolsByRange(#startMarker,#endMarker)">
        			<tr><th concordion:assertEquals="#symbol">Marker Symbol</th></tr>
                	<tr><td>Ighmbp2</td></tr>
					<tr><td>D19Mit32</td></tr>
					<tr><td>Lfp3</td></tr>
					<tr><td>Tgct2</td></tr>
					<tr><td>Mrpl21</td></tr>
					<tr><td>Gm27771</td></tr>
					<tr><td>Gm7810</td></tr>
					<tr><td>Cpt1a</td></tr>
					<tr><td>4833417J20Rik</td></tr>
					<tr><td>Mtl5</td></tr>
					<tr><td>Gal</td></tr>
					<tr><td>Ppp6r3</td></tr>
					<tr><td>Gm23940</td></tr>
					<tr><td>Gm19205</td></tr>
					<tr><td>n-Ta9</td></tr>
					<tr><td>Lrp5</td></tr>
					<tr><td>D19Mit68</td></tr>
					<tr><td>Drinksac2</td></tr>
					<tr><td>1810055G02Rik</td></tr>
					<tr><td>Gm2141</td></tr>
					<tr><td>Suv420h1</td></tr>
					<tr><td>Gm16066</td></tr>
					<tr><td>Gm19209</td></tr>
					<tr><td>Chka</td></tr>
					<tr><td>Tcirg1</td></tr>
					<tr><td>Ndufs8</td></tr>
					<tr><td>Aldh3b1</td></tr>
					<tr><td>Unc93b1</td></tr>
					<tr><td>1700055N04Rik</td></tr>
					<tr><td>Aldh3b2</td></tr>
					<tr><td>Acy3</td></tr>
					<tr><td>Tbx10</td></tr>
                </table>
                </li>
    		</ol>
    		</li>

    		<li>Submitting an invalid marker range (one or both markers are missing coordinates; markers are on different chromosomes) returns error messages.
 			<p concordion:assertTrue="false">new</p>
 			<ol>
 				<li>Input <b>Aff3</b> and <b>Tbx10</b>.</li>
 				<li>Verify result returns <b>Marker Aff3(1) not on same chromosome as Tbx10(19)</b>.</li>
 				<li>Input <b>Tbx10</b> and <b>D19Mit127</b>.</li>
 				<li>Verify result returns <b>Error: Your marker range contains at least one marker (D19Mit127) which has no genome coordinates. To query by marker range, both markers must have assigned genome coordinates.</b></li>
 			</ol>
 			</li>

		</ol>

   		<h4>Gene Ontology classifications</h4>
		<ol>
    		<li>Submitting a GO ID returns markers associated with the term and its descendants, if any.
 			<p concordion:assertTrue="false">new</p>
    		<ol>
                <li>Enter <b concordion:set="#go">GO:0060710</b> (chorio-allantoic fusion) and submit the query. (has no descendants)</li>
                <li>Verify <b>7</b> genes are returned.</li>
                <li>Verify the genes returned include:
                <table cc:verifySupersetOf="#symbol : getSymbolsByGo(#go)">
        			<tr><th concordion:assertEquals="#symbol">Marker Symbol</th></tr>
                	<tr><td>Cyr61</td></tr>
                	<tr><td>Dnajb6</td></tr>
                	<tr><td>Erf</td></tr>
                	<tr><td>Itga4</td></tr>
                	<tr><td>Lef1</td></tr>
                	<tr><td>Vcam1</td></tr>
                	<tr><td>Wnt7b</td></tr>
                </table>
                </li>
                <li>Enter <b concordion:set="#go">GO:0070486</b> (leukocyte aggregation) and submit the query. (has annotated descendants; only Sema4d is annotated to the parent term)</li>
                <li>Verify <b>11</b> genes are returned.</li>
                <li>Verify the genes returned include:
                <table cc:verifySupersetOf="#symbol : getSymbolsByGo(#go)">
        			<tr><th concordion:assertEquals="#symbol">Marker Symbol</th></tr>
                	<tr><td>Adam8</td></tr>
                	<tr><td>Bmp7</td></tr>
                	<tr><td>Cd44</td></tr>
                	<tr><td>Gstp1</td></tr>
                	<tr><td>Has2</td></tr>
                	<tr><td>Il1b</td></tr>
                	<tr><td>S100a8</td></tr>
                	<tr><td>S100a9</td></tr>
                	<tr><td>Sema4d</td></tr>
                	<tr><td>Spn</td></tr>
                	<tr><td>Stk10</td></tr>
                </table>
                </li>
    		</ol>
    		</li>

    		<li>Submitting an exact GO term returns markers associated with the term and its descendants, if any.
 			<p concordion:assertTrue="false">new</p>
    		<ol>
                <li>Enter <b concordion:set="#go">leukocyte aggregation</b> (GO:0070486) and submit the query. (has annotated descendants; only Sema4d is annotated to the parent term)</li>
                <li>Verify <b>11</b> genes are returned.</li>
                <li>Verify the genes returned include:
                <table cc:verifySupersetOf="#symbol : getSymbolsByGo(#go)">
        			<tr><th concordion:assertEquals="#symbol">Marker Symbol</th></tr>
                	<tr><td>Adam8</td></tr>
                	<tr><td>Bmp7</td></tr>
                	<tr><td>Cd44</td></tr>
                	<tr><td>Gstp1</td></tr>
                	<tr><td>Has2</td></tr>
                	<tr><td>Il1b</td></tr>
                	<tr><td>S100a8</td></tr>
                	<tr><td>S100a9</td></tr>
                	<tr><td>Sema4d</td></tr>
                	<tr><td>Spn</td></tr>
                	<tr><td>Stk10</td></tr>
                </table>
                </li>
    		</ol>
    		</li>

    		<li>Submitting an exact GO synonym returns markers associated with the term and its descendants, if any.
 			<p concordion:assertTrue="false">new</p>
    		<ol>
                <li>Enter <b>leucocyte aggregation</b> (synonym for GO:0070486) and submit the query. (has annotated descendants; only Sema4d is annotated to the parent term)</li>
                <li>Verify <b>11</b> genes are returned.</li>
                <li>Verify the genes returned include:
                <ul>
                	<li>Adam8</li>
                	<li>Bmp7</li>
                	<li>Cd44</li>
                	<li>Gstp1</li>
                	<li>Has2</li>
                	<li>Il1b</li>
                	<li>S100a8</li>
                	<li>S100a9</li>
                	<li>Sema4d</li>
                	<li>Spn</li>
                	<li>Stk10</li>
                </ul>
                </li>
    		</ol>
    		</li>

    		<li>Submitting text which appears in more than one GO term returns markers associated with each term and its descendants, if any.
 			<p concordion:assertTrue="false">new</p>
    		<ol>
                <li>Enter <b>reticulum</b> in the GO input text field and submit the search.<br/>(Appears in 3 Molecular Function, 96 Biological Process, and 64 Cellular Component terms.) </li>
                <li>Verify the genes returned include the following:
	                <table>
						<tr>
							<th>Gene</th>
							<th>Annotated Term</th>
							<th>GO sub-DAG</th>
						</tr>
						<tr>
							<td>Srp14</td>
							<td>endoplasmic reticulum signal peptide binding</td>
							<td>Molecular Function</td>
						</tr>
						<tr>
							<td>Srp54a</td>
							<td>endoplasmic reticulum signal peptide binding</td>
							<td>Molecular Function</td>
						</tr>
						<tr>
							<td>Rab10</td>
							<td>endoplasmic reticulum tubular network organization</td>
							<td>Biological Process</td>
						</tr>
						<tr>
							<td>Reep1</td>
							<td>endoplasmic reticulum tubular network organization</td>
							<td>Biological Process</td>
						</tr>
						<tr>
							<td>H13</td>
							<td>integral component of cytoplasmic side of endoplasmic reticulum membrane</td>
							<td>Cellular Component</td>
						</tr>
						<tr>
							<td>Pkd2</td>
							<td>integral component of cytoplasmic side of endoplasmic reticulum membrane</td>
							<td>Cellular Component</td>
						</tr>
					</table>
				</li>
    		</ol>
    		</li>

    		<li>Submitting text and checking a box for particular DAG(s) within the GO returns markers associated with matching terms in only the checked DAG(s).
 			<p concordion:assertTrue="false">new</p>
    		<ol>
                <li>Enter <b>reticulum</b> in the GO input text field.<br/>(Appears in 3 Molecular Function, 96 Biological Process, and 64 Cellular Component terms.) </li>
				<li>Check only the box for <b>Molecular Function</b> and submit the search.</li>
                <li>Verify the genes returned include or exclude the following, as indicated:
                <table>
					<tr>
						<th>Gene</th>
						<th>Annotated Term</th>
						<th>Included / Excluded</th>
						<th>GO sub-DAG</th>
					</tr>
					<tr>
						<td>Srp14</td>
						<td>endoplasmic reticulum signal peptide binding</td>
						<td>included</td>
						<td>Molecular Function</td>
					</tr>
					<tr>
						<td>Srp54a</td>
						<td>endoplasmic reticulum signal peptide binding</td>
						<td>included</td>
						<td>Molecular Function</td>
					</tr>
					<tr>
						<td>Rab10</td>
						<td>endoplasmic reticulum tubular network organization</td>
						<td>excluded</td>
						<td>Biological Process</td>
					</tr>
					<tr>
						<td>Reep1</td>
						<td>endoplasmic reticulum tubular network organization</td>
						<td>excluded</td>
						<td>Biological Process</td>
					</tr>
					<tr>
						<td>H13</td>
						<td>integral component of cytoplasmic side of endoplasmic
							reticulum membrane</td>
						<td>excluded</td>
						<td>Cellular Component</td>
					</tr>
					<tr>
						<td>Pkd2</td>
						<td>integral component of cytoplasmic side of endoplasmic
							reticulum membrane</td>
						<td>excluded</td>
						<td>Cellular Component</td>
					</tr>
				</table>
				</li>
                <li>Enter <b>reticulum</b> in the GO input text field.<br/>(Appears in 3 Molecular Function, 96 Biological Process, and 64 Cellular Component terms.) </li>
				<li>Check only the box for <b>Biological Process</b> and submit the search.</li>
                <li>Verify the genes returned include or exclude the following, as indicated:
                <table>
					<tr>
						<th>Gene</th>
						<th>Annotated Term</th>
						<th>Included / Excluded</th>
						<th>GO sub-DAG</th>
					</tr>
					<tr>
						<td>Srp14</td>
						<td>endoplasmic reticulum signal peptide binding</td>
						<td>excluded</td>
						<td>Molecular Function</td>
					</tr>
					<tr>
						<td>Srp54a</td>
						<td>endoplasmic reticulum signal peptide binding</td>
						<td>excluded</td>
						<td>Molecular Function</td>
					</tr>
					<tr>
						<td>Rab10</td>
						<td>endoplasmic reticulum tubular network organization</td>
						<td>included</td>
						<td>Biological Process</td>
					</tr>
					<tr>
						<td>Reep1</td>
						<td>endoplasmic reticulum tubular network organization</td>
						<td>included</td>
						<td>Biological Process</td>
					</tr>
					<tr>
						<td>H13</td>
						<td>integral component of cytoplasmic side of endoplasmic
							reticulum membrane</td>
						<td>excluded</td>
						<td>Cellular Component</td>
					</tr>
					<tr>
						<td>Pkd2</td>
						<td>integral component of cytoplasmic side of endoplasmic
							reticulum membrane</td>
						<td>excluded</td>
						<td>Cellular Component</td>
					</tr>
				</table>
				</li>
                <li>Enter <b>reticulum</b> in the GO input text field.<br/>(Appears in 3 Molecular Function, 96 Biological Process, and 64 Cellular Component terms.) </li>
				<li>Check only the box for <b>Cellular Component</b> and submit the search.</li>
                <li>Verify the genes returned include or exclude the following, as indicated:
                <table>
					<tr>
						<th>Gene</th>
						<th>Annotated Term</th>
						<th>Included / Excluded</th>
						<th>GO sub-DAG</th>
					</tr>
					<tr>
						<td>Srp14</td>
						<td>endoplasmic reticulum signal peptide binding</td>
						<td>excluded</td>
						<td>Molecular Function</td>
					</tr>
					<tr>
						<td>Srp54a</td>
						<td>endoplasmic reticulum signal peptide binding</td>
						<td>excluded</td>
						<td>Molecular Function</td>
					</tr>
					<tr>
						<td>Rab10</td>
						<td>endoplasmic reticulum tubular network organization</td>
						<td>excluded</td>
						<td>Biological Process</td>
					</tr>
					<tr>
						<td>Reep1</td>
						<td>endoplasmic reticulum tubular network organization</td>
						<td>excluded</td>
						<td>Biological Process</td>
					</tr>
					<tr>
						<td>H13</td>
						<td>integral component of cytoplasmic side of endoplasmic
							reticulum membrane</td>
						<td>included</td>
						<td>Cellular Component</td>
					</tr>
					<tr>
						<td>Pkd2</td>
						<td>integral component of cytoplasmic side of endoplasmic
							reticulum membrane</td>
						<td>included</td>
						<td>Cellular Component</td>
					</tr>
				</table>
				</li>
                <li>Enter <b>reticulum</b> in the GO input text field.<br/>(Appears in 3 Molecular Function, 96 Biological Process, and 64 Cellular Component terms.) </li>
				<li>Check only the boxes for <b>Molecular Function</b> and <b>Cellular Component</b> and submit the search.</li>
                <li>Verify the genes returned include or exclude the following, as indicated:
                <table>
					<tr>
						<th>Gene</th>
						<th>Annotated Term</th>
						<th>Included / Excluded</th>
						<th>GO sub-DAG</th>
					</tr>
					<tr>
						<td>Srp14</td>
						<td>endoplasmic reticulum signal peptide binding</td>
						<td>included</td>
						<td>Molecular Function</td>
					</tr>
					<tr>
						<td>Srp54a</td>
						<td>endoplasmic reticulum signal peptide binding</td>
						<td>included</td>
						<td>Molecular Function</td>
					</tr>
					<tr>
						<td>Rab10</td>
						<td>endoplasmic reticulum tubular network organization</td>
						<td>excluded</td>
						<td>Biological Process</td>
					</tr>
					<tr>
						<td>Reep1</td>
						<td>endoplasmic reticulum tubular network organization</td>
						<td>excluded</td>
						<td>Biological Process</td>
					</tr>
					<tr>
						<td>H13</td>
						<td>integral component of cytoplasmic side of endoplasmic
							reticulum membrane</td>
						<td>included</td>
						<td>Cellular Component</td>
					</tr>
					<tr>
						<td>Pkd2</td>
						<td>integral component of cytoplasmic side of endoplasmic
							reticulum membrane</td>
						<td>included</td>
						<td>Cellular Component</td>
					</tr>
				</table>
				</li>
    		</ol>
    		</li>
    		
		</ol>
   		<h4>Protein domains</h4>
		<ol>
    		<li>Submitting an InterPro ID returns markers associated with the term (we do not represent InterPro as a DAG or hierarchy, even though it is one).
 			<p concordion:assertTrue="false">new</p>
    		<ol>
                <li>Enter <b>IPR000001</b> (Kringle) and submit the query. (has no descendants)</li>
                <li>Verify <b>15</b> genes are returned.</li>
                <li>Verify the genes returned include:
                <table >
        			<tr><th concordion:assertEquals="#symbol">Marker Symbol</th></tr>
                	<tr><td>F12</td></tr>
                	<tr><td>F2</td></tr>
                	<tr><td>Habp2</td></tr>
                	<tr><td>Hgfac</td></tr>
                	<tr><td>Mst1</td></tr>
                	<tr><td>Plau</td></tr>
                	<tr><td>Ror1</td></tr>
                	<tr><td>Ror2</td></tr>
                </table>
                </li>
                <li>Enter <b>IPR000562</b> (Fibronectin, type II, collagen-binding) and submit the query. (has no descendants)</li>
                <li>Verify <b>13</b> genes are returned.</li>
                <li>Verify the genes returned include:
                <table >
        			<tr><th concordion:assertEquals="#symbol">Marker Symbol</th></tr>
                	<tr><td>Bsph1</td></tr>
                	<tr><td>F12</td></tr>
                	<tr><td>Hgfac</td></tr>
                	<tr><td>Ly75</td></tr>
                	<tr><td>Mmp2</td></tr>
                	<tr><td>Mmp9</td></tr>
                	<tr><td>Pla2r1</td></tr>
                	<tr><td>Sel1l</td></tr>
                </table>
                </li>
                <li>Enter <b>IPR013806</b> (Kringle-like fold) and submit the query. (has annotated descendants; is the parent term of the previous two queried terms; has no genes annotated ONLY to the parent term)</li>
                <li>Verify <b>25</b> genes are returned.</li>
                <li>Verify the genes returned include:
                <table >
        			<tr><th concordion:assertEquals="#symbol">Marker Symbol</th></tr>
                	<tr><td>Bsph1</td></tr>
                	<tr><td>F12</td></tr>
                	<tr><td>F2</td></tr>
                	<tr><td>Habp2</td></tr>
                	<tr><td>Hgfac</td></tr>
                	<tr><td>Mmp2</td></tr>
                	<tr><td>Mmp9</td></tr>
                	<tr><td>Mst1</td></tr>
                	<tr><td>Plau</td></tr>
                	<tr><td>Pla2r1</td></tr>
                	<tr><td>Ror1</td></tr>
                	<tr><td>Ror2</td></tr>
                	<tr><td>Sel1l</td></tr>
                </table>
                </li>
    		</ol>
    		</li>

    		<li>Submitting an exact InterPro term returns markers associated with the term.
 			<p concordion:assertTrue="false">new</p>
    		<ol>
                <li>Enter <b>Kringle-like fold</b> (IPR013806) and submit the query. (has annotated descendants; is the parent term of the previous two queried terms; has no genes annotated ONLY to the parent term)</li>
                <li>Verify <b>25</b> genes are returned.</li>
                <li>Verify the genes returned include:
                <table >
        			<tr><th concordion:assertEquals="#symbol">Marker Symbol</th></tr>
                	<tr><td>Bsph1</td></tr>
                	<tr><td>F12</td></tr>
                	<tr><td>F2</td></tr>
                	<tr><td>Habp2</td></tr>
                	<tr><td>Hgfac</td></tr>
                	<tr><td>Mmp2</td></tr>
                	<tr><td>Mmp9</td></tr>
                	<tr><td>Mst1</td></tr>
                	<tr><td>Plau</td></tr>
                	<tr><td>Pla2r1</td></tr>
                	<tr><td>Ror1</td></tr>
                	<tr><td>Ror2</td></tr>
                	<tr><td>Sel1l</td></tr>
                </table>
                </li>
    		</ol>
    		</li>

    		<li>Submitting text which appears in more than one InterPro term returns markers associated with any matching term(s).
 			<p concordion:assertTrue="false">new</p>
    		<ol>
                <li>Enter <b>fibronec</b> (piece of text which appears in many protein domain names which have annotated genes) and submit the query.</li>
                <li>Verify <b>231</b> genes are returned.</li>
                <li>Verify the genes returned include:
                <table >
        			<tr><th concordion:assertEquals="#symbol">Marker Symbol</th></tr>
        			<tr><td>Abi3bp</td></tr>
                	<tr><td>Bsph1</td></tr>
                	<tr><td>Epha1</td></tr>
                	<tr><td>F12</td></tr>
                	<tr><td>Hgfac</td></tr>
                	<tr><td>Mmp2</td></tr>
                	<tr><td>Mmp9</td></tr>
                	<tr><td>Mylk</td></tr>
                	<tr><td>Pla2r1</td></tr>
                	<tr><td>Prlr</td></tr>
                	<tr><td>Robo1</td></tr>
                	<tr><td>Robo2</td></tr>
                	<tr><td>Tnn</td></tr>
                </table>
                </li>
    		</ol>
    		</li>
    		
		</ol>

   		<h4>Mouse phenotypes &amp; mouse models of human disease</h4>
		<ol>
    		<li>Test
 			<p concordion:assertTrue="false">new</p>
    		<ol>
                <li>Step</li>
                <li>Step</li>
    		</ol>
    		</li>

    		<li>Test
 			<p concordion:assertTrue="false">new</p>
    		<ol>
                <li>Step</li>
                <li>Step</li>
    		</ol>
    		</li>
    		
		</ol>
	</div>

	<div id="auto">
		<h3>Automated tests</h3>
		<ol>
    		<li>Nomen Query
    		<ol>
               	<li>Testing that AND, OR, and NOT aren't treated as Booleans
				<ol>
					<li>open the Marker query form</li>
					<li>Type <b concordion:set="#query">pax6 and shh</b> in the nomenclature field and press the search button</li>
					<li>verify that <b concordion:assertEquals="getMarkerCountByNomen(#query)">0</b> matching genes are returned</li>
					<li>Type <b concordion:set="#query">pax6 or shh</b> in the nomenclature field and press the search button</li>
					<li>verify that <b concordion:assertEquals="getMarkerCountByNomen(#query)">0</b> matching genes are returned</li>
					<li>Type <b concordion:set="#query">paired AND box</b> in the nomenclature field and press the search button</li>
					<li>verify that <b concordion:assertEquals="getMarkerCountByNomen(#query)">0</b> matching genes are returned</li>
					<li>Type <b concordion:set="#query">paired OR box</b> in the nomenclature field and press the search button</li>
					<li>verify that <b concordion:assertEquals="getMarkerCountByNomen(#query)">0</b> matching genes are returned</li>
					<li>Type <b concordion:set="#query">paired NOT box</b> in the nomenclature field and press the search button</li>
					<li>verify that <b concordion:assertEquals="getMarkerCountByNomen(#query)">0</b> matching genes are returned</li>
				</ol>
				</li>
				<li>Testing that punctuation and extraneous wildcards are removed (ignored)
				<ol>
					<li>open the Marker query form</li>
					<li>Type <b concordion:set="#query">dopachrome@#$%^&amp;*()_-+={}[]\|`~;':",./&lt;&gt;?</b> in the nomenclature field and press the search button</li>
					<li>verify that only <b concordion:assertEquals="getMarkerCountByNomen(#query)">2</b> markers returned</li>
					<li>Type <b concordion:set="#query">*dopachrome ***tautome*</b> in the nomenclature field and press the search button</li>
					<li>verify that only <b concordion:assertEquals="getMarkerCountByNomen(#query)">2</b> markers is returned</li>
					<li>Type <b concordion:set="#query">*****</b> in the nomenclature field and press the search button</li>
					<li>verify that <b cc:assertEquals="getMarkerCountByNomen(#query)">0</b> matching markers are returned</li>
				</ol>
				</li>
    		</ol>
    		</li>
    		<li>Feature Type Query
    		<ol>
               	<li>Querying all feature types should bring back all genes
				<ol>
					<li>open the Marker query form</li>
					<li>Select <b>all feature types</b> and search</li>
					<li>Summary should return all <b cc:assertEquals="getMarkerCountAllMcv()"><b cc:dynamicData="allMarkers"></b></b></li>
				</ol>
				</li>
    		</ol>
    		</li>
		</ol>
 	</div>
</div>

</body>
</html>